// @medusajs/medusa/dist/api/admin/rbac/policies/[id]/roles/route.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GET = void 0;
const utils_1 = require("@medusajs/framework/utils");
const rbac_1 = __importDefault(require("../../../../../../feature-flags/rbac"));
/**
 * Lists the roles that include the given policy.
 *
 * @ignore
 * @featureFlag rbac
 */
const GET = async (req, res) => {
    const policyId = req.params.id;
    const query = req.scope.resolve(utils_1.ContainerRegistrationKeys.QUERY);
    await query.graph({
        entity: "rbac_policy",
        fields: ["id"],
        filters: { id: policyId },
    }, {
        throwIfKeyNotFound: true,
    });
    const { data: links, metadata } = await query.graph({
        entity: "rbac_role_policy",
        fields: req.queryConfig.fields,
        filters: { policy_id: policyId },
        pagination: req.queryConfig.pagination,
    });
    const roles = links
        .map((link) => link.role)
        .filter((role) => !!role);
    res.status(200).json({
        roles,
        count: metadata?.count ?? 0,
        offset: metadata?.skip ?? 0,
        limit: metadata?.take ?? 0,
    });
};
exports.GET = GET;
(0, utils_1.defineFileConfig)({
    isDisabled: () => !utils_1.FeatureFlag.isFeatureEnabled(rbac_1.default.key),
});
