// @medusajs/auth/dist/loaders/providers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const awilix_1 = require("@medusajs/framework/awilix");
const modules_sdk_1 = require("@medusajs/framework/modules-sdk");
const utils_1 = require("@medusajs/framework/utils");
const _types_1 = require("../types");
const recovery_code_1 = require("../providers/mfa/recovery-code");
const totp_1 = require("../providers/mfa/totp");
const token_1 = require("../providers/verification/token");
const medusa_cloud_auth_1 = require("../providers/medusa-cloud-auth");
const validateCloudOptions = (options) => {
    const { oauth_authorize_endpoint, oauth_token_endpoint, oauth_audience, environment_handle, sandbox_handle, api_key, callback_url, } = options ?? {};
    if (!environment_handle && !sandbox_handle) {
        return false;
    }
    if (!oauth_authorize_endpoint ||
        !oauth_token_endpoint ||
        !oauth_audience ||
        !api_key ||
        !callback_url) {
        return false;
    }
    return true;
};
const registrationFn = async (klass, container, pluginOptions) => {
    container.register({
        [_types_1.AuthProviderRegistrationPrefix + pluginOptions.id]: (0, awilix_1.asFunction)((cradle) => new klass(cradle, pluginOptions.options ?? {}), {
            lifetime: klass.LIFE_TIME || awilix_1.Lifetime.SINGLETON,
        }),
    });
    container.registerAdd(_types_1.AuthIdentifiersRegistrationName, (0, awilix_1.asValue)(pluginOptions.id));
};
const mfaRegistrationFn = async (klass, container, pluginOptions) => {
    if (!klass?.identifier) {
        throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_ARGUMENT, `Trying to register an MFA provider without a provider identifier.`);
    }
    const id = pluginOptions.id ?? klass.identifier;
    const key = _types_1.AuthMfaProviderRegistrationPrefix + id;
    container.register({
        [key]: (0, awilix_1.asFunction)((cradle) => new klass(cradle, pluginOptions.options ?? {}), {
            lifetime: klass.LIFE_TIME || awilix_1.Lifetime.SINGLETON,
        }),
    });
    container.registerAdd(_types_1.AuthMfaIdentifiersRegistrationName, (0, awilix_1.asValue)(id));
};
const verificationRegistrationFn = async (klass, container, pluginOptions) => {
    if (!klass?.identifier) {
        throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_ARGUMENT, `Trying to register a verification provider without a provider identifier.`);
    }
    const id = pluginOptions.id ?? klass.identifier;
    const key = _types_1.AuthVerificationProviderRegistrationPrefix + id;
    container.register({
        [key]: (0, awilix_1.asFunction)((cradle) => new klass(cradle, pluginOptions.options ?? {}), {
            lifetime: klass.LIFE_TIME || awilix_1.Lifetime.SINGLETON,
        }),
    });
    container.registerAdd(_types_1.AuthVerificationIdentifiersRegistrationName, (0, awilix_1.asValue)(id));
};
const getMfaProviderOptions = (options, id) => {
    return options?.mfa?.providers?.find((provider) => provider.id === id)
        ?.options;
};
const getVerificationProviderOptions = (options, id) => {
    return options?.verification?.providers?.find((provider) => provider.id === id)
        ?.options;
};
const hasProviderResolver = (provider) => {
    return !!provider.resolve;
};
const validateMfaProviderConfigs = (providers = []) => {
    const configurableProviderIds = new Set([totp_1.TotpMfaProvider.identifier]);
    const invalidProvider = providers.find((provider) => !provider.resolve && !configurableProviderIds.has(provider.id));
    if (invalidProvider) {
        throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_ARGUMENT, `MFA provider "${invalidProvider.id}" must include a provider resolver.`);
    }
};
const validateVerificationProviderConfigs = (providers = []) => {
    const configurableProviderIds = new Set([token_1.TokenVerificationProvider.identifier]);
    const invalidProvider = providers.find((provider) => !provider.resolve && !configurableProviderIds.has(provider.id));
    if (invalidProvider) {
        throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_ARGUMENT, `Verification provider "${invalidProvider.id}" must include a provider resolver.`);
    }
};
exports.default = async ({ container, options, }) => {
    validateMfaProviderConfigs(options?.mfa?.providers);
    validateVerificationProviderConfigs(options?.verification?.providers);
    const totpOptions = getMfaProviderOptions(options, totp_1.TotpMfaProvider.identifier);
    await mfaRegistrationFn(totp_1.TotpMfaProvider, container, {
        options: {
            encryption_key: options?.mfa?.encryption_key,
            ...(totpOptions ?? {}),
        },
        id: totp_1.TotpMfaProvider.identifier,
    });
    await mfaRegistrationFn(recovery_code_1.RecoveryCodeMfaProvider, container, {
        options: options?.mfa,
        id: recovery_code_1.RecoveryCodeMfaProvider.identifier,
    });
    await (0, modules_sdk_1.moduleProviderLoader)({
        container,
        providers: (options?.mfa?.providers || []).filter(hasProviderResolver),
        registerServiceFn: mfaRegistrationFn,
    });
    const tokenOptions = getVerificationProviderOptions(options, token_1.TokenVerificationProvider.identifier);
    await verificationRegistrationFn(token_1.TokenVerificationProvider, container, {
        options: tokenOptions ?? {},
        id: token_1.TokenVerificationProvider.identifier,
    });
    await (0, modules_sdk_1.moduleProviderLoader)({
        container,
        providers: (options?.verification?.providers || []).filter(hasProviderResolver),
        registerServiceFn: verificationRegistrationFn,
    });
    if (validateCloudOptions(options?.cloud) && !options?.cloud?.disabled) {
        await registrationFn(medusa_cloud_auth_1.MedusaCloudAuthService, container, {
            options: options?.cloud,
            id: "cloud",
        });
    }
    await (0, modules_sdk_1.moduleProviderLoader)({
        container,
        providers: options?.providers || [],
        registerServiceFn: registrationFn,
    });
};
