// @medusajs/auth/dist/providers/verification/token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenVerificationProvider = void 0;
const utils_1 = require("@medusajs/framework/utils");
const verification_token_1 = require("../../utils/verification-token");
class TokenVerificationProvider {
    constructor({ authVerificationService }, options = {}) {
        this.identifier = TokenVerificationProvider.identifier;
        this.authVerificationService_ = authVerificationService;
        this.options_ = options;
    }
    async request(data, sharedContext = {}) {
        const existingVerifications = await this.authVerificationService_.list({
            auth_identity_id: data.auth_identity_id,
            entity_id: data.entity_id,
            entity_type: data.entity_type,
        }, { take: 1, skip: 0 }, sharedContext);
        if (existingVerifications.length && existingVerifications[0].verified_at) {
            return existingVerifications[0];
        }
        const token = (0, verification_token_1.generateVerificationToken)();
        const tokenHash = (0, verification_token_1.hashVerificationToken)(token);
        const requestedAt = new Date(Date.now());
        const expiresAt = new Date(requestedAt.getTime() + this.getTokenTtlMs_());
        let verification;
        if (existingVerifications.length) {
            verification = await this.authVerificationService_.update({
                id: existingVerifications[0].id,
                code_provider: data.code_provider,
                provider_metadata: { token_hash: tokenHash },
                requested_at: requestedAt,
                verified_at: null,
            }, sharedContext);
        }
        else {
            verification = await this.authVerificationService_.create({
                auth_identity_id: data.auth_identity_id,
                entity_id: data.entity_id,
                entity_type: data.entity_type,
                code_provider: data.code_provider,
                provider_metadata: { token_hash: tokenHash },
                requested_at: requestedAt,
                metadata: data.metadata ?? null,
            }, sharedContext);
        }
        return {
            ...verification,
            code: token,
            expires_at: expiresAt,
        };
    }
    async confirm(data, sharedContext = {}) {
        if (!data.code) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, "Verification code is required");
        }
        const [verification] = await this.authVerificationService_.list({
            provider_metadata: {
                token_hash: (0, verification_token_1.hashVerificationToken)(data.code),
            },
        }, {}, sharedContext);
        if (!verification || verification.verified_at) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_ALLOWED, "Verification code is invalid or already used");
        }
        if (data.code_provider &&
            data.code_provider !== verification.code_provider) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_ALLOWED, `Verification code does not belong to provider "${data.code_provider}"`);
        }
        const expiresAt = new Date(verification.requested_at).getTime() + this.getTokenTtlMs_();
        if (expiresAt <= Date.now()) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_ALLOWED, "Verification code has expired");
        }
        return await this.authVerificationService_.update({
            id: verification.id,
            verified_at: new Date(Date.now()),
        }, sharedContext);
    }
    getTokenTtlMs_() {
        return (0, verification_token_1.getVerificationTokenTtlMs)(this.options_.ttl_seconds ?? 900);
    }
}
exports.TokenVerificationProvider = TokenVerificationProvider;
TokenVerificationProvider.identifier = "token";
