// @medusajs/auth/dist/models/auth-password-reset-token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthPasswordResetToken = void 0;
const utils_1 = require("@medusajs/framework/utils");
const auth_identity_1 = require("./auth-identity");
const provider_identity_1 = require("./provider-identity");
exports.AuthPasswordResetToken = utils_1.model
    .define("auth_password_reset_token", {
    id: utils_1.model.id({ prefix: "authprt" }).primaryKey(),
    auth_identity: utils_1.model.belongsTo(() => auth_identity_1.AuthIdentity, {
        mappedBy: "password_reset_tokens",
    }),
    provider_identity: utils_1.model.belongsTo(() => provider_identity_1.ProviderIdentity, {
        mappedBy: "password_reset_tokens",
    }),
    entity_id: utils_1.model.text(),
    token_hash: utils_1.model.text(),
    expires_at: utils_1.model.dateTime(),
})
    .indexes([
    {
        name: "IDX_auth_password_reset_token_provider_identity_id",
        on: ["provider_identity_id"],
        where: "deleted_at IS NULL",
    },
    {
        name: "IDX_auth_password_reset_token_token_hash",
        on: ["token_hash"],
        where: "deleted_at IS NULL",
    },
    {
        name: "IDX_auth_password_reset_token_expires_at",
        on: ["expires_at"],
        where: "deleted_at IS NULL",
    },
]);
