// @medusajs/auth/dist/models/auth-identity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthIdentity = void 0;
const utils_1 = require("@medusajs/framework/utils");
const auth_mfa_factor_1 = require("./auth-mfa-factor");
const auth_mfa_recovery_code_1 = require("./auth-mfa-recovery-code");
const auth_password_reset_token_1 = require("./auth-password-reset-token");
const auth_verification_1 = require("./auth-verification");
const provider_identity_1 = require("./provider-identity");
exports.AuthIdentity = utils_1.model
    .define("auth_identity", {
    id: utils_1.model.id({ prefix: "authid" }).primaryKey(),
    provider_identities: utils_1.model.hasMany(() => provider_identity_1.ProviderIdentity, {
        mappedBy: "auth_identity",
    }),
    mfa_factors: utils_1.model.hasMany(() => auth_mfa_factor_1.AuthMfaFactor, {
        mappedBy: "auth_identity",
    }),
    mfa_recovery_codes: utils_1.model.hasMany(() => auth_mfa_recovery_code_1.AuthMfaRecoveryCode, {
        mappedBy: "auth_identity",
    }),
    verifications: utils_1.model.hasMany(() => auth_verification_1.AuthVerification, {
        mappedBy: "auth_identity",
    }),
    password_reset_tokens: utils_1.model.hasMany(() => auth_password_reset_token_1.AuthPasswordResetToken, {
        mappedBy: "auth_identity",
    }),
    app_metadata: utils_1.model.json().nullable(),
})
    .cascades({
    delete: [
        "provider_identities",
        "mfa_factors",
        "mfa_recovery_codes",
        "verifications",
        "password_reset_tokens",
    ],
});
