// @medusajs/auth/dist/models/auth-verification.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthVerification = void 0;
const utils_1 = require("@medusajs/framework/utils");
const auth_identity_1 = require("./auth-identity");
exports.AuthVerification = utils_1.model
    .define("auth_verification", {
    id: utils_1.model.id({ prefix: "authver" }).primaryKey(),
    auth_identity: utils_1.model.belongsTo(() => auth_identity_1.AuthIdentity, {
        mappedBy: "verifications",
    }),
    entity_id: utils_1.model.text(),
    entity_type: utils_1.model.text(),
    code_provider: utils_1.model.text(),
    verified_at: utils_1.model.dateTime().nullable(),
    requested_at: utils_1.model.dateTime(),
    provider_metadata: utils_1.model.json().nullable(),
    metadata: utils_1.model.json().nullable(),
})
    .indexes([
    {
        name: "IDX_auth_verification_unique_auth_identity_entity_id_entity_type",
        on: ["auth_identity_id", "entity_id", "entity_type"],
        unique: true,
        where: "deleted_at IS NULL",
    },
]);
