// @medusajs/medusa/dist/api/admin/rbac/policies/query-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listRbacPolicyRolesTransformQueryConfig = exports.defaultAdminRbacPolicyRolesFields = exports.listTransformQueryConfig = exports.retrieveTransformQueryConfig = exports.defaultAdminRbacPolicyFields = void 0;
exports.defaultAdminRbacPolicyFields = [
    "id",
    "key",
    "resource",
    "operation",
    "name",
    "description",
    "metadata",
    "created_at",
    "updated_at",
    "deleted_at",
];
exports.retrieveTransformQueryConfig = {
    defaults: exports.defaultAdminRbacPolicyFields,
    isList: false,
};
exports.listTransformQueryConfig = {
    ...exports.retrieveTransformQueryConfig,
    defaultLimit: 20,
    isList: true,
};
exports.defaultAdminRbacPolicyRolesFields = [
    "id",
    "role.id",
    "role.name",
    "role.description",
    "role.created_at",
    "role.updated_at",
    "role.users.id",
];
exports.listRbacPolicyRolesTransformQueryConfig = {
    defaults: exports.defaultAdminRbacPolicyRolesFields,
    defaultLimit: 10,
    isList: true,
};
