// @medusajs/medusa/dist/api/auth/token/refresh/route.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.POST = void 0;
const utils_1 = require("@medusajs/framework/utils");
const generate_jwt_token_1 = require("../../utils/generate-jwt-token");
// Retrieve a newly generated JWT token. Checking the existing token is valid already happens in the auth middleware.
// Note: We probably want to disallow refreshes if the password changes, and require reauth.
const POST = async (req, res) => {
    const service = req.scope.resolve(utils_1.Modules.AUTH);
    if (!req.auth_context.auth_provider) {
        throw new utils_1.MedusaError(utils_1.MedusaError.Types.UNAUTHORIZED, "The auth provider is not set while refreshing token");
    }
    /* A request would have the actor ID set on it only after it:
      - Creates an actor after registration
      - Passes MFA and other verifications
    Once that is done we can safely just regenerate a token without performing any additional checks.
  
    However, if the actor is not set we have to perform the MFA and verification checks - for the refresh
    call we don't return verification/mfa required, but instead throw as the call to refresh the token is
    not expected until mfa/verifications are completed.
    */
    if (req.auth_context.actor_id) {
        const { http } = req.scope.resolve(utils_1.ContainerRegistrationKeys.CONFIG_MODULE).projectConfig;
        const authIdentity = await service.retrieveAuthIdentity(req.auth_context.auth_identity_id, { relations: ["provider_identities"] });
        if (!authIdentity) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.UNAUTHORIZED, "Invalid auth identity while refreshing token");
        }
        const token = await (0, generate_jwt_token_1.generateJwtTokenForAuthIdentity)({
            authIdentity,
            actorType: req.auth_context.actor_type,
            authProvider: req.auth_context.auth_provider,
            container: req.scope,
        }, {
            secret: http.jwtSecret,
            expiresIn: http.jwtExpiresIn,
            options: http.jwtOptions,
        });
        return res.json({ token });
    }
    // This will get results (and perform checks), similar to the authenticate call.
    const { authIdentity, mfaChallenge } = await service.validateAuthIdentity(req.auth_context.auth_identity_id, req.auth_context.auth_provider, { relations: ["provider_identities"] });
    if (!authIdentity) {
        throw new utils_1.MedusaError(utils_1.MedusaError.Types.UNAUTHORIZED, "Invalid auth identity while refreshing token");
    }
    const result = await (0, generate_jwt_token_1.generateJwtTokenWithChecks)(req.scope, {
        authIdentity,
        mfaChallenge,
        actorType: req.auth_context.actor_type,
        authProvider: req.auth_context.auth_provider,
    });
    return res.json(result);
};
exports.POST = POST;
