// @medusajs/medusa/dist/commands/user.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = default_1;
const utils_1 = require("@medusajs/framework/utils");
const telemetry_1 = require("@medusajs/telemetry");
const express_1 = __importDefault(require("express"));
const loaders_1 = __importDefault(require("../loaders"));
async function default_1({ directory, id, email, password, keepAlive, invite, }) {
    (0, telemetry_1.track)("CLI_USER", { with_id: !!id });
    const app = (0, express_1.default)();
    try {
        process.env.MEDUSA_WORKER_MODE = "server";
        const { container } = await (0, loaders_1.default)({
            directory,
            expressApp: app,
            skipLoadingEntryPoints: true,
        });
        const logger = container.resolve(utils_1.ContainerRegistrationKeys.LOGGER);
        const authService = container.resolve(utils_1.Modules.AUTH);
        const workflowService = container.resolve(utils_1.Modules.WORKFLOW_ENGINE);
        const provider = "emailpass";
        // Check if RBAC is enabled and get super admin role
        let userRoles = [];
        const rbacEnabled = utils_1.FeatureFlag.isFeatureEnabled("rbac");
        if (rbacEnabled) {
            const rbacService = container.resolve(utils_1.Modules.RBAC);
            const superAdminRoles = await rbacService.listRbacRoles({
                id: "role_super_admin",
            });
            if (superAdminRoles.length > 0) {
                userRoles = [superAdminRoles[0].id];
            }
        }
        if (invite) {
            const { result: invites } = await workflowService.run("create-invite-step", {
                input: {
                    invites: [
                        {
                            email,
                            roles: userRoles,
                        },
                    ],
                },
            });
            const createdInvite = invites[0];
            logger.info(`
      Invite token: ${createdInvite.token}
      Open the invite in Medusa Admin at: [your-admin-url]/invite?token=${createdInvite.token}`);
        }
        else {
            if (userRoles.length > 0) {
                logger.info("Assigning super admin role to user.");
            }
            const { result: users } = await workflowService.run("create-users-workflow", {
                input: {
                    users: [
                        {
                            email,
                            roles: userRoles,
                        },
                    ],
                },
            });
            const user = users[0];
            const { authIdentity, error } = await authService.register(provider, {
                actor_type: "user",
                body: {
                    email,
                    password,
                },
            });
            if (error) {
                logger.error(error);
                process.exit(1);
            }
            await authService.updateAuthIdentities({
                id: authIdentity.id,
                app_metadata: {
                    user_id: user.id,
                },
            });
            logger.info("User created successfully." +
                (userRoles.length > 0 ? " Super admin role assigned." : ""));
        }
    }
    catch (err) {
        console.error(err);
        process.exit(1);
    }
    (0, telemetry_1.track)("CLI_USER_COMPLETED", { with_id: !!id });
    if (!keepAlive) {
        process.exit();
    }
}
