// @medusajs/core-flows/dist/auth/workflows/generate-reset-password-token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateResetPasswordTokenWorkflow = void 0;
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
const common_1 = require("../../common");
const steps_1 = require("../steps");
const RESET_PASSWORD_TOKEN_TTL_SECONDS = 15 * 60;
/**
 * This workflow generates a reset password token for a user. It's used by the
 * [Generate Reset Password Token for Admin](https://docs.medusajs.com/api/admin#auth_postactor_typeauth_providerresetpassword)
 * and [Generate Reset Password Token for Customer](https://docs.medusajs.com/api/store#auth_postactor_typeauth_providerresetpassword)
 * API Routes.
 *
 * The workflow emits the `auth.password_reset` event, which you can listen to in
 * a [subscriber](https://docs.medusajs.com/learn/fundamentals/events-and-subscribers). Follow
 * [this guide](https://docs.medusajs.com/resources/commerce-modules/auth/reset-password) to learn
 * how to handle this event.
 *
 * You can use this workflow within your customizations or your own custom workflows, allowing you to
 * generate reset password tokens within your custom flows.
 *
 * @example
 * const { result } = await generateResetPasswordTokenWorkflow(container)
 * .run({
 *   input: {
 *     entityId: "example@gmail.com",
 *     actorType: "customer",
 *     provider: "emailpass",
 *     secret: "jwt_123" // jwt secret
 *   }
 * })
 *
 * @summary
 *
 * Generate a reset password token for a user or customer.
 *
 * @since 2.16.0
 */
exports.generateResetPasswordTokenWorkflow = (0, workflows_sdk_1.createWorkflow)("generate-reset-password-token", (input) => {
    const resetToken = (0, steps_1.createPasswordResetTokenStep)({
        entityId: input.entityId,
        provider: input.provider,
        ttlSeconds: RESET_PASSWORD_TOKEN_TTL_SECONDS,
    });
    const token = (0, workflows_sdk_1.transform)({ input, resetToken }, ({ input, resetToken }) => {
        return (0, utils_1.generateJwtToken)({
            entity_id: input.entityId,
            provider: input.provider,
            actor_type: input.actorType,
            purpose: "reset",
        }, {
            secret: input.secret,
            expiresIn: `${RESET_PASSWORD_TOKEN_TTL_SECONDS}s`,
            jwtOptions: {
                ...input.jwtOptions,
                jwtid: resetToken.jti,
            },
        });
    });
    const payload = (0, workflows_sdk_1.transform)({
        input,
        token,
    }, (data) => {
        return {
            entity_id: data.input.entityId,
            actor_type: data.input.actorType,
            token: data.token,
            metadata: data.input.metadata ?? {},
        };
    });
    (0, common_1.emitEventStep)({
        eventName: utils_1.AuthWorkflowEvents.PASSWORD_RESET,
        data: payload,
    });
    return new workflows_sdk_1.WorkflowResponse(token);
});
