// @medusajs/core-flows/dist/auth/workflows/request-verification.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.requestVerificationWorkflow = exports.requestVerificationWorkflowId = void 0;
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
const common_1 = require("../../common");
const steps_1 = require("../steps");
/**
 * The ID of the request verification workflow.
 *
 * @since 2.15.5
 */
exports.requestVerificationWorkflowId = "request-verification-workflow";
/**
 * This workflow requests authentication verification.
 *
 * @since 2.15.5
 */
exports.requestVerificationWorkflow = (0, workflows_sdk_1.createWorkflow)(exports.requestVerificationWorkflowId, (input) => {
    const result = (0, steps_1.requestVerificationStep)(input);
    const eventData = (0, workflows_sdk_1.transform)({ input, result }, ({ input, result }) => {
        return {
            entity_id: result.entity_id,
            entity_type: result.entity_type,
            code_provider: result.code_provider,
            auth_identity_id: result.auth_identity_id,
            code: result.code,
            expires_at: result.expires_at,
            metadata: result.metadata ?? {},
        };
    });
    (0, common_1.emitEventStep)({
        eventName: utils_1.AuthWorkflowEvents.VERIFICATION_REQUESTED,
        data: eventData,
    });
    const verification = (0, workflows_sdk_1.transform)({ result }, (data) => {
        return {
            ...data.result,
            code: undefined,
            expires_at: undefined,
        };
    });
    return new workflows_sdk_1.WorkflowResponse(verification);
});
