// @medusajs/auth/dist/services/auth-module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("@medusajs/framework/types");
const node_crypto_1 = __importDefault(require("node:crypto"));
const utils_1 = require("@medusajs/framework/utils");
const _models_1 = require("../models");
const joiner_config_1 = require("../joiner-config");
class AuthModuleService extends (0, utils_1.MedusaService)({ AuthIdentity: _models_1.AuthIdentity, ProviderIdentity: _models_1.ProviderIdentity, AuthVerification: _models_1.AuthVerification }) {
    constructor({ authIdentityService, authVerificationService, authPasswordResetTokenService, authMfaFactorService, authMfaRecoveryCodeService, providerIdentityService, authProviderService, authMfaProviderService, authVerificationProviderService, baseRepository, cache, }, moduleOptions = {}, moduleDeclaration) {
        // @ts-ignore
        super(...arguments);
        this.moduleDeclaration = moduleDeclaration;
        this.baseRepository_ = baseRepository;
        this.authIdentityService_ = authIdentityService;
        this.authVerificationService_ = authVerificationService;
        this.authPasswordResetTokenService_ = authPasswordResetTokenService;
        this.authMfaFactorService_ = authMfaFactorService;
        this.authMfaRecoveryCodeService_ = authMfaRecoveryCodeService;
        this.authProviderService_ = authProviderService;
        this.authMfaProviderService_ = authMfaProviderService;
        this.authVerificationProviderService_ = authVerificationProviderService;
        this.providerIdentityService_ = providerIdentityService;
        this.cache_ = cache;
        this.moduleOptions_ = moduleOptions;
    }
    __joinerConfig() {
        return joiner_config_1.joinerConfig;
    }
    async createAuthIdentities(data, sharedContext = {}) {
        const authIdentities = await this.authIdentityService_.create(data, sharedContext);
        return await this.baseRepository_.serialize(authIdentities, {
            populate: true,
        });
    }
    // @ts-expect-error
    async updateAuthIdentities(data, sharedContext = {}) {
        const updatedUsers = await this.authIdentityService_.update(data, sharedContext);
        const serializedUsers = await this.baseRepository_.serialize(updatedUsers, {
            populate: true,
        });
        return serializedUsers;
    }
    async register(provider, authenticationData) {
        try {
            const response = await this.authProviderService_.register(provider, authenticationData, this.getAuthIdentityProviderService(provider));
            return await this.applyMfaRequirement_(response, {
                actor_type: authenticationData.actor_type,
                auth_provider: provider,
            });
        }
        catch (error) {
            return { success: false, error: error.message };
        }
    }
    // @ts-expect-error
    async createProviderIdentities(data, sharedContext = {}) {
        const providerIdentities = await this.providerIdentityService_.create(data, sharedContext);
        return await this.baseRepository_.serialize(providerIdentities);
    }
    // @ts-expect-error
    async updateProviderIdentities(data, sharedContext = {}) {
        const updatedProviders = await this.providerIdentityService_.update(data, sharedContext);
        const serializedProviders = await this.baseRepository_.serialize(updatedProviders);
        return serializedProviders;
    }
    async updateProvider(provider, data) {
        try {
            return await this.authProviderService_.update(provider, data, this.getAuthIdentityProviderService(provider));
        }
        catch (error) {
            return { success: false, error: error.message };
        }
    }
    async authenticate(provider, authenticationData) {
        try {
            const response = await this.authProviderService_.authenticate(provider, authenticationData, this.getAuthIdentityProviderService(provider));
            return await this.applyMfaRequirement_(response, {
                actor_type: authenticationData.actor_type,
                auth_provider: provider,
            });
        }
        catch (error) {
            return { success: false, error: error.message };
        }
    }
    async validateCallback(provider, authenticationData) {
        try {
            const response = await this.authProviderService_.validateCallback(provider, authenticationData, this.getAuthIdentityProviderService(provider));
            return await this.applyMfaRequirement_(response, {
                actor_type: authenticationData.actor_type,
                auth_provider: provider,
            });
        }
        catch (error) {
            return { success: false, error: error.message };
        }
    }
    async validateAuthIdentity(id, provider, config, sharedContext) {
        try {
            const authIdentity = await this.authIdentityService_.retrieve(id, config, sharedContext);
            return await this.applyMfaRequirement_({
                success: true,
                authIdentity: authIdentity,
            }, {
                auth_provider: provider,
            });
        }
        catch (error) {
            return { success: false, error: error.message };
        }
    }
    async startAuthMfa(data, sharedContext = {}) {
        return await this.startAuthMfa_(data, sharedContext);
    }
    async startAuthMfa_(data, sharedContext = {}) {
        await this.authIdentityService_.retrieve(data.auth_identity_id, {}, sharedContext);
        return await this.authMfaProviderService_.start(data.provider, data, sharedContext);
    }
    async verifyAuthMfa(data, sharedContext = {}) {
        return await this.verifyAuthMfa_(data, sharedContext);
    }
    async verifyAuthMfa_(data, sharedContext = {}) {
        const factor = await this.authMfaFactorService_.retrieve(data.id, {}, sharedContext);
        if (data.auth_identity_id &&
            factor.auth_identity_id !== data.auth_identity_id) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_FOUND, `MFA factor with id "${data.id}" was not found`);
        }
        const verifiedFactor = await this.authMfaProviderService_.verifySetup(factor.provider, data, sharedContext);
        return verifiedFactor;
    }
    async createAuthMfaChallenge(data, sharedContext = {}) {
        return await this.createAuthMfaChallenge_(data, sharedContext);
    }
    async createAuthMfaChallenge_(data, sharedContext = {}) {
        await this.authIdentityService_.retrieve(data.auth_identity_id, {}, sharedContext);
        const methods = await this.getAvailableMfaChallengeMethods_(data.auth_identity_id, sharedContext);
        if (!methods.length) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_ALLOWED, "Auth identity does not have any enabled MFA methods");
        }
        const challengeConfig = this.getMfaChallengeConfig_();
        const challenge = {
            id: (0, utils_1.generateEntityId)(undefined, "authmfachal"),
            auth_identity_id: data.auth_identity_id,
            actor_type: data.actor_type ?? null,
            auth_provider: data.auth_provider ?? null,
            methods,
            expires_at: new Date(Date.now() + challengeConfig.ttlSeconds * 1000),
            attempts: 0,
            max_attempts: challengeConfig.maxAttempts,
            completed_at: null,
            metadata: data.metadata ?? null,
        };
        await this.setMfaChallenge_(challenge, challengeConfig.ttlSeconds);
        return challenge;
    }
    async verifyAuthMfaChallenge(data, sharedContext = {}) {
        return await this.verifyAuthMfaChallenge_(data, sharedContext);
    }
    async verifyAuthMfaChallenge_(data, sharedContext = {}) {
        const challenge = await this.retrieveMfaChallenge_(data.id);
        this.assertMfaChallengeCanBeVerified_(challenge, data.method);
        const valid = await this.authMfaProviderService_.verify(data.method, {
            auth_identity_id: challenge.auth_identity_id,
            code: data.code,
        }, sharedContext);
        if (!valid) {
            const attempts = challenge.attempts + 1;
            await this.setMfaChallenge_({
                ...challenge,
                attempts,
            });
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_ALLOWED, attempts >= challenge.max_attempts
                ? "MFA challenge has too many failed attempts"
                : "Invalid MFA challenge code");
        }
        const completedChallenge = {
            ...challenge,
            completed_at: new Date(),
        };
        await this.setMfaChallenge_(completedChallenge);
        return completedChallenge;
    }
    async disableAuthMfa(data, sharedContext = {}) {
        const factor = await this.authMfaFactorService_.retrieve(data.id, {}, sharedContext);
        if (factor.status === "enabled" &&
            this.getMfaDisablePolicy_() === "challenge") {
            if (!data.method || !data.code) {
                throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, "MFA verification code is required to disable MFA");
            }
            const valid = await this.authMfaProviderService_.verify(data.method, {
                auth_identity_id: factor.auth_identity_id,
                code: data.code,
            }, sharedContext);
            if (!valid) {
                throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_ALLOWED, "Invalid MFA verification code");
            }
        }
        const disabledFactor = await this.authMfaFactorService_.update({
            id: data.id,
            status: "disabled",
        }, sharedContext);
        return await this.serializeMfaFactor_(disabledFactor);
    }
    async retrieveAuthMfa(selector, config = {}, sharedContext = {}) {
        const filters = typeof selector === "string"
            ? { id: [selector] }
            : {
                id: [selector.id],
                auth_identity_id: selector.auth_identity_id,
            };
        const [factor] = await this.authMfaFactorService_.list(filters, config, sharedContext);
        if (!factor) {
            const id = typeof selector === "string" ? selector : selector.id;
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_FOUND, `MFA factor with id "${id}" was not found`);
        }
        return await this.serializeMfaFactor_(factor);
    }
    async listAuthMfa(filters = {}, config = {}, sharedContext = {}) {
        const factors = await this.authMfaFactorService_.list(filters, config, sharedContext);
        return await Promise.all(factors.map((factor) => this.serializeMfaFactor_(factor)));
    }
    async generateAuthMfaRecoveryCodes(data, sharedContext = {}) {
        return await this.generateAuthMfaRecoveryCodes_(data, sharedContext);
    }
    async generateAuthMfaRecoveryCodes_(data, sharedContext = {}) {
        const count = data.count ?? this.moduleOptions_.mfa?.recovery_code_count ?? 10;
        if (!Number.isInteger(count) || count < 1 || count > 50) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, "Recovery code count must be between 1 and 50");
        }
        await this.authIdentityService_.retrieve(data.auth_identity_id, {}, sharedContext);
        const codes = await this.authMfaProviderService_.generateCodes("recovery_code", {
            auth_identity_id: data.auth_identity_id,
            count,
        }, sharedContext);
        return { codes };
    }
    async useAuthMfaRecoveryCode(data, sharedContext = {}) {
        return await this.useAuthMfaRecoveryCode_(data, sharedContext);
    }
    async useAuthMfaRecoveryCode_(data, sharedContext = {}) {
        const valid = await this.authMfaProviderService_.verify("recovery_code", data, sharedContext);
        if (!valid) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_ALLOWED, "Recovery code is invalid or already used");
        }
    }
    async requestAuthVerification(data, sharedContext = {}) {
        if (!data.code_provider) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, "Verification provider is required");
        }
        return await this.authVerificationProviderService_.request(data.code_provider, data, sharedContext);
    }
    async confirmAuthVerification(data, sharedContext = {}) {
        const codeProvider = data.code_provider ?? "token";
        return await this.authVerificationProviderService_.confirm(codeProvider, { ...data, code_provider: codeProvider }, sharedContext);
    }
    async applyMfaRequirement_(response, context) {
        if (!response.success || !response.authIdentity || response.location) {
            return response;
        }
        const methods = await this.getAvailableMfaChallengeMethods_(response.authIdentity.id);
        if (!methods.length) {
            return response;
        }
        const mfaChallenge = await this.createAuthMfaChallenge_({
            auth_identity_id: response.authIdentity.id,
            actor_type: context.actor_type ?? null,
            auth_provider: context.auth_provider ?? null,
        });
        return {
            success: true,
            mfaChallenge: mfaChallenge,
            authIdentity: response.authIdentity,
        };
    }
    async getAvailableMfaChallengeMethods_(authIdentityId, sharedContext = {}) {
        const factors = await this.authMfaFactorService_.list({
            auth_identity_id: authIdentityId,
            status: "enabled",
        }, { select: ["provider"] }, sharedContext);
        const factorMethods = Array.from(new Set(factors.map((factor) => factor.provider)));
        if (!factorMethods.length) {
            return factorMethods;
        }
        const methods = [];
        for (const method of factorMethods) {
            const canVerify = await this.authMfaProviderService_.canVerifyForAuthIdentity(method, {
                auth_identity_id: authIdentityId,
            }, sharedContext);
            if (canVerify) {
                methods.push(method);
            }
        }
        if (!methods.length) {
            return methods;
        }
        const hasRecoveryCodes = await this.authMfaProviderService_.canVerifyForAuthIdentity("recovery_code", {
            auth_identity_id: authIdentityId,
        }, sharedContext);
        if (hasRecoveryCodes) {
            methods.push("recovery_code");
        }
        return methods;
    }
    async serializeMfaFactor_(factor) {
        const serialized = await this.baseRepository_.serialize(factor);
        delete serialized.provider_metadata;
        return serialized;
    }
    assertMfaChallengeCanBeVerified_(challenge, method) {
        if (challenge.completed_at) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_ALLOWED, "MFA challenge has already been completed");
        }
        if (new Date(challenge.expires_at).getTime() <= Date.now()) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_ALLOWED, "MFA challenge has expired");
        }
        if (challenge.attempts >= challenge.max_attempts) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_ALLOWED, "MFA challenge has too many failed attempts");
        }
        if (!challenge.methods.includes(method)) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `MFA challenge does not support method "${method}"`);
        }
    }
    async createPasswordResetToken(data, sharedContext = {}) {
        return await this.createPasswordResetToken_(data, sharedContext);
    }
    async createPasswordResetToken_(data, sharedContext = {}) {
        const [providerIdentity] = await this.providerIdentityService_.list({ provider: data.provider, entity_id: data.entity_id }, {}, sharedContext);
        if (!providerIdentity) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `Provider identity with entity_id ${data.entity_id} and provider ${data.provider} not found`);
        }
        await this.invalidatePasswordResetTokens_(providerIdentity.id, sharedContext);
        const jti = node_crypto_1.default.randomUUID();
        const expiresAt = new Date(Date.now() + this.getPasswordResetTokenTtlMs_(data.ttl_seconds ?? 900));
        await this.authPasswordResetTokenService_.create({
            auth_identity_id: providerIdentity.auth_identity_id,
            provider_identity_id: providerIdentity.id,
            entity_id: providerIdentity.entity_id,
            token_hash: this.hashVerificationToken_(jti),
            expires_at: expiresAt,
        }, sharedContext);
        return { jti, expires_at: expiresAt };
    }
    async consumePasswordResetToken(data, sharedContext = {}) {
        return await this.consumePasswordResetToken_(data, sharedContext);
    }
    async consumePasswordResetToken_(data, sharedContext = {}) {
        const [resetToken] = await this.authPasswordResetTokenService_.list({ token_hash: this.hashVerificationToken_(data.jti) }, {}, sharedContext);
        if (!resetToken) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.UNAUTHORIZED, "Invalid token");
        }
        if (new Date(resetToken.expires_at).getTime() <= Date.now()) {
            await this.authPasswordResetTokenService_.delete(resetToken.id, sharedContext);
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.UNAUTHORIZED, "Invalid token");
        }
        const providerIdentity = await this.providerIdentityService_.retrieve(resetToken.provider_identity_id, {}, sharedContext);
        if (providerIdentity.provider !== data.provider ||
            providerIdentity.entity_id !== data.entity_id) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.UNAUTHORIZED, "Invalid token");
        }
        await this.authPasswordResetTokenService_.delete(resetToken.id, sharedContext);
        return {
            auth_identity_id: resetToken.auth_identity_id,
            provider_identity_id: resetToken.provider_identity_id,
            entity_id: resetToken.entity_id,
        };
    }
    async invalidatePasswordResetTokens_(providerIdentityId, sharedContext = {}) {
        const existing = await this.authPasswordResetTokenService_.list({ provider_identity_id: providerIdentityId }, { select: ["id"] }, sharedContext);
        const ids = existing.map((token) => token.id);
        if (ids.length) {
            await this.authPasswordResetTokenService_.delete(ids, sharedContext);
        }
    }
    getPasswordResetTokenTtlMs_(ttlSeconds) {
        if (!Number.isInteger(ttlSeconds) || ttlSeconds < 1) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, "Password reset token TTL must be a positive integer");
        }
        return ttlSeconds * 1000;
    }
    hashVerificationToken_(token) {
        return node_crypto_1.default.createHash("sha256").update(token).digest("hex");
    }
    getMfaChallengeConfig_() {
        const ttlSeconds = this.moduleOptions_.mfa?.challenge_ttl_seconds ?? 300;
        const maxAttempts = this.moduleOptions_.mfa?.challenge_max_attempts ?? 5;
        if (!Number.isInteger(ttlSeconds) || ttlSeconds < 1) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, "MFA challenge TTL must be a positive integer");
        }
        if (!Number.isInteger(maxAttempts) || maxAttempts < 1) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, "MFA challenge max attempts must be a positive integer");
        }
        return {
            ttlSeconds,
            maxAttempts,
        };
    }
    getMfaDisablePolicy_() {
        const policy = this.moduleOptions_.mfa?.disable_policy ?? "session";
        if (policy !== "challenge" && policy !== "session") {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, 'MFA disable policy must be either "challenge" or "session"');
        }
        return policy;
    }
    async retrieveMfaChallenge_(id) {
        const challenge = await this.getCache_().get(this.getMfaChallengeCacheKey_(id));
        if (!challenge) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_FOUND, `MFA challenge with id "${id}" was not found`);
        }
        return {
            ...challenge,
            expires_at: new Date(challenge.expires_at),
            completed_at: challenge.completed_at
                ? new Date(challenge.completed_at)
                : null,
        };
    }
    async setMfaChallenge_(challenge, ttlSeconds) {
        const ttl = ttlSeconds ??
            Math.ceil((new Date(challenge.expires_at).getTime() - Date.now()) / 1000);
        if (ttl < 1) {
            return;
        }
        await this.getCache_().set(this.getMfaChallengeCacheKey_(challenge.id), challenge, ttl);
    }
    getMfaChallengeCacheKey_(id) {
        return `auth:mfa:challenge:${id}`;
    }
    getCache_() {
        if (!this.cache_) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_ARGUMENT, "Cache module dependency is required when using MFA challenges");
        }
        return this.cache_;
    }
    getAuthIdentityProviderService(provider) {
        return {
            retrieve: async ({ entity_id }) => {
                const authIdentities = await this.authIdentityService_.list({
                    provider_identities: {
                        entity_id,
                        provider,
                    },
                }, {
                    relations: ["provider_identities"],
                });
                if (!authIdentities.length) {
                    throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_FOUND, `AuthIdentity with entity_id "${entity_id}" not found`);
                }
                if (authIdentities.length > 1) {
                    throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `Multiple authIdentities found for entity_id "${entity_id}"`);
                }
                return await this.baseRepository_.serialize(authIdentities[0]);
            },
            create: async (data) => {
                const normalizedRequest = {
                    provider_identities: [
                        {
                            entity_id: data.entity_id,
                            provider_metadata: data.provider_metadata,
                            user_metadata: data.user_metadata,
                            provider,
                        },
                    ],
                };
                const createdAuthIdentity = await this.authIdentityService_.create(normalizedRequest);
                return await this.baseRepository_.serialize(createdAuthIdentity);
            },
            update: async (entity_id, data) => {
                const authIdentities = await this.authIdentityService_.list({
                    provider_identities: {
                        entity_id,
                        provider,
                    },
                }, {
                    relations: ["provider_identities"],
                });
                if (!authIdentities.length) {
                    throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_FOUND, `AuthIdentity with entity_id "${entity_id}" not found`);
                }
                if (authIdentities.length > 1) {
                    throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `Multiple authIdentities found for entity_id "${entity_id}"`);
                }
                const providerIdentityData = authIdentities[0].provider_identities.find((pi) => pi.provider === provider);
                if (!providerIdentityData) {
                    throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_FOUND, `ProviderIdentity with entity_id "${entity_id}" not found`);
                }
                const updatedProviderIdentity = await this.providerIdentityService_.update({
                    id: providerIdentityData.id,
                    ...data,
                });
                const serializedResponse = await this.baseRepository_.serialize(authIdentities[0]);
                const serializedProviderIdentity = await this.baseRepository_.serialize(updatedProviderIdentity);
                serializedResponse.provider_identities = [
                    ...(serializedResponse.provider_identities?.filter((p) => p.provider !== provider) ?? []),
                    serializedProviderIdentity,
                ];
                return serializedResponse;
            },
            setState: async (key, value) => {
                if (!this.cache_) {
                    throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_ARGUMENT, "Cache module dependency is required when using OAuth providers that require state");
                }
                // 20 minutes. Can be made configurable if necessary, but this is a good default.
                this.cache_.set(key, value, 1200);
            },
            getState: async (key) => {
                if (!this.cache_) {
                    throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_ARGUMENT, "Cache module dependency is required when using OAuth providers that require state");
                }
                return await this.cache_.get(key);
            },
        };
    }
}
exports.default = AuthModuleService;
__decorate([
    (0, utils_1.InjectManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "createAuthIdentities", null);
__decorate([
    (0, utils_1.InjectManager)()
    // @ts-expect-error
    ,
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "updateAuthIdentities", null);
__decorate([
    (0, utils_1.InjectManager)()
    // @ts-expect-error
    ,
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "createProviderIdentities", null);
__decorate([
    (0, utils_1.InjectManager)()
    // @ts-expect-error
    ,
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "updateProviderIdentities", null);
__decorate([
    (0, utils_1.InjectManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "startAuthMfa", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "startAuthMfa_", null);
__decorate([
    (0, utils_1.InjectManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "verifyAuthMfa", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "verifyAuthMfa_", null);
__decorate([
    (0, utils_1.InjectManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "createAuthMfaChallenge", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "createAuthMfaChallenge_", null);
__decorate([
    (0, utils_1.InjectManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "verifyAuthMfaChallenge", null);
__decorate([
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "verifyAuthMfaChallenge_", null);
__decorate([
    (0, utils_1.InjectManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "disableAuthMfa", null);
__decorate([
    (0, utils_1.InjectManager)(),
    __param(2, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "retrieveAuthMfa", null);
__decorate([
    (0, utils_1.InjectManager)(),
    __param(2, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "listAuthMfa", null);
__decorate([
    (0, utils_1.InjectManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "generateAuthMfaRecoveryCodes", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "generateAuthMfaRecoveryCodes_", null);
__decorate([
    (0, utils_1.InjectManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "useAuthMfaRecoveryCode", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "useAuthMfaRecoveryCode_", null);
__decorate([
    (0, utils_1.InjectManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "requestAuthVerification", null);
__decorate([
    (0, utils_1.InjectManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "confirmAuthVerification", null);
__decorate([
    (0, utils_1.InjectManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "createPasswordResetToken", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "createPasswordResetToken_", null);
__decorate([
    (0, utils_1.InjectManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "consumePasswordResetToken", null);
__decorate([
    (0, utils_1.InjectTransactionManager)(),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], AuthModuleService.prototype, "consumePasswordResetToken_", null);
