// @medusajs/auth/dist/services/verification-provider.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AuthVerificationProviderService_logger;
Object.defineProperty(exports, "__esModule", { value: true });
const _types_1 = require("../types");
class AuthVerificationProviderService {
    constructor(container) {
        _AuthVerificationProviderService_logger.set(this, void 0);
        this.dependencies = container;
        __classPrivateFieldSet(this, _AuthVerificationProviderService_logger, container["logger"]
            ? container.logger
            : console, "f");
    }
    retrieveProviderRegistration(provider) {
        try {
            return this.dependencies[`${_types_1.AuthVerificationProviderRegistrationPrefix}${provider}`];
        }
        catch (err) {
            if (err.name === "AwilixResolutionError") {
                const errMessage = `
Unable to retrieve the verification provider with id: ${provider}
Please make sure that the provider is registered in the container and it is configured correctly in your project configuration file.`;
                __classPrivateFieldGet(this, _AuthVerificationProviderService_logger, "f").error(`AwilixResolutionError: ${err.message}`, err);
                throw new Error(errMessage);
            }
            const errMessage = `Unable to retrieve the verification provider with id: ${provider}, the following error occurred: ${err.message}`;
            __classPrivateFieldGet(this, _AuthVerificationProviderService_logger, "f").error(errMessage);
            throw new Error(errMessage);
        }
    }
    async request(provider, data, sharedContext) {
        return await this.retrieveProviderRegistration(provider).request(data, sharedContext);
    }
    async confirm(provider, data, sharedContext) {
        return await this.retrieveProviderRegistration(provider).confirm(data, sharedContext);
    }
}
_AuthVerificationProviderService_logger = new WeakMap();
exports.default = AuthVerificationProviderService;
