// @medusajs/medusa/dist/migration-scripts/create-super-admin-role.js
#!/usr/bin/env node
"use strict";
/**
 * Migration Script: Assign Super Admin Role to All Admin Users
 *
 * This script assigns the pre-created super admin role
 * to all existing admin users in your Medusa instance.
 *
 * Usage: npx tsx packages/medusa/src/migration-scripts/create-super-admin-role.ts
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = createSuperAdminRole;
const modules_sdk_1 = require("@medusajs/framework/modules-sdk");
const utils_1 = require("@medusajs/framework/utils");
const dotenv_1 = __importDefault(require("dotenv"));
dotenv_1.default.config();
async function assignSuperAdminRoleToUsers(container) {
    try {
        const logger = container.resolve(utils_1.ContainerRegistrationKeys.LOGGER);
        logger.info("Starting super admin role assignment script...");
        const userModuleService = container.resolve(utils_1.Modules.USER);
        const rbacModuleService = container.resolve(utils_1.Modules.RBAC);
        const users = await userModuleService.listUsers({});
        if (!users.length) {
            logger.info("No users found. Exiting.");
            return;
        }
        logger.info(`Found ${users.length} users`);
        logger.info("\nUsers found:");
        users.forEach((user, index) => {
            logger.info(`  ${index + 1}. ${user.email || user.id} (${user.id})`);
        });
        logger.info("\nLooking for super admin role...");
        // Get the pre-created super admin role
        let superAdminRole;
        try {
            const existingRoles = await rbacModuleService.listRbacRoles({
                id: "role_super_admin",
            });
            if (existingRoles.length) {
                superAdminRole = existingRoles[0];
                logger.info(`Found super admin role: ${superAdminRole.name} (${superAdminRole.id})`);
            }
            else {
                throw new Error("Super admin role not found. Please ensure RBAC module is loaded.");
            }
        }
        catch (error) {
            logger.error("Could not find super admin role. Please ensure RBAC module is loaded.", error);
            throw error;
        }
        // Verify the role has the wildcard policy
        try {
            const rolePolicies = await rbacModuleService.listRbacRolePolicies({
                role_id: superAdminRole.id,
            });
            if (!rolePolicies.length) {
                logger.warn("Super admin role has no policies assigned. This might indicate an issue with the migration.");
            }
            else {
                logger.info(`Super admin role has ${rolePolicies.length} policies assigned`);
            }
        }
        catch (error) {
            logger.warn("Could not verify role policies:", error);
        }
        logger.info("Assigning super admin role to all users...");
        let successCount = 0;
        let errorCount = 0;
        const link = container.resolve(utils_1.ContainerRegistrationKeys.LINK);
        for (const user of users) {
            try {
                logger.info(`  Processing user: ${user.email || user.id}`);
                // Link the user to the super admin role
                await link.create({
                    [utils_1.Modules.USER]: {
                        user_id: user.id,
                    },
                    [utils_1.Modules.RBAC]: {
                        rbac_role_id: superAdminRole.id,
                    },
                });
                logger.info(`    Assigned super admin role to user`);
                successCount++;
            }
            catch (error) {
                console.error(`    Failed to assign role to user ${user.id}:`, error.message);
                errorCount++;
            }
        }
        logger.info(`  Successfully assigned role: ${successCount} users`);
        if (errorCount > 0) {
            logger.info(`  Failed to assign role: ${errorCount} users`);
        }
        if (successCount > 0) {
            logger.log("\nSuper admin role assignment completed successfully!");
            const totalAssigned = successCount;
            logger.info(`Super admin role is now assigned to ${totalAssigned} out of ${users.length} users`);
        }
    }
    catch (error) {
        console.error("\nFatal error:", error.message);
        console.error("Stack trace:", error.stack);
        throw error;
    }
}
async function createSuperAdminRole({ container }) {
    if (!modules_sdk_1.MedusaModule.isInstalled(utils_1.Modules.USER) ||
        !modules_sdk_1.MedusaModule.isInstalled(utils_1.Modules.AUTH) ||
        !modules_sdk_1.MedusaModule.isInstalled(utils_1.Modules.RBAC)) {
        const logger = container.resolve(utils_1.ContainerRegistrationKeys.LOGGER);
        logger.info("Required modules (USER, AUTH, RBAC) not installed. Skipping.");
        return;
    }
    try {
        await assignSuperAdminRoleToUsers(container);
    }
    catch (error) {
        const logger = container.resolve(utils_1.ContainerRegistrationKeys.LOGGER);
        logger.error("Failed to assign super admin role:", error);
        throw error;
    }
}
(0, utils_1.defineFileConfig)({
    isDisabled: () => !utils_1.FeatureFlag.isFeatureEnabled("rbac"),
});
