// @medusajs/core-flows/dist/rbac/steps/validate-user-permissions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateUserPermissionsStep = exports.validateUserPermissionsStepId = void 0;
const framework_1 = require("@medusajs/framework");
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
/**
 * @ignore
 * @featureFlag rbac
 */
exports.validateUserPermissionsStepId = "validate-user-permissions";
/**
 * Validates that a user has access to all the policies they are trying to assign.
 * A user can only create roles and add policies that they themselves have access to.
 * @ignore
 * @featureFlag rbac
 */
exports.validateUserPermissionsStep = (0, workflows_sdk_1.createStep)(exports.validateUserPermissionsStepId, async (data, { container }) => {
    const { actor_id, actor, policy_ids, actions } = data;
    if (!policy_ids?.length && !actions?.length) {
        return;
    }
    const query = container.resolve(utils_1.ContainerRegistrationKeys.QUERY);
    const { data: users } = await query.graph({
        entity: actor ?? "user",
        fields: ["rbac_roles.id"],
        filters: { id: actor_id },
    });
    const roleIds = users?.[0]?.rbac_roles?.map((r) => r.id).filter(Boolean) ?? [];
    if (!roleIds.length) {
        throw new utils_1.MedusaError(utils_1.MedusaError.Types.FORBIDDEN, "Forbidden");
    }
    let actionsToCheck = [];
    if (policy_ids?.length) {
        const { data: targetPolicies } = await query.graph({
            entity: "rbac_policy",
            fields: ["id", "resource", "operation"],
            filters: { id: policy_ids },
        });
        // A user cannot grant a policy that doesn't exist.
        const inexistentPolicies = (0, utils_1.arrayDifference)(policy_ids, targetPolicies.map((p) => p.id));
        if (inexistentPolicies.length) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_FOUND, `The following policies do not exist: ${inexistentPolicies.join(", ")}`);
        }
        actionsToCheck = targetPolicies.map((p) => ({
            resource: p.resource,
            operation: p.operation,
        }));
    }
    else if (actions?.length) {
        actionsToCheck = actions;
    }
    const allowed = await (0, framework_1.hasPermission)({
        roles: roleIds,
        actions: actionsToCheck,
        container,
    });
    if (!allowed) {
        throw new utils_1.MedusaError(utils_1.MedusaError.Types.FORBIDDEN, "You do not have access to some of the policies you are trying to assign.");
    }
});
