// @medusajs/medusa/dist/api/auth/[actor_type]/[auth_provider]/register/route.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.POST = void 0;
const utils_1 = require("@medusajs/framework/utils");
const generate_jwt_token_1 = require("../../../utils/generate-jwt-token");
const POST = async (req, res) => {
    const { actor_type, auth_provider } = req.params;
    const service = req.scope.resolve(utils_1.Modules.AUTH);
    const authData = {
        actor_type,
        url: req.url,
        headers: req.headers,
        query: req.query,
        body: req.body,
        protocol: req.protocol,
    };
    const { success, error, authIdentity } = await service.register(auth_provider, authData);
    if (success && authIdentity) {
        const { http } = req.scope.resolve(utils_1.ContainerRegistrationKeys.CONFIG_MODULE).projectConfig;
        /*
          At registration time the auth identity doesn't have an actor attached to it, so we return the actorless token.
          The token can be used for few, auth-related operations, such as creating an actor, completing MFA,
          and completing other verifications.
    
          Once those are done the user either needs to use the returned token with an actor attached (eg. after verifying an MFA challenge)
          Or request a token refresh (such as after passing a verification)
         */
        const token = await (0, generate_jwt_token_1.generateJwtTokenForAuthIdentity)({
            authIdentity,
            actorType: actor_type,
            authProvider: auth_provider,
            container: req.scope,
        }, {
            secret: http.jwtSecret,
            expiresIn: http.jwtExpiresIn,
            options: http.jwtOptions,
        });
        return res.status(200).json({ token });
    }
    throw new utils_1.MedusaError(utils_1.MedusaError.Types.UNAUTHORIZED, error || "Authentication failed");
};
exports.POST = POST;
