// @medusajs/medusa/dist/api/admin/orders/validators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdminCreateOrderCreditLines = exports.AdminUpdateOrder = exports.AdminTransferOrder = exports.AdminOrderChangesParams = exports.AdminOrderCancelFulfillment = exports.OrderCancelFulfillment = exports.AdminMarkOrderFulfillmentAsDelivered = exports.AdminOrderCreateShipment = exports.OrderCreateShipment = exports.AdminOrderCreateFulfillment = exports.OrderCreateFulfillment = exports.AdminCompleteOrder = exports.AdminGetOrdersParams = exports.AdminGetOrderShippingOptionList = exports.AdminGetOrdersOrderItemsParams = exports.AdminGetOrdersOrderParams = void 0;
const zod_1 = require("@medusajs/framework/zod");
const common_validators_1 = require("../../utils/common-validators");
const validators_1 = require("../../utils/validators");
const utils_1 = require("@medusajs/framework/utils");
exports.AdminGetOrdersOrderParams = (0, validators_1.createSelectParams)().merge(zod_1.z.object({
    version: zod_1.z.preprocess((val) => {
        if ((0, utils_1.isString)(val) && val) {
            return parseInt(val);
        }
        return val;
    }, zod_1.z.number().optional()),
}));
exports.AdminGetOrdersOrderItemsParams = (0, validators_1.createSelectParams)().merge(zod_1.z.object({
    id: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    item_id: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    version: zod_1.z.preprocess((val) => {
        if ((0, utils_1.isString)(val) && val) {
            return parseInt(val);
        }
        return val;
    }, zod_1.z.number().optional()),
}));
exports.AdminGetOrderShippingOptionList = zod_1.z.object({});
const AdminGetOrdersParamsBase = (0, validators_1.createFindParams)({
    limit: 15,
    offset: 0,
}).merge(zod_1.z.object({
    id: zod_1.z
        .union([zod_1.z.string(), zod_1.z.array(zod_1.z.string()), (0, validators_1.createOperatorMap)()])
        .optional(),
    status: zod_1.z
        .union([zod_1.z.string(), zod_1.z.array(zod_1.z.string()), (0, validators_1.createOperatorMap)()])
        .optional(),
    sales_channel_id: zod_1.z.array(zod_1.z.string()).optional(),
    region_id: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    customer_id: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    q: zod_1.z.string().optional(),
    created_at: (0, validators_1.createOperatorMap)().optional(),
    updated_at: (0, validators_1.createOperatorMap)().optional(),
    total: (0, validators_1.createOperatorMap)().optional(),
}));
const AdminGetOrdersParamsTransform = (v) => {
    const { total, ...rest } = v;
    return {
        ...rest,
        ...(total ? { summary: { totals: { current_order_total: total } } } : {}),
    };
};
exports.AdminGetOrdersParams = AdminGetOrdersParamsBase.transform(AdminGetOrdersParamsTransform);
exports.AdminCompleteOrder = (0, validators_1.WithAdditionalData)(zod_1.z.object({}));
const Item = zod_1.z.object({
    id: zod_1.z.string(),
    quantity: zod_1.z.number(),
});
exports.OrderCreateFulfillment = zod_1.z.object({
    items: zod_1.z.array(Item).min(1),
    location_id: zod_1.z.string().nullish(),
    shipping_option_id: zod_1.z.string().optional(),
    no_notification: zod_1.z.boolean().optional(),
    metadata: zod_1.z.record(zod_1.z.string(), zod_1.z.unknown()).nullish(),
});
exports.AdminOrderCreateFulfillment = (0, validators_1.WithAdditionalData)(exports.OrderCreateFulfillment);
const Label = zod_1.z.object({
    tracking_number: zod_1.z.string(),
    tracking_url: common_validators_1.safeHttpUrl,
    label_url: common_validators_1.safeHttpUrl,
});
exports.OrderCreateShipment = zod_1.z.object({
    items: zod_1.z.array(Item),
    labels: zod_1.z.array(Label).optional(),
    no_notification: zod_1.z.boolean().optional(),
    metadata: zod_1.z.record(zod_1.z.string(), zod_1.z.unknown()).nullish(),
});
exports.AdminOrderCreateShipment = (0, validators_1.WithAdditionalData)(exports.OrderCreateShipment);
exports.AdminMarkOrderFulfillmentAsDelivered = zod_1.z.object({
    no_notification: zod_1.z.boolean().optional(),
});
exports.OrderCancelFulfillment = zod_1.z.object({
    no_notification: zod_1.z.boolean().optional(),
});
exports.AdminOrderCancelFulfillment = (0, validators_1.WithAdditionalData)(exports.OrderCancelFulfillment);
exports.AdminOrderChangesParams = (0, validators_1.createSelectParams)().merge(zod_1.z.object({
    id: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    status: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    change_type: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    created_at: (0, validators_1.createOperatorMap)().optional(),
    updated_at: (0, validators_1.createOperatorMap)().optional(),
    deleted_at: (0, validators_1.createOperatorMap)().optional(),
}));
exports.AdminTransferOrder = zod_1.z.object({
    customer_id: zod_1.z.string(),
    description: zod_1.z.string().optional(),
    internal_note: zod_1.z.string().optional(),
    update_order_email: zod_1.z.boolean().optional(),
});
exports.AdminUpdateOrder = zod_1.z.object({
    email: zod_1.z.string().optional(),
    shipping_address: common_validators_1.AddressPayload.optional(),
    billing_address: common_validators_1.AddressPayload.optional(),
    locale: zod_1.z.string().nullish(),
    metadata: zod_1.z.record(zod_1.z.string(), zod_1.z.unknown()).nullish(),
});
exports.AdminCreateOrderCreditLines = zod_1.z.object({
    amount: zod_1.z.number(),
    reference: zod_1.z.string(),
    reference_id: zod_1.z.string(),
    metadata: zod_1.z.record(zod_1.z.string(), zod_1.z.unknown()).nullish(),
});
