// @ariakit/react-components/dist/hovercard/hovercard.js
"use client";
import { createDialogComponent } from "../dialog/dialog.js";
import { usePopover } from "../popover/popover.js";
import { HovercardScopedContextProvider, useHovercardProviderContext } from "./hovercard-context.js";
import { getElementPolygon, getEventPoint, isPointInPolygon } from "./utils/polygon.js";
import { createElement, createHook, forwardRef, useBooleanEvent, useEvent, useIsMouseMoving, useLiveRef, useMergeRefs, usePortalRef, useSafeLayoutEffect, useWrapElement } from "@ariakit/react-utils";
import { addGlobalEventListener, chain, contains, hasFocusWithin, invariant, isFalsyBooleanCallback } from "@ariakit/utils";
import { createContext, useCallback, useContext, useEffect, useRef, useState } from "react";
import { jsx } from "react/jsx-runtime";
import { useStoreState } from "@ariakit/react-store";
import { sync } from "@ariakit/store";
//#region src/hovercard/hovercard.tsx
const TagName = "div";
function isMovingOnHovercard(target, card, anchor, nested) {
	if (hasFocusWithin(card)) return true;
	if (!target) return false;
	if (contains(card, target)) return true;
	if (anchor && contains(anchor, target)) return true;
	if (nested?.some((card) => hasFocusWithin(card) || contains(card, target))) return true;
	return false;
}
function useAutoFocusOnHide({ store, ...props }) {
	const [autoFocusOnHide, setAutoFocusOnHide] = useState(false);
	const mounted = useStoreState(store, "mounted");
	useEffect(() => {
		if (!mounted) setAutoFocusOnHide(false);
	}, [mounted]);
	const onFocusProp = props.onFocus;
	const onFocus = useEvent((event) => {
		onFocusProp?.(event);
		if (event.defaultPrevented) return;
		setAutoFocusOnHide(true);
	});
	const finalFocusRef = useRef(null);
	useEffect(() => {
		return sync(store, ["anchorElement"], (state) => {
			finalFocusRef.current = state.anchorElement;
		});
	}, [store]);
	props = {
		autoFocusOnHide,
		finalFocus: finalFocusRef,
		...props,
		onFocus
	};
	return props;
}
const NestedHovercardContext = createContext(null);
/**
* Returns props to create a `Hovercard` component.
* @see https://ariakit.com/components/hovercard
* @example
* ```jsx
* const store = useHovercardStore();
* const props = useHovercard({ store });
* <HovercardAnchor store={store}>@username</HovercardAnchor>
* <Role {...props}>Details</Role>
* ```
*/
const useHovercard = createHook(function useHovercard({ store, modal = false, portal = modal, hideOnEscape = true, hideOnHoverOutside = true, disablePointerEventsOnApproach = !!hideOnHoverOutside, ...props }) {
	const context = useHovercardProviderContext();
	store = store || context;
	invariant(store, process.env.NODE_ENV !== "production" && "Hovercard must receive a `store` prop or be wrapped in a HovercardProvider component.");
	const ref = useRef(null);
	const [nestedHovercards, setNestedHovercards] = useState([]);
	const hideTimeoutRef = useRef(0);
	const enterPointRef = useRef(null);
	const { portalRef, domReady } = usePortalRef(portal, props.portalRef);
	const isMouseMoving = useIsMouseMoving();
	const mayHideOnHoverOutside = !!hideOnHoverOutside;
	const hideOnHoverOutsideProp = useBooleanEvent(hideOnHoverOutside);
	const mayDisablePointerEvents = !!disablePointerEventsOnApproach;
	const disablePointerEventsProp = useBooleanEvent(disablePointerEventsOnApproach);
	const open = useStoreState(store, "open");
	const mounted = useStoreState(store, "mounted");
	useEffect(() => {
		if (!domReady) return;
		if (!mounted) return;
		if (!mayHideOnHoverOutside && !mayDisablePointerEvents) return;
		const element = ref.current;
		if (!element) return;
		const onMouseMove = (event) => {
			if (!store) return;
			if (!isMouseMoving()) return;
			const { anchorElement, hideTimeout, timeout } = store.getState();
			const enterPoint = enterPointRef.current;
			const [target] = event.composedPath();
			const anchor = anchorElement;
			if (isMovingOnHovercard(target, element, anchor, nestedHovercards)) {
				enterPointRef.current = target && anchor && contains(anchor, target) ? getEventPoint(event) : null;
				window.clearTimeout(hideTimeoutRef.current);
				hideTimeoutRef.current = 0;
				return;
			}
			if (hideTimeoutRef.current) return;
			if (enterPoint) {
				const currentPoint = getEventPoint(event);
				if (isPointInPolygon(currentPoint, getElementPolygon(element, enterPoint))) {
					enterPointRef.current = currentPoint;
					if (!disablePointerEventsProp(event)) return;
					event.preventDefault();
					event.stopPropagation();
					return;
				}
			}
			if (!hideOnHoverOutsideProp(event)) return;
			hideTimeoutRef.current = window.setTimeout(() => {
				hideTimeoutRef.current = 0;
				store?.hide();
			}, hideTimeout ?? timeout);
		};
		return chain(addGlobalEventListener("mousemove", onMouseMove, true), () => clearTimeout(hideTimeoutRef.current));
	}, [
		store,
		isMouseMoving,
		domReady,
		mounted,
		mayHideOnHoverOutside,
		mayDisablePointerEvents,
		nestedHovercards,
		disablePointerEventsProp,
		hideOnHoverOutsideProp
	]);
	useEffect(() => {
		if (!domReady) return;
		if (!mounted) return;
		if (!mayDisablePointerEvents) return;
		const disableEvent = (event) => {
			const element = ref.current;
			if (!element) return;
			const enterPoint = enterPointRef.current;
			if (!enterPoint) return;
			const polygon = getElementPolygon(element, enterPoint);
			if (isPointInPolygon(getEventPoint(event), polygon)) {
				if (!disablePointerEventsProp(event)) return;
				event.preventDefault();
				event.stopPropagation();
			}
		};
		return chain(addGlobalEventListener("mouseenter", disableEvent, true), addGlobalEventListener("mouseover", disableEvent, true), addGlobalEventListener("mouseout", disableEvent, true), addGlobalEventListener("mouseleave", disableEvent, true));
	}, [
		domReady,
		mounted,
		mayDisablePointerEvents,
		disablePointerEventsProp
	]);
	useEffect(() => {
		if (!domReady) return;
		if (open) return;
		store?.setAutoFocusOnShow(false);
	}, [
		store,
		domReady,
		open
	]);
	const openRef = useLiveRef(open);
	useEffect(() => {
		if (!domReady) return;
		return () => {
			if (!openRef.current) store?.setAutoFocusOnShow(false);
		};
	}, [store, domReady]);
	const registerOnParent = useContext(NestedHovercardContext);
	useSafeLayoutEffect(() => {
		if (modal) return;
		if (!portal) return;
		if (!mounted) return;
		if (!domReady) return;
		const element = ref.current;
		if (!element) return;
		return registerOnParent?.(element);
	}, [
		modal,
		portal,
		mounted,
		domReady
	]);
	const registerNestedHovercard = useCallback((element) => {
		setNestedHovercards((prevElements) => [...prevElements, element]);
		const parentUnregister = registerOnParent?.(element);
		return () => {
			setNestedHovercards((prevElements) => prevElements.filter((item) => item !== element));
			parentUnregister?.();
		};
	}, [registerOnParent]);
	props = useWrapElement(props, (element) => /* @__PURE__ */ jsx(HovercardScopedContextProvider, {
		value: store,
		children: /* @__PURE__ */ jsx(NestedHovercardContext.Provider, {
			value: registerNestedHovercard,
			children: element
		})
	}), [store, registerNestedHovercard]);
	props = {
		...props,
		ref: useMergeRefs(ref, props.ref)
	};
	props = useAutoFocusOnHide({
		store,
		...props
	});
	const autoFocusOnShow = useStoreState(store, (state) => modal || state.autoFocusOnShow);
	props = usePopover({
		store,
		modal,
		portal,
		autoFocusOnShow,
		...props,
		portalRef,
		hideOnEscape(event) {
			if (isFalsyBooleanCallback(hideOnEscape, event)) return false;
			requestAnimationFrame(() => {
				requestAnimationFrame(() => {
					store?.hide();
				});
			});
			return true;
		}
	});
	return props;
});
/**
* Renders a hovercard element, which is a popover that's usually made visible
* by hovering the mouse cursor over a
* [`HovercardAnchor`](https://ariakit.com/reference/hovercard-anchor).
* @see https://ariakit.com/components/hovercard
* @example
* ```jsx {3}
* <HovercardProvider>
*   <HovercardAnchor>@username</HovercardAnchor>
*   <Hovercard>Details</Hovercard>
* </HovercardProvider>
* ```
*/
const Hovercard = createDialogComponent(forwardRef(function Hovercard(props) {
	return createElement(TagName, useHovercard(props));
}), useHovercardProviderContext);
//#endregion
export { Hovercard, useHovercard };
