// @medusajs/index/dist/utils/default-schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultSchema = void 0;
const utils_1 = require("@medusajs/utils");
exports.defaultSchema = `
  type Product @Listeners(values: ["${utils_1.Modules.PRODUCT}.product.created", "${utils_1.Modules.PRODUCT}.product.updated", "${utils_1.Modules.PRODUCT}.product.deleted", "${utils_1.Modules.PRODUCT}.product.restored"]) {
    id: ID
    title: String
    handle: String
    status: String
    type_id: String
    collection_id: String
    is_giftcard: Boolean
    external_id: String
    created_at: DateTime
    updated_at: DateTime

    variants: [ProductVariant]
    sales_channels: [SalesChannel]
  }

  type ProductVariant @Listeners(values: ["${utils_1.Modules.PRODUCT}.product-variant.created", "${utils_1.Modules.PRODUCT}.product-variant.updated", "${utils_1.Modules.PRODUCT}.product-variant.deleted", "${utils_1.Modules.PRODUCT}.product-variant.restored"]) {
    id: ID
    product_id: String
    sku: String

    prices: [Price]
  }
  
  type Price @Listeners(values: ["${utils_1.Modules.PRICING}.price.created", "${utils_1.Modules.PRICING}.price.updated", "${utils_1.Modules.PRICING}.price.deleted", "${utils_1.Modules.PRICING}.price.restored"]) {
    id: ID
    amount: Float
    currency_code: String
    price_list_id: String
  }

  type SalesChannel @Listeners(values: ["${utils_1.Modules.SALES_CHANNEL}.sales-channel.created", "${utils_1.Modules.SALES_CHANNEL}.sales-channel.updated", "${utils_1.Modules.SALES_CHANNEL}.sales-channel.deleted", "${utils_1.Modules.SALES_CHANNEL}.sales-channel.restored"]) {
    id: ID
    is_disabled: Boolean
  }
`;
