// @medusajs/js-sdk/dist/auth/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Auth = void 0;
class Auth {
    constructor(client, config) {
        /**
         * Methods for managing and completing multi-factor authentication (MFA).
         */
        this.mfa = {
            /**
             * This method retrieves the MFA factors configured for the authenticated
             * identity. It sends a request to the
             * [List MFA Factors](https://docs.medusajs.com/api/admin#auth_getmfa_factors)
             * API route.
             *
             * @param headers - Headers to pass in the request.
             * @returns The configured MFA factors.
             *
             * @tags auth
             *
             * @example
             * const { mfa_factors } = await sdk.auth.mfa.list()
             */
            list: async (headers) => {
                return await this.client.fetch("/auth/mfa/factors", {
                    headers,
                });
            },
            /**
             * This method starts MFA setup for the authenticated identity. It sends a
             * request to the
             * [Create MFA Factor](https://docs.medusajs.com/api/admin#auth_postmfa_factors)
             * API route.
             *
             * @param body - The MFA setup details.
             * @param headers - Headers to pass in the request.
             * @returns The pending MFA factor and any setup details returned by the provider.
             *
             * @tags auth
             *
             * @example
             * const setup = await sdk.auth.mfa.start({
             *   provider: "totp",
             *   label: "Authenticator app"
             * })
             *
             * // Render setup.otpauth_url as a QR code or show setup.secret manually.
             */
            start: async (body, headers) => {
                return await this.client.fetch("/auth/mfa/factors", {
                    method: "POST",
                    body,
                    headers,
                });
            },
            /**
             * This method verifies a pending MFA factor setup. It sends a request to the
             * [Verify MFA Factor](https://docs.medusajs.com/api/admin#auth_postmfa_factorsidverify)
             * API route.
             *
             * @param id - The ID of the MFA factor to verify.
             * @param body - The verification details.
             * @param headers - Headers to pass in the request.
             * @returns The verified MFA factor.
             *
             * @tags auth
             *
             * @example
             * const { mfa_factor } = await sdk.auth.mfa.verify("authmfa_123", {
             *   code: "123456"
             * })
             */
            verify: async (id, body, headers) => {
                return await this.client.fetch(`/auth/mfa/factors/${id}/verify`, {
                    method: "POST",
                    body,
                    headers,
                });
            },
            /**
             * This method disables an MFA factor for the authenticated identity. It
             * sends a request to the
             * [Delete MFA Factor](https://docs.medusajs.com/api/admin#auth_deletemfa_factorsid)
             * API route.
             *
             * @param id - The ID of the MFA factor to disable.
             * @param body - Optional verification details required by the server configuration.
             * @param headers - Headers to pass in the request.
             * @returns The disabled MFA factor.
             *
             * @tags auth
             *
             * @example
             * const { mfa_factor } = await sdk.auth.mfa.disable("authmfa_123")
             */
            disable: async (id, body, headers) => {
                return await this.client.fetch(`/auth/mfa/factors/${id}`, {
                    method: "DELETE",
                    body: body ?? {},
                    headers,
                });
            },
            /**
             * This method generates new recovery codes for the authenticated identity.
             * It sends a request to the
             * [Generate MFA Recovery Codes](https://docs.medusajs.com/api/admin#auth_postmfa_recovery-codes)
             * API route.
             *
             * @param body - Optional recovery code generation details.
             * @param headers - Headers to pass in the request.
             * @returns The generated recovery codes.
             *
             * @tags auth
             *
             * @example
             * const { recovery_codes } = await sdk.auth.mfa.generateRecoveryCodes()
             */
            generateRecoveryCodes: async (body, headers) => {
                return await this.client.fetch("/auth/mfa/recovery-codes", {
                    method: "POST",
                    body: body ?? {},
                    headers,
                });
            },
            /**
             * This method verifies an MFA challenge returned from `sdk.auth.login` or
             * `sdk.auth.callback`. It sends a request to the
             * [Verify MFA Challenge](https://docs.medusajs.com/api/admin#auth_postmfa_challengesidverify)
             * API route.
             *
             * If verification succeeds, the returned token is stored based on the SDK's
             * auth configuration, matching `sdk.auth.login`.
             *
             * @param id - The ID of the MFA challenge to verify.
             * @param body - The challenge verification details.
             * @param headers - Headers to pass in the request.
             * @returns The authentication JWT token.
             *
             * @tags auth
             *
             * @example
             * const result = await sdk.auth.login("user", "emailpass", {
             *   email: "user@example.com",
             *   password: "secret"
             * })
             *
             * if (typeof result === "object" && "mfa_challenge" in result) {
             *   await sdk.auth.mfa.verifyChallenge(result.mfa_challenge.id, {
             *     method: "totp",
             *     code: "123456"
             *   })
             * }
             */
            verifyChallenge: async (id, body, headers) => {
                const { token } = await this.client.fetch(`/auth/mfa/challenges/${id}/verify`, {
                    method: "POST",
                    body,
                    headers,
                });
                await this.setToken_(token);
                return token;
            },
        };
        /**
         * Methods for requesting and confirming verification.
         *
         * @since 2.15.5
         */
        this.verification = {
            /**
             * This method requests a verification token for an auth identity.
             *
             * @param body - The verification request details.
             * @param headers - Headers to pass in the request. Must include a bearer token for the auth identity.
             *
             * @tags auth
             */
            request: async (body, headers) => {
                return await this.client.fetch(`/auth/verification/request`, {
                    method: "POST",
                    body,
                    headers,
                });
            },
            /**
             * This method confirms a verification code.
             *
             * @param body - The verification code details.
             * @param headers - Headers to pass in the request.
             *
             * @tags auth
             */
            confirm: async (body, headers) => {
                return await this.client.fetch(`/auth/verification/confirm`, {
                    method: "POST",
                    body,
                    headers,
                });
            },
        };
        /**
         * This method is used to retrieve a registration JWT token for a user, customer, or custom actor type. It sends a request to the
         * [Retrieve Registration Token API route](https://docs.medusajs.com/api/store#auth_postactor_typeauth_provider_register).
         *
         * Then, it stores the returned token and passes it in the header of subsequent requests. So, you can call the
         * [store.customer.create](https://docs.medusajs.com/resources/references/js-sdk/store/customer#create) method,
         * for example, after calling this method.
         *
         * Learn more in the [JS SDK Authentication](https://docs.medusajs.com/resources/js-sdk/auth/overview) guide.
         *
         * @param actor - The actor type. For example, `user` for admin user, or `customer` for customer.
         * @param method - The authentication provider to use. For example, `emailpass` or `google`.
         * @param payload - The data to pass in the request's body for authentication. When using the `emailpass` provider,
         * you pass the email and password.
         * @param options - Optional behavior for multi-step registration responses.
         * @returns The JWT token used for registration later.
         *
         * @tags auth
         *
         * @example
         * await sdk.auth.register(
         *   "customer",
         *   "emailpass",
         *   {
         *     email: "customer@gmail.com",
         *     password: "supersecret"
         *   }
         * )
         *
         * // all subsequent requests will use the token in the header
         * const { customer } = await sdk.store.customer.create({
         *   email: "customer@gmail.com",
         *   password: "supersecret"
         * })
         */
        this.register = async (actor, method, payload) => {
            const { token } = await this.client.fetch(`/auth/${actor}/${method}/register`, {
                method: "POST",
                body: payload,
            });
            if (!token) {
                throw new Error("Unexpected registration response");
            }
            // The reason we don't use setToken_ (i.e. start a session) here is because the token doesn't have any actor types attached to it yet.
            // The token should be sent in a separate request to create an actor type, after which we can initiate the session.
            this.client.setToken(token);
            return token;
        };
        /**
         * This method retrieves the JWT authenticated token for an admin user, customer, or custom
         * actor type. It sends a request to the [Authenticate API Route](https://docs.medusajs.com/api/admin#auth_postactor_typeauth_provider).
         *
         * ### Third-Party Authentication
         *
         * If the API route returns a `location` property, it means that the authentication requires additional steps,
         * typically in a third-party service. The `location` property is returned so that you
         * can redirect the user to the appropriate page.
         *
         * :::note
         *
         * For an example of implementing third-party authentication, refer to the
         * [Third-Party Login in Storefront](https://docs.medusajs.com/resources/storefront-development/customers/third-party-login) guide.
         *
         * :::
         *
         * ### Session Authentication
         *
         * If the `auth.type` of the SDK is set to `session`, this method will also send a request to the
         * [Set Authentication Session API route](https://docs.medusajs.com/api/admin#auth_postsession).
         *
         * Learn more in the [JS SDK Authentication](https://docs.medusajs.com/resources/js-sdk/auth/overview) guide.
         *
         * ### Automatic Authentication
         *
         * If the authentication was successful, subsequent requests using the SDK will automatically have the necessary authentication headers / session
         * set, based on your JS SDK authentication configurations.
         *
         * Learn more in the [JS SDK Authentication](https://docs.medusajs.com/resources/js-sdk/auth/overview) guide.
         *
         * @param actor - The actor type. For example, `user` for admin user, or `customer` for customer.
         * @param method - The authentication provider to use. For example, `emailpass` or `google`.
         * @param payload - The data to pass in the request's body for authentication. When using the `emailpass` provider,
         * you pass the email and password.
         * @returns The authentication JWT token
         *
         * @tags auth
         *
         * @example
         * const result = await sdk.auth.login(
         *   "customer",
         *   "emailpass",
         *   {
         *     email: "customer@gmail.com",
         *     password: "supersecret"
         *   }
         * )
         *
         * if (typeof result !== "string") {
         *   alert("Authentication requires additional steps")
         *   // replace with the redirect logic of your application
         *   window.location.href = result.location
         *   return
         * }
         *
         * // customer is now authenticated
         * // all subsequent requests will use the token in the header
         * const { customer } = await sdk.store.customer.retrieve()
         */
        this.login = async (actor, method, payload) => {
            // There will either be token, location, MFA challenge, or verification returned from the backend.
            const { token, location, mfa_challenge, verification_required, verification, } = await this.client.fetch(`/auth/${actor}/${method}`, {
                method: "POST",
                body: payload,
            });
            // In the case of an oauth login, we return the redirect location to the caller.
            // They can decide if they do an immediate redirect or put it in an <a> tag.
            if (location) {
                return { location };
            }
            if (!token) {
                throw new Error("Unexpected authentication response");
            }
            await this.setToken_(token);
            if (verification_required) {
                return {
                    verification_required: true,
                    verification,
                    token,
                };
            }
            if (mfa_challenge) {
                return {
                    mfa_required: true,
                    mfa_challenge,
                    token,
                };
            }
            return token;
        };
        /**
         * This method is used to validate an Oauth callback from a third-party service, such as Google, for an admin user, customer, or custom actor types.
         * It sends a request to the [Validate Authentication Callback](https://docs.medusajs.com/api/admin#auth_postactor_typeauth_providercallback).
         *
         * The method stores the returned token and passes it in the header of subsequent requests. So, you can call the
         * [store.customer.create](https://docs.medusajs.com/resources/references/js-sdk/store/customer#create) or {@link refresh} methods,
         * for example, after calling this method.
         *
         * Learn more in the [JS SDK Authentication](https://docs.medusajs.com/resources/js-sdk/auth/overview) guide.
         *
         * @param actor - The actor type. For example, `user` for admin user, or `customer`.
         * @param method - The authentication provider to use. For example, `google`.
         * @param query - The query parameters from the Oauth callback, which should be passed to the API route. This includes query parameters like
         * `code` and `state`.
         * @returns The authentication JWT token
         *
         * @tags auth
         *
         * @example
         * await sdk.auth.callback(
         *   "customer",
         *   "google",
         *   {
         *     code: "123",
         *     state: "456"
         *   }
         * )
         *
         * // all subsequent requests will use the token in the header
         * const { customer } = await sdk.store.customer.create({
         *   email: "customer@gmail.com",
         *   password: "supersecret"
         * })
         *
         * @privateRemarks
         * The callback expects all query parameters from the Oauth callback to be passed to
         * the backend, and the provider is in charge of parsing and validating them
         */
        this.callback = async (actor, method, query) => {
            const { token, mfa_challenge, verification_required, verification } = await this.client.fetch(`/auth/${actor}/${method}/callback`, {
                method: "GET",
                query,
            });
            if (!token) {
                throw new Error("Unexpected authentication callback response");
            }
            await this.setToken_(token);
            if (verification_required) {
                return {
                    verification_required: true,
                    verification,
                    token,
                };
            }
            if (mfa_challenge && token) {
                return {
                    mfa_required: true,
                    mfa_challenge,
                    token,
                };
            }
            return token;
        };
        /**
         * This method refreshes a JWT authentication token, which is useful after validating the Oauth callback
         * with {@link callback}. It sends a request to the [Refresh Authentication Token API route](https://docs.medusajs.com/api/admin#auth_postadminauthtokenrefresh).
         *
         * The method stores the returned token and passes it in the header of subsequent requests. So, you can call other
         * methods that require authentication after calling this method.
         *
         * Learn more in the [JS SDK Authentication](https://docs.medusajs.com/resources/js-sdk/auth/overview) guide.
         *
         * For an example of implementing third-party authentication, refer to the
         * [Third-Party Login in Storefront](https://docs.medusajs.com/resources/storefront-development/customers/third-party-login) guide.
         *
         * @param headers - Headers to pass in the request
         *
         * @returns The refreshed JWT authentication token, or any requirements such as MFA or verification.
         *
         * @tags auth
         *
         * @example
         * const token = await sdk.auth.refresh()
         *
         * // all subsequent requests will use the token in the header
         * const { customer } = await sdk.store.customer.retrieve()
         */
        this.refresh = async (headers) => {
            const response = await this.client.fetch("/auth/token/refresh", {
                method: "POST",
                headers,
            });
            // Putting the token in session after refreshing is only useful when the new token has updated info (eg. actor_id).
            // Ideally we don't use the full JWT in session as key, but just store a pseudorandom key that keeps the rest of the auth context as value.
            await this.setToken_(response.token);
            return response;
        };
        /**
         * This method logs out the currently authenticated user based on your JS SDK authentication configurations.
         *
         * If the `auth.type` of the SDK is set to `session`, this method will also send a request to the
         * [Delete Authentication Session API route](https://docs.medusajs.com/api/admin#auth_deletesession).
         *
         * The method also clears any stored tokens or sessions, based on your JS SDK authentication configurations.
         *
         * Learn more in the [JS SDK Authentication](https://docs.medusajs.com/resources/js-sdk/auth/overview) guide.
         *
         * @tags auth
         *
         * @example
         * await sdk.auth.logout()
         *
         * // user is now logged out
         * // you can't send any requests that require authentication
         */
        this.logout = async () => {
            if (this.config?.auth?.type === "session") {
                await this.client.fetch("/auth/session", {
                    method: "DELETE",
                });
            }
            this.client.clearToken();
        };
        /**
         * This method requests a reset password token for an admin user, customer, or custom actor type.
         * It sends a request to the [Generate Reset Password Token API route](https://docs.medusajs.com/api/admin#auth_postactor_typeauth_providerresetpassword).
         *
         * To reset the password later using the token delivered to the user, use the {@link updateProvider} method.
         *
         * Related guide: [How to allow customers to reset their passwords in a storefront](https://docs.medusajs.com/resources/storefront-development/customers/reset-password).
         *
         * @param actor - The actor type. For example, `user` for admin user, or `customer` for customer.
         * @param provider - The authentication provider to use. For example, `emailpass`.
         * @param body - The data required to identify the user.
         *
         * @tags auth
         *
         * @example
         * sdk.auth.resetPassword(
         *   "customer",
         *   "emailpass",
         *   {
         *     identifier: "customer@gmail.com"
         *   }
         * )
         * .then(() => {
         *   // user receives token
         * })
         */
        this.resetPassword = async (actor, provider, body) => {
            await this.client.fetch(`/auth/${actor}/${provider}/reset-password`, {
                method: "POST",
                body,
                headers: { accept: "text/plain" }, // 201 Created response
            });
        };
        /**
         * This method is used to update user-related data authentication data.
         *
         * More specifically, use this method when updating the password of an admin user, customer, or
         * custom actor type after requesting to reset their password with {@link resetPassword}.
         *
         * This method sends a request to [this API route](https://docs.medusajs.com/api/admin#auth_postactor_typeauth_providerupdate).
         *
         * Related guide: [How to allow customers to reset their passwords in a storefront](https://docs.medusajs.com/resources/storefront-development/customers/reset-password).
         *
         * @param actor - The actor type. For example, `user` for admin user, or `customer` for customer.
         * @param provider - The authentication provider to use. For example, `emailpass`.
         * @param body - The data necessary to update the user's authentication data. When resetting the user's password,
         * send the `password` property.
         *
         * @tags auth
         *
         * @example
         * sdk.auth.updateProvider(
         *   "customer",
         *   "emailpass",
         *   {
         *     password: "supersecret"
         *   },
         *   token
         * )
         * .then(() => {
         *   // password updated
         * })
         */
        this.updateProvider = async (actor, provider, body, token) => {
            await this.client.fetch(`/auth/${actor}/${provider}/update`, {
                method: "POST",
                body,
                headers: { Authorization: `Bearer ${token}` },
            });
        };
        /**
         * @ignore
         */
        this.setToken_ = async (token) => {
            // By default we just set the token in the configured storage, if configured to use sessions we convert it into session storage instead.
            if (this.config?.auth?.type === "session") {
                await this.client.fetch("/auth/session", {
                    method: "POST",
                    headers: { Authorization: `Bearer ${token}` },
                });
            }
            else {
                this.client.setToken(token);
            }
        };
        this.client = client;
        this.config = config;
    }
}
exports.Auth = Auth;
