// @medusajs/cli/dist/commands/new.js
"use strict";
/*
 * Adapted from https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-cli/src/init-starter.ts
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.newStarter = exports.setupDB = exports.createDatabase = void 0;
const execa_1 = __importDefault(require("execa"));
const fs_exists_cached_1 = require("fs-exists-cached");
const fs_extra_1 = __importDefault(require("fs-extra"));
const hosted_git_info_1 = __importDefault(require("hosted-git-info"));
const is_valid_path_1 = __importDefault(require("is-valid-path"));
const path_1 = __importDefault(require("path"));
const prompts_1 = __importDefault(require("prompts"));
const pg_1 = require("@medusajs/deps/pg");
const url_1 = __importDefault(require("url"));
const telemetry_1 = require("@medusajs/telemetry");
// @ts-ignore
const inquirer_1 = __importDefault(require("inquirer"));
const utils_1 = require("@medusajs/utils");
const reporter_1 = __importDefault(require("../reporter"));
const panic_handler_1 = require("../reporter/panic-handler");
const package_manager_1 = require("../util/package-manager");
const STARTER_PATH = `medusajs/dtc-starter`;
const removeUndefined = (obj) => {
    return Object.fromEntries(Object.entries(obj)
        .filter(([_, v]) => v != null)
        .map(([k, v]) => [k, v === Object(v) ? removeUndefined(v) : v]));
};
const spawnWithArgs = async (file, args, options) => (0, execa_1.default)(file, args, { stdio: `inherit`, preferLocal: false, ...options });
const spawn = async (cmd, options) => {
    const [file, ...args] = cmd.split(/\s+/);
    return spawnWithArgs(file, args, options);
};
// Executes `pnpm install`, `yarn install`, or `npm install` in rootPath.
const install = async (rootPath) => {
    const prevDir = process.cwd();
    reporter_1.default.info(`Installing packages...`);
    console.log(); // Add some space
    process.chdir(rootPath);
    try {
        await spawn(`pnpm install`, {});
    }
    finally {
        process.chdir(prevDir);
    }
};
// Clones starter from URI.
const clone = async (rootPath, inputBranch) => {
    const hostedInfo = hosted_git_info_1.default.fromUrl(STARTER_PATH);
    let url;
    // Let people use private repos accessed over SSH.
    if (hostedInfo.getDefaultRepresentation() === `sshurl`) {
        url = hostedInfo.ssh({ noCommittish: true });
        // Otherwise default to normal git syntax.
    }
    else {
        url = hostedInfo.https({ noCommittish: true, noGitPlus: true });
    }
    const branch = [`-b`, inputBranch || "main"];
    const createAct = reporter_1.default.activity(`Creating new project from git: ${url}`);
    const args = [
        `clone`,
        ...branch,
        url,
        rootPath,
        `--recursive`,
        `--depth=1`,
    ].filter((arg) => Boolean(arg));
    await (0, execa_1.default)(`git`, args, {})
        .then(() => {
        reporter_1.default.success(createAct, `Created starter directory layout`);
    })
        .catch((err) => {
        reporter_1.default.failure(createAct, `Failed to clone repository`);
        throw err;
    });
    await fs_extra_1.default.remove(path_1.default.join(rootPath, `.git`));
    await fs_extra_1.default.remove(path_1.default.join(rootPath, `.github`));
    // remove storefront from monorepo
    await fs_extra_1.default.remove(path_1.default.join(rootPath, `apps`, `storefront`));
    await install(rootPath);
};
const getPaths = async (rootPath) => {
    // if no args are passed, prompt user for path and starter
    if (!rootPath) {
        const response = await prompts_1.default.prompt([
            {
                type: `text`,
                name: `path`,
                message: `What is your project called?`,
                initial: `my-medusa-store`,
            },
        ]);
        // exit gracefully if responses aren't provided
        if (!response.path.trim()) {
            throw new Error(`Please mention both starter package and project name along with path(if its not in the root)`);
        }
        rootPath = response.path;
    }
    // set defaults if no root or starter has been set yet
    rootPath = rootPath || process.cwd();
    return { rootPath };
};
const successMessage = (path) => {
    reporter_1.default.info(`Your new Medusa project is ready for you! To start developing run:

  cd ${path}
  pnpm dev
`);
};
const defaultDBCreds = {
    user: process.env.USER || "postgres",
    database: "postgres",
    password: "",
    port: 5432,
    host: "localhost",
};
const verifyPgCreds = async (creds) => {
    const pool = new pg_1.Pool(creds);
    return new Promise((resolve, reject) => {
        pool.query("SELECT NOW()", async (err, res) => {
            await pool.end();
            if (err) {
                reject(err);
            }
            else {
                resolve(res);
            }
        });
    });
};
const interactiveDbCreds = async (dbName, dbCreds = {}) => {
    const credentials = Object.assign({}, defaultDBCreds, dbCreds);
    const collecting = true;
    while (collecting) {
        const result = await inquirer_1.default
            .prompt([
            {
                type: "list",
                name: "continueWithDefault",
                message: `

Will attempt to setup Postgres database "${dbName}" with credentials:
  user: ${credentials.user}
  password: ***
  port: ${credentials.port}
  host: ${credentials.host}
Do you wish to continue with these credentials?

          `,
                choices: [`Continue`, `Change credentials`, `Skip database setup`],
            },
            {
                type: "input",
                when: ({ continueWithDefault }) => continueWithDefault === `Change credentials`,
                name: "user",
                default: credentials.user,
                message: `DB user`,
            },
            {
                type: "password",
                when: ({ continueWithDefault }) => continueWithDefault === `Change credentials`,
                name: "password",
                default: credentials.password,
                message: `DB password`,
            },
            {
                type: "number",
                when: ({ continueWithDefault }) => continueWithDefault === `Change credentials`,
                name: "port",
                default: credentials.port,
                message: `DB port`,
            },
            {
                type: "input",
                when: ({ continueWithDefault }) => continueWithDefault === `Change credentials`,
                name: "host",
                default: credentials.host,
                message: `DB host`,
            },
        ])
            .then(async (answers) => {
            const collectedCreds = Object.assign({}, credentials, {
                user: answers.user,
                password: answers.password,
                host: answers.host,
                port: answers.port,
            });
            switch (answers.continueWithDefault) {
                case "Continue": {
                    const done = await verifyPgCreds(credentials).catch((_) => false);
                    if (done) {
                        return credentials;
                    }
                    return false;
                }
                case "Change credentials": {
                    const done = await verifyPgCreds(collectedCreds).catch((_) => false);
                    if (done) {
                        return collectedCreds;
                    }
                    return false;
                }
                default:
                    return null;
            }
        });
        if (result !== false) {
            return result;
        }
        console.log("\n\nCould not verify DB credentials - please try again\n\n");
    }
    return;
};
const createDatabase = async (dbName, credentials) => {
    const pool = new pg_1.Pool(credentials);
    try {
        // `dbName` is always internally generated (`medusa-db-<base36>`) and is never
        // user-supplied, so interpolating it here is safe. CREATE DATABASE also cannot
        // use a parameterized identifier, so interpolation is the only option.
        await pool.query(`CREATE DATABASE "${dbName}"`);
    }
    finally {
        await pool.end();
    }
};
exports.createDatabase = createDatabase;
const setupDB = async (dbName, dbCreds = {}) => {
    const credentials = Object.assign({}, defaultDBCreds, dbCreds);
    const dbActivity = reporter_1.default.activity(`Setting up database "${dbName}"...`);
    await (0, exports.createDatabase)(dbName, credentials)
        .then(() => {
        reporter_1.default.success(dbActivity, `Created database "${dbName}"`);
    })
        .catch((err) => {
        // Postgres raises 42P04 (duplicate_database) when the database already exists
        if (err.code === "42P04") {
            reporter_1.default.success(dbActivity, `Database ${dbName} already exists; skipping setup`);
        }
        else {
            reporter_1.default.failure(dbActivity, `Skipping database setup.`);
            reporter_1.default.warn(`Failed to setup database; install PostgresQL or make sure to manage your database connection manually`);
            console.error(err);
        }
    });
};
exports.setupDB = setupDB;
const setupEnvVars = async (rootPath, dbName, dbCreds = {}) => {
    const templatePath = path_1.default.join(rootPath, ".env.template");
    const destination = path_1.default.join(rootPath, ".env");
    if ((0, fs_exists_cached_1.sync)(templatePath)) {
        fs_extra_1.default.renameSync(templatePath, destination);
    }
    const credentials = Object.assign({}, defaultDBCreds, dbCreds);
    let dbUrl = "";
    if (credentials.user !== defaultDBCreds.user ||
        credentials.password !== defaultDBCreds.password) {
        dbUrl = `postgres://${credentials.user}:${credentials.password}@${credentials.host}:${credentials.port}/${dbName}`;
    }
    else {
        dbUrl = `postgres://${credentials.host}:${credentials.port}/${dbName}`;
    }
    fs_extra_1.default.appendFileSync(destination, `DATABASE_URL=${dbUrl}\n`);
};
const runMigrations = async (rootPath) => {
    const migrationActivity = reporter_1.default.activity("Applying database migrations...");
    const cliPath = path_1.default.join(`node_modules`, `@medusajs`, `medusa-cli`, `cli.js`);
    return await (0, execa_1.default)(cliPath, [`migrations`, `run`], {
        cwd: rootPath,
    })
        .then(() => {
        reporter_1.default.success(migrationActivity, "Database migrations completed.");
    })
        .catch((err) => {
        reporter_1.default.failure(migrationActivity, "Failed to migrate database you must complete migration manually before starting your server.");
        console.error(err);
    });
};
const attemptSeed = async (rootPath) => {
    const seedActivity = reporter_1.default.activity("Seeding database");
    const pkgPath = path_1.default.resolve(rootPath, "package.json");
    if ((0, fs_exists_cached_1.sync)(pkgPath)) {
        const pkg = require(pkgPath);
        if (pkg.scripts && pkg.scripts.seed) {
            const proc = (0, execa_1.default)((0, package_manager_1.getPackageManager)(), [`run`, `seed`], {
                cwd: rootPath,
            });
            // Useful for development
            // proc.stdout.pipe(process.stdout)
            await proc
                .then(() => {
                reporter_1.default.success(seedActivity, "Seed completed");
            })
                .catch((err) => {
                reporter_1.default.failure(seedActivity, "Failed to complete seed; skipping");
                console.error(err);
            });
        }
        else {
            reporter_1.default.failure(seedActivity, "Starter doesn't provide a seed command; skipping.");
        }
    }
    else {
        reporter_1.default.failure(seedActivity, "Could not find package.json");
    }
};
/**
 * Main function that clones or copies the starter.
 */
const newStarter = async (args) => {
    const nodeVersion = (0, utils_1.getNodeVersion)();
    if (nodeVersion < utils_1.MIN_SUPPORTED_NODE_VERSION) {
        reporter_1.default.error(`Medusa requires at least v20 of Node.js. You're using v${nodeVersion}. Please install at least v20 and try again: https://nodejs.org/en/download`);
        process.exit(1);
    }
    (0, telemetry_1.track)("CLI_NEW");
    const { root, skipDb, skipMigrations, skipEnv, seed, useDefaults, dbUser, dbDatabase, dbPass, dbPort, dbHost, branch, } = args;
    const dbCredentials = removeUndefined({
        user: dbUser,
        database: dbDatabase,
        password: dbPass,
        port: dbPort,
        host: dbHost,
    });
    const { rootPath } = await getPaths(root);
    const urlObject = url_1.default.parse(rootPath);
    if (urlObject.protocol && urlObject.host) {
        if (/medusa-starter/gi.test(rootPath)) {
            reporter_1.default.panic({
                id: panic_handler_1.PanicId.InvalidProjectName,
                context: {
                    starter: STARTER_PATH,
                    rootPath,
                },
            });
            return;
        }
        reporter_1.default.panic({
            id: panic_handler_1.PanicId.InvalidProjectName,
            context: {
                rootPath,
            },
        });
        return;
    }
    if (!(0, is_valid_path_1.default)(rootPath)) {
        reporter_1.default.panic({
            id: panic_handler_1.PanicId.InvalidPath,
            context: {
                path: path_1.default.resolve(rootPath),
            },
        });
        return;
    }
    if ((0, fs_exists_cached_1.sync)(path_1.default.join(rootPath, `package.json`))) {
        reporter_1.default.panic({
            id: panic_handler_1.PanicId.AlreadyNodeProject,
            context: {
                rootPath,
            },
        });
        return;
    }
    await clone(rootPath, branch);
    (0, telemetry_1.track)("CLI_NEW_LAYOUT_COMPLETED");
    let creds = dbCredentials;
    const dbName = `medusa-db-${Math.random().toString(36).substring(2, 7)}`; // generate random 5 character string
    if (!useDefaults && !skipDb && !skipEnv) {
        creds = await interactiveDbCreds(dbName, dbCredentials);
    }
    if (creds === null) {
        reporter_1.default.info("Skipping automatic database setup. Please note that you need to create a database and run migrations before you can run your Medusa backend");
    }
    else {
        if (!skipDb) {
            (0, telemetry_1.track)("CLI_NEW_SETUP_DB");
            await (0, exports.setupDB)(dbName, creds);
        }
        if (!skipEnv) {
            (0, telemetry_1.track)("CLI_NEW_SETUP_ENV");
            await setupEnvVars(rootPath, dbName, creds);
        }
        if (!skipMigrations) {
            (0, telemetry_1.track)("CLI_NEW_RUN_MIGRATIONS");
            await runMigrations(rootPath);
        }
        if (seed) {
            (0, telemetry_1.track)("CLI_NEW_SEED_DB");
            await attemptSeed(rootPath);
        }
    }
    successMessage(rootPath);
    (0, telemetry_1.track)("CLI_NEW_SUCCEEDED");
};
exports.newStarter = newStarter;
