// @ariakit/react-components/dist/form/form-control.js
"use client";
import { useCollectionItem } from "../collection/collection-item.js";
import { useFormContext } from "./form-context.js";
import { createElement, createHook, forwardRef, memo, useBooleanEvent, useEvent, useId, useMergeRefs } from "@ariakit/react-utils";
import { cx, getDocument, invariant } from "@ariakit/utils";
import { useCallback, useRef } from "react";
import { useStoreState } from "@ariakit/react-store";
//#region src/form/form-control.tsx
const TagName = "input";
function getNamedElement(ref, name) {
	const element = ref.current;
	if (!element) return null;
	if (element.name === name) return element;
	if (element.form) return element.form.elements.namedItem(name);
	return getDocument(element).getElementsByName(name)[0];
}
function useItem(store, name, type) {
	return useStoreState(store, (state) => state.items.find((item) => item.type === type && item.name === name));
}
/**
* Returns props to create a `FormControl` component. Unlike `useFormInput`,
* this hook doesn't automatically returns the `value` and `onChange` props.
* This is so we can use it not only for native form elements but also for
* custom components whose value is not controlled by the native `value` and
* `onChange` props.
* @see https://ariakit.com/components/form
* @example
* ```jsx
* const store = useFormStore({ defaultValues: { content: "" } });
* const props = useFormControl({ store, name: store.names.content });
* const value = store.useValue(store.names.content);
*
* <Form store={store}>
*   <FormLabel name={store.names.content}>Content</FormLabel>
*   <Role
*     {...props}
*     value={value}
*     onChange={(value) => store.setValue(store.names.content, value)}
*     render={<Editor />}
*   />
* </Form>
* ```
*/
const useFormControl = createHook(function useFormControl({ store, name: nameProp, getItem: getItemProp, touchOnBlur = true, ...props }) {
	const context = useFormContext();
	store = store || context;
	invariant(store, process.env.NODE_ENV !== "production" && "FormControl must be wrapped in a Form component.");
	const name = String(nameProp);
	const id = useId(props.id);
	const ref = useRef(null);
	store.useValidate(async () => {
		const element = getNamedElement(ref, name);
		if (!element) return;
		await Promise.resolve();
		if ("validity" in element && !element.validity.valid) store?.setError(name, element.validationMessage);
	});
	const getItem = useCallback((item) => {
		const nextItem = {
			...item,
			id: id || item.id,
			name,
			type: "field"
		};
		if (getItemProp) return getItemProp(nextItem);
		return nextItem;
	}, [
		id,
		name,
		getItemProp
	]);
	const onBlurProp = props.onBlur;
	const touchOnBlurProp = useBooleanEvent(touchOnBlur);
	const onBlur = useEvent((event) => {
		onBlurProp?.(event);
		if (event.defaultPrevented) return;
		if (!touchOnBlurProp(event)) return;
		store?.setFieldTouched(name, true);
	});
	const label = useItem(store, name, "label");
	const error = useItem(store, name, "error");
	const description = useItem(store, name, "description");
	const describedBy = cx(error?.id, description?.id, props["aria-describedby"]);
	const invalid = useStoreState(store, () => !!store?.getError(name) && store.getFieldTouched(name));
	props = {
		"aria-labelledby": props["aria-label"] != null ? void 0 : label?.id,
		"aria-invalid": invalid,
		...props,
		id,
		"aria-describedby": describedBy || void 0,
		ref: useMergeRefs(ref, props.ref),
		onBlur
	};
	props = useCollectionItem({
		store,
		...props,
		name,
		getItem
	});
	return props;
});
/**
* Abstract component that renders a form control. Unlike
* [`FormInput`](https://ariakit.com/reference/form-input), this component
* doesn't automatically pass the `value` and `onChange` props down to the
* underlying element. This is so we can use it not only for native form
* elements but also for custom components whose value is not controlled by the
* native `value` and `onChange` props.
* @see https://ariakit.com/components/form
* @example
* ```jsx {11-19}
* const form = useFormStore({
*   defaultValues: {
*     content: "",
*   },
* });
*
* const value = form.useValue(form.names.content);
*
* <Form store={form}>
*   <FormLabel name={form.names.content}>Content</FormLabel>
*   <FormControl
*     name={form.names.content}
*     render={
*       <Editor
*         value={value}
*         onChange={(value) => form.setValue(form.names.content, value)}
*       />
*     }
*   />
* </Form>
* ```
*/
const FormControl = memo(forwardRef(function FormControl(props) {
	return createElement(TagName, useFormControl(props));
}));
//#endregion
export { FormControl, useFormControl };
