// @medusajs/medusa/dist/api/auth/mfa/challenges/[id]/verify/route.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.POST = void 0;
const utils_1 = require("@medusajs/framework/utils");
const generate_jwt_token_1 = require("../../../../utils/generate-jwt-token");
const POST = async (req, res) => {
    const { id } = req.params;
    const { method, code } = req.validatedBody;
    const authService = req.scope.resolve(utils_1.Modules.AUTH);
    const challenge = await authService.verifyAuthMfaChallenge({
        id,
        method,
        code,
    });
    if (!challenge.auth_identity_id) {
        throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, "MFA challenge is missing an auth identity");
    }
    if (!challenge.actor_type) {
        throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, "MFA challenge is missing an actor type");
    }
    const authIdentity = await authService.retrieveAuthIdentity(challenge.auth_identity_id, { relations: ["provider_identities"] });
    if (challenge.auth_provider &&
        !authIdentity.provider_identities?.some((identity) => identity.provider === challenge.auth_provider)) {
        throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `Auth identity does not have a provider identity for "${challenge.auth_provider}"`);
    }
    const { http } = req.scope.resolve(utils_1.ContainerRegistrationKeys.CONFIG_MODULE).projectConfig;
    // We don't do additinal checks here as initial authentication has to be done before doing the MFA challenge.
    const token = await (0, generate_jwt_token_1.generateJwtTokenForAuthIdentity)({
        authIdentity,
        actorType: challenge.actor_type,
        authProvider: req.auth_context.auth_provider ?? challenge.auth_provider ?? undefined,
        container: req.scope,
    }, {
        secret: http.jwtSecret,
        expiresIn: http.jwtExpiresIn,
        options: http.jwtOptions,
    });
    return res.status(200).json({ token });
};
exports.POST = POST;
