// @medusajs/js-sdk/dist/esm/admin/rbac-policy.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
/**
 * This class is used to send requests to the RBAC Policy API routes.
 *
 * @since 2.15.5
 */
export class RbacPolicy {
    /**
     * @ignore
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * This method retrieves a paginated list of RBAC policies. It sends a request to the
     * List RBAC Policies API route.
     *
     * @param queryParams - Filters and pagination configurations.
     * @param headers - Headers to pass in the request.
     * @returns The paginated list of RBAC policies.
     *
     * @example
     * To retrieve the list of policies:
     *
     * ```ts
     * sdk.admin.rbacPolicy.list()
     * .then(({ policies, count, limit, offset }) => {
     *   console.log(policies)
     * })
     * ```
     *
     * To configure the pagination, pass the `limit` and `offset` query parameters.
     *
     * For example, to retrieve only 10 items and skip 10 items:
     *
     * ```ts
     * sdk.admin.rbacPolicy.list({
     *   limit: 10,
     *   offset: 10
     * })
     * .then(({ policies, count, limit, offset }) => {
     *   console.log(policies)
     * })
     * ```
     *
     * Using the `fields` query parameter, you can specify the fields and relations to retrieve
     * in each policy:
     *
     * ```ts
     * sdk.admin.rbacPolicy.list({
     *   fields: "id,key,resource,operation"
     * })
     * .then(({ policies, count, limit, offset }) => {
     *   console.log(policies)
     * })
     * ```
     *
     * @tags rbac
     * @since 2.15.5
     */
    list(queryParams, headers) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.client.fetch(`/admin/rbac/policies`, {
                query: queryParams,
                headers,
            });
        });
    }
    /**
     * This method retrieves an RBAC policy's details. It sends a request to the
     * Get RBAC Policy API route.
     *
     * @param id - The policy's ID.
     * @param query - Configure the fields to retrieve in the policy.
     * @param headers - Headers to pass in the request.
     * @returns The policy's details.
     *
     * @example
     * sdk.admin.rbacPolicy.retrieve("policy_123")
     * .then(({ policy }) => {
     *   console.log(policy)
     * })
     *
     * @tags rbac
     * @since 2.15.5
     */
    retrieve(id, query, headers) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.client.fetch(`/admin/rbac/policies/${id}`, {
                query,
                headers,
            });
        });
    }
    /**
     * This method creates an RBAC policy. It sends a request to the Create RBAC Policy
     * API route.
     *
     * @param body - The policy's details.
     * @param query - Configure the fields to retrieve in the created policy.
     * @param headers - Headers to pass in the request
     * @returns The created policy
     *
     * @example
     * sdk.admin.rbacPolicy.create({
     *   key: "product.read",
     *   resource: "product",
     *   operation: "read"
     * })
     * .then(({ policy }) => {
     *   console.log(policy)
     * })
     *
     * @tags rbac
     * @since 2.15.5
     */
    create(body, query, headers) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.client.fetch(`/admin/rbac/policies`, {
                method: "POST",
                headers,
                body,
                query,
            });
        });
    }
    /**
     * This method updates an RBAC policy's details. It sends a request to the
     * Update RBAC Policy API route.
     *
     * @param id - The policy's ID.
     * @param body - The data to update in the policy.
     * @param query - Configure the fields to retrieve in the policy.
     * @param headers - Headers to pass in the request.
     * @returns The policy's details.
     *
     * @example
     * sdk.admin.rbacPolicy.update("policy_123", {
     *   name: "Read Products",
     *   description: "Allows reading product information"
     * })
     * .then(({ policy }) => {
     *   console.log(policy)
     * })
     *
     * @tags rbac
     * @since 2.15.5
     */
    update(id, body, query, headers) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.client.fetch(`/admin/rbac/policies/${id}`, {
                method: "POST",
                headers,
                body,
                query,
            });
        });
    }
    /**
     * This method deletes an RBAC policy by its ID. It sends a request to the
     * Delete RBAC Policy API route.
     *
     * @param id - The policy's ID.
     * @param headers - Headers to pass in the request.
     * @returns The deletion's details.
     *
     * @example
     * sdk.admin.rbacPolicy.delete("policy_123")
     * .then(({ deleted }) => {
     *   console.log(deleted)
     * })
     *
     * @tags rbac
     * @since 2.15.5
     */
    delete(id, headers) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.client.fetch(`/admin/rbac/policies/${id}`, {
                method: "DELETE",
                headers,
            });
        });
    }
    /**
     * Lists the roles that include the given policy.
     */
    listRoles(id, queryParams, headers) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.client.fetch(`/admin/rbac/policies/${id}/roles`, {
                query: queryParams,
                headers,
            });
        });
    }
    /**
     * Lists the policies the authenticated actor is allowed to assign.
     *
     * @param queryParams - Filters and pagination configurations.
     * @param headers - Headers to pass in the request.
     * @returns The assignable policies.
     */
    listAssignable(queryParams, headers) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.client.fetch(`/admin/rbac/policies/assignable`, {
                method: "GET",
                query: queryParams,
                headers,
            });
        });
    }
}
