// @ariakit/react/cjs/__chunks/D6USTXKA.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _KDVG735Hcjs = require('./KDVG735H.cjs');



var _2WBRQ3I7cjs = require('./2WBRQ3I7.cjs');


var _JBOLBTVUcjs = require('./JBOLBTVU.cjs');

// ../ariakit-react-components/dist/select/select-context.js
var _react = require('react');
var ctx = _JBOLBTVUcjs.createStoreContext.call(void 0, [_KDVG735Hcjs.PopoverContextProvider, _2WBRQ3I7cjs.CompositeContextProvider], [_KDVG735Hcjs.PopoverScopedContextProvider, _2WBRQ3I7cjs.CompositeScopedContextProvider]);
var useSelectContext = ctx.useContext;
var useSelectScopedContext = ctx.useScopedContext;
var useSelectProviderContext = ctx.useProviderContext;
var SelectContextProvider = ctx.ContextProvider;
var SelectScopedContextProvider = ctx.ScopedContextProvider;
var SelectItemCheckedContext = _react.createContext.call(void 0, false);
var SelectHeadingContext = _react.createContext.call(void 0, null);









exports.useSelectContext = useSelectContext; exports.useSelectScopedContext = useSelectScopedContext; exports.useSelectProviderContext = useSelectProviderContext; exports.SelectContextProvider = SelectContextProvider; exports.SelectScopedContextProvider = SelectScopedContextProvider; exports.SelectItemCheckedContext = SelectItemCheckedContext; exports.SelectHeadingContext = SelectHeadingContext;
