// @ariakit/react/cjs/__chunks/FNYFMCD7.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _KDVG735Hcjs = require('./KDVG735H.cjs');



var _2WBRQ3I7cjs = require('./2WBRQ3I7.cjs');


var _JBOLBTVUcjs = require('./JBOLBTVU.cjs');

// ../ariakit-react-components/dist/combobox/combobox-context.js
var _react = require('react');
var ComboboxListRoleContext = _react.createContext.call(void 0, void 0);
var ctx = _JBOLBTVUcjs.createStoreContext.call(void 0, [_KDVG735Hcjs.PopoverContextProvider, _2WBRQ3I7cjs.CompositeContextProvider], [_KDVG735Hcjs.PopoverScopedContextProvider, _2WBRQ3I7cjs.CompositeScopedContextProvider]);
var useComboboxContext = ctx.useContext;
var useComboboxScopedContext = ctx.useScopedContext;
var useComboboxProviderContext = ctx.useProviderContext;
var ComboboxContextProvider = ctx.ContextProvider;
var ComboboxScopedContextProvider = ctx.ScopedContextProvider;
var ComboboxItemValueContext = _react.createContext.call(void 0, void 0);
var ComboboxItemCheckedContext = _react.createContext.call(void 0, false);










exports.ComboboxListRoleContext = ComboboxListRoleContext; exports.useComboboxContext = useComboboxContext; exports.useComboboxScopedContext = useComboboxScopedContext; exports.useComboboxProviderContext = useComboboxProviderContext; exports.ComboboxContextProvider = ComboboxContextProvider; exports.ComboboxScopedContextProvider = ComboboxScopedContextProvider; exports.ComboboxItemValueContext = ComboboxItemValueContext; exports.ComboboxItemCheckedContext = ComboboxItemCheckedContext;
