// @ariakit/react/cjs/__chunks/P6Y2DILA.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _HCYQ5Z7Zcjs = require('./HCYQ5Z7Z.cjs');


var _2ENQD3AScjs = require('./2ENQD3AS.cjs');



var _ULDO57DJcjs = require('./ULDO57DJ.cjs');




var _JBOLBTVUcjs = require('./JBOLBTVU.cjs');

// ../ariakit-react-components/dist/hovercard/hovercard-description.js
var TagName = "p";
var useHovercardDescription = _JBOLBTVUcjs.createHook.call(void 0, function useHovercardDescription2(props) {
  props = _2ENQD3AScjs.usePopoverDescription.call(void 0, props);
  return props;
});
var HovercardDescription = _JBOLBTVUcjs.forwardRef.call(void 0, function HovercardDescription2(props) {
  return _JBOLBTVUcjs.createElement.call(void 0, TagName, useHovercardDescription(props));
});

// ../ariakit-react-components/dist/hovercard/hovercard-dismiss.js
var TagName2 = "button";
var useHovercardDismiss = _JBOLBTVUcjs.createHook.call(void 0, function useHovercardDismiss2({ store, ...props }) {
  const context = _HCYQ5Z7Zcjs.useHovercardScopedContext.call(void 0, );
  store = store || context;
  props = _ULDO57DJcjs.usePopoverDismiss.call(void 0, {
    store,
    ...props
  });
  return props;
});
var HovercardDismiss = _JBOLBTVUcjs.forwardRef.call(void 0, function HovercardDismiss2(props) {
  return _JBOLBTVUcjs.createElement.call(void 0, TagName2, useHovercardDismiss(props));
});

// ../ariakit-react-components/dist/hovercard/hovercard-heading.js
var TagName3 = "h1";
var useHovercardHeading = _JBOLBTVUcjs.createHook.call(void 0, function useHovercardHeading2(props) {
  props = _ULDO57DJcjs.usePopoverHeading.call(void 0, props);
  return props;
});
var HovercardHeading = _JBOLBTVUcjs.forwardRef.call(void 0, function HovercardHeading2(props) {
  return _JBOLBTVUcjs.createElement.call(void 0, TagName3, useHovercardHeading(props));
});








exports.useHovercardDescription = useHovercardDescription; exports.HovercardDescription = HovercardDescription; exports.useHovercardDismiss = useHovercardDismiss; exports.HovercardDismiss = HovercardDismiss; exports.useHovercardHeading = useHovercardHeading; exports.HovercardHeading = HovercardHeading;
