// @ariakit/react/cjs/__chunks/P7AY4KL3.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _TMY4OIYCcjs = require('./TMY4OIYC.cjs');







var _JBOLBTVUcjs = require('./JBOLBTVU.cjs');

// ../ariakit-react-components/dist/button/button.js
var _react = require('react');
var TagName = "button";
var useButton = _JBOLBTVUcjs.createHook.call(void 0, function useButton2(props) {
  const ref = _react.useRef.call(void 0, null);
  const tagName = _JBOLBTVUcjs.useTagName.call(void 0, ref, TagName);
  const [isNativeButton, setIsNativeButton] = _react.useState.call(void 0, () => !!tagName && _JBOLBTVUcjs.isButton.call(void 0, {
    tagName,
    type: props.type
  }));
  _react.useEffect.call(void 0, () => {
    if (!ref.current) return;
    setIsNativeButton(_JBOLBTVUcjs.isButton.call(void 0, ref.current));
  }, []);
  props = {
    role: !isNativeButton && tagName !== "a" ? "button" : void 0,
    ...props,
    ref: _JBOLBTVUcjs.useMergeRefs.call(void 0, ref, props.ref)
  };
  props = _TMY4OIYCcjs.useCommand.call(void 0, props);
  return props;
});
var Button = _JBOLBTVUcjs.forwardRef.call(void 0, function Button2(props) {
  return _JBOLBTVUcjs.createElement.call(void 0, TagName, useButton(props));
});




exports.useButton = useButton; exports.Button = Button;
