// @ariakit/react/cjs/__chunks/ULDO57DJ.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _KDVG735Hcjs = require('./KDVG735H.cjs');



var _6XKLGFUOcjs = require('./6XKLGFUO.cjs');




var _JBOLBTVUcjs = require('./JBOLBTVU.cjs');

// ../ariakit-react-components/dist/popover/popover-dismiss.js
var TagName = "button";
var usePopoverDismiss = _JBOLBTVUcjs.createHook.call(void 0, function usePopoverDismiss2({ store, ...props }) {
  const context = _KDVG735Hcjs.usePopoverScopedContext.call(void 0, );
  store = store || context;
  props = _6XKLGFUOcjs.useDialogDismiss.call(void 0, {
    store,
    ...props
  });
  return props;
});
var PopoverDismiss = _JBOLBTVUcjs.forwardRef.call(void 0, function PopoverDismiss2(props) {
  return _JBOLBTVUcjs.createElement.call(void 0, TagName, usePopoverDismiss(props));
});

// ../ariakit-react-components/dist/popover/popover-heading.js
var TagName2 = "h1";
var usePopoverHeading = _JBOLBTVUcjs.createHook.call(void 0, function usePopoverHeading2(props) {
  props = _6XKLGFUOcjs.useDialogHeading.call(void 0, props);
  return props;
});
var PopoverHeading = _JBOLBTVUcjs.forwardRef.call(void 0, function PopoverHeading2(props) {
  return _JBOLBTVUcjs.createElement.call(void 0, TagName2, usePopoverHeading(props));
});






exports.usePopoverDismiss = usePopoverDismiss; exports.PopoverDismiss = PopoverDismiss; exports.usePopoverHeading = usePopoverHeading; exports.PopoverHeading = PopoverHeading;
