// @medusajs/dashboard/dist/chunk-FLVBJYEJ.mjs
import {
  useRegions
} from "./chunk-HP2JH45P.mjs";
import {
  useSalesChannels
} from "./chunk-MUOBXNL6.mjs";

// src/hooks/table/filters/use-order-table-filters.tsx
import { useTranslation } from "react-i18next";
import { useMemo } from "react";
var useOrderTableFilters = () => {
  const { t } = useTranslation();
  const { regions } = useRegions({
    limit: 1e3,
    fields: "id,name"
  });
  const { sales_channels } = useSalesChannels({
    limit: 1e3,
    fields: "id,name"
  });
  return useMemo(() => {
    const filters = [];
    if (regions) {
      const regionFilter = {
        key: "region_id",
        label: t("fields.region"),
        type: "select",
        options: regions.map((r) => ({
          label: r.name,
          value: r.id
        })),
        multiple: true,
        searchable: true
      };
      filters.push(regionFilter);
    }
    if (sales_channels) {
      const salesChannelFilter = {
        key: "sales_channel_id",
        label: t("fields.salesChannel"),
        type: "select",
        multiple: true,
        searchable: true,
        options: sales_channels.map((s) => ({
          label: s.name,
          value: s.id
        }))
      };
      filters.push(salesChannelFilter);
    }
    const dateFilters = [
      { label: t("fields.createdAt"), key: "created_at" },
      { label: t("fields.updatedAt"), key: "updated_at" }
    ].map((f) => ({
      key: f.key,
      label: f.label,
      type: "date"
    }));
    filters.push(...dateFilters);
    return filters;
  }, [regions, sales_channels, t]);
};

export {
  useOrderTableFilters
};
