// @medusajs/dashboard/dist/chunk-AFU4RQQF.mjs
import {
  useExtension
} from "./chunk-5SZFF255.mjs";

// src/components/layout-composer/layout-composer.tsx
import { Outlet } from "react-router-dom";

// src/components/layout-composer/constants.ts
var CORE_LAYOUT_IDS = {
  SINGLE_COLUMN: "core:single-column",
  TWO_COLUMN: "core:two-column"
};
var BEFORE_ORDER = 0;
var CORE_CONTENT_ORDER = 100;
var AFTER_ORDER = 999;

// src/components/layout-composer/layout-composer.tsx
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var LayoutComposer = ({
  widgetsZonePrefix,
  preferredLayoutId,
  sections,
  data,
  hasOutlet = true
}) => {
  const { getWidgetsForSections, getLayout } = useExtension();
  const layoutId = preferredLayoutId;
  const layout = getLayout(layoutId);
  const widgetsBySection = getWidgetsForSections(
    widgetsZonePrefix,
    layout?.sections?.map((s) => s.id) ?? []
  );
  const widgetProps = { data };
  const renderedSections = {};
  for (const { id } of layout?.sections ?? []) {
    const widgets = widgetsBySection[id] ?? [];
    const before = widgets.filter((w) => w.order < CORE_CONTENT_ORDER);
    const after = widgets.filter((w) => w.order >= CORE_CONTENT_ORDER);
    renderedSections[id] = /* @__PURE__ */ jsxs(Fragment, { children: [
      before.map(({ Component }, i) => /* @__PURE__ */ jsx(Component, { ...widgetProps }, i)),
      sections[id],
      after.map(({ Component }, i) => /* @__PURE__ */ jsx(Component, { ...widgetProps }, i))
    ] });
  }
  const LayoutComponent = layout?.Component;
  if (!LayoutComponent) return null;
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(LayoutComponent, { sections: renderedSections, data }),
    hasOutlet && /* @__PURE__ */ jsx(Outlet, {})
  ] });
};

export {
  CORE_LAYOUT_IDS,
  BEFORE_ORDER,
  AFTER_ORDER,
  LayoutComposer
};
