// @medusajs/dashboard/dist/chunk-JHQC6AFI.mjs
import {
  queryKeysFactory
} from "./chunk-774WSTCC.mjs";
import {
  sdk
} from "./chunk-NFEK63OE.mjs";

// src/hooks/api/rbac-policies.tsx
import {
  useMutation,
  useQuery
} from "@tanstack/react-query";
var RBAC_POLICIES_QUERY_KEY = "rbac_policies";
var _rbacPoliciesQueryKeys = queryKeysFactory(
  RBAC_POLICIES_QUERY_KEY
);
_rbacPoliciesQueryKeys.roles = function(policyId, query) {
  return [this.detail(policyId), "roles", query].filter(Boolean);
};
var rbacPoliciesQueryKeys = _rbacPoliciesQueryKeys;
var useRbacPolicy = (id, query, options) => {
  const { data, ...rest } = useQuery({
    queryFn: () => sdk.admin.rbacPolicy.retrieve(id, query),
    queryKey: rbacPoliciesQueryKeys.detail(id, query),
    ...options
  });
  return { ...data, ...rest };
};
var useRbacPolicies = (query, options) => {
  const { data, ...rest } = useQuery({
    queryFn: () => sdk.admin.rbacPolicy.list(query),
    queryKey: rbacPoliciesQueryKeys.list(query),
    ...options
  });
  return { ...data, ...rest };
};
var useRbacPolicyRoles = (policyId, query, options) => {
  const { data, ...rest } = useQuery({
    queryFn: () => sdk.admin.rbacPolicy.listRoles(policyId, query),
    queryKey: rbacPoliciesQueryKeys.roles(policyId, query),
    ...options
  });
  return { ...data, ...rest };
};
var ASSIGNABLE_POLICIES_QUERY_KEY = ["rbac_assignable_policies"];
var useRbacAssignablePolicies = (query, options) => {
  return useQuery({
    queryFn: () => sdk.admin.rbacPolicy.listAssignable(query),
    queryKey: [...ASSIGNABLE_POLICIES_QUERY_KEY, query],
    staleTime: 60 * 1e3,
    ...options
  });
};

export {
  rbacPoliciesQueryKeys,
  useRbacPolicy,
  useRbacPolicies,
  useRbacPolicyRoles,
  useRbacAssignablePolicies
};
