// @medusajs/dashboard/dist/login-SMUUQFIP.mjs
import {
  AvatarBox
} from "./chunk-O333RR6K.mjs";
import "./chunk-KIIT4BNH.mjs";
import {
  isFetchError
} from "./chunk-ONB3JEHR.mjs";
import "./chunk-BPIG7PI6.mjs";
import {
  useExtension
} from "./chunk-5SZFF255.mjs";
import {
  Form
} from "./chunk-OBQI23QM.mjs";
import {
  callbackWithCloudAuth,
  useCloudAuthEnabled,
  useCreateCloudAuthUser,
  useVerifyAuthMfaChallenge
} from "./chunk-BVANDDMY.mjs";
import "./chunk-FAB3VINY.mjs";
import {
  useSignInWithEmailPass
} from "./chunk-EKTC7HGF.mjs";
import "./chunk-HI6URQ7H.mjs";
import "./chunk-6CLQKVAU.mjs";
import "./chunk-Q7YBE4IP.mjs";
import "./chunk-KI7TOXBR.mjs";
import "./chunk-QBIQEVTU.mjs";
import "./chunk-WAXMT4IY.mjs";
import "./chunk-CN7JXSGW.mjs";
import "./chunk-JHQC6AFI.mjs";
import "./chunk-OCESTHZJ.mjs";
import "./chunk-LGNTHZ5Y.mjs";
import "./chunk-CHDKZVIP.mjs";
import "./chunk-5U6AZEO7.mjs";
import "./chunk-KWWLGCSY.mjs";
import "./chunk-5AFMB7XQ.mjs";
import "./chunk-A63RZVX6.mjs";
import "./chunk-HBXV7ENS.mjs";
import "./chunk-ZBMENGLV.mjs";
import "./chunk-LGLPAUIO.mjs";
import "./chunk-CDORR33H.mjs";
import "./chunk-DE44HCDH.mjs";
import "./chunk-HP2JH45P.mjs";
import "./chunk-VCSSQVPD.mjs";
import "./chunk-SQDIZZDW.mjs";
import "./chunk-MUOBXNL6.mjs";
import "./chunk-BEQ5XTWH.mjs";
import "./chunk-DYDGGABK.mjs";
import "./chunk-LA2ILUXH.mjs";
import "./chunk-7AXHHXCX.mjs";
import "./chunk-FXYH54JP.mjs";
import "./chunk-774WSTCC.mjs";
import {
  sdk
} from "./chunk-NFEK63OE.mjs";
import "./chunk-QZ7TP4HQ.mjs";

// src/routes/login/login.tsx
import { zodResolver } from "@hookform/resolvers/zod";
import { Alert, Button as Button3, Heading as Heading2, Hint as Hint2, Input as Input2, Text as Text2 } from "@medusajs/ui";
import { useState as useState3 } from "react";
import { useForm } from "react-hook-form";
import { Trans, useTranslation as useTranslation3 } from "react-i18next";
import { Link as Link2, useLocation, useNavigate as useNavigate2 } from "react-router-dom";
import * as z from "zod";

// src/routes/login/components/cloud-auth-login.tsx
import { Spinner } from "@medusajs/icons";
import { Button, toast } from "@medusajs/ui";
import { useCallback, useEffect, useRef, useState } from "react";
import { useTranslation } from "react-i18next";
import { decodeToken } from "react-jwt";
import { useNavigate, useSearchParams } from "react-router-dom";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var CLOUD_AUTH_PROVIDER = "cloud";
var CloudAuthLogin = ({ onMfaChallenge }) => {
  const { t } = useTranslation();
  const [searchParams] = useSearchParams();
  const { data: cloudAuth } = useCloudAuthEnabled();
  const isAutoLogin = searchParams.get("auth_provider") === CLOUD_AUTH_PROVIDER && searchParams.get("auto") === "true";
  const isCallback = searchParams.get("auth_provider") === CLOUD_AUTH_PROVIDER && (searchParams.has("code") || searchParams.has("error"));
  const { handleLogin, isLoginPending } = useHandleLogin(isAutoLogin);
  const { handleCallback, isCallbackPending } = useAuthCallback(
    searchParams,
    onMfaChallenge
  );
  const actionInitiated = useRef(false);
  useEffect(() => {
    if (actionInitiated.current) {
      return;
    }
    if (isAutoLogin) {
      actionInitiated.current = true;
      handleLogin();
    } else if (isCallback) {
      actionInitiated.current = true;
      handleCallback();
    }
  }, [isAutoLogin, isCallback, handleLogin, handleCallback]);
  if (isAutoLogin || isCallback) {
    return /* @__PURE__ */ jsx("div", { className: "bg-ui-bg-subtle fixed inset-0 z-50 flex items-center justify-center", children: /* @__PURE__ */ jsx(Spinner, { className: "text-ui-fg-subtle animate-spin" }) });
  }
  if (!cloudAuth?.enabled) {
    return null;
  }
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx("hr", { className: "bg-ui-border-base my-4" }),
    /* @__PURE__ */ jsx(
      Button,
      {
        variant: "secondary",
        onClick: handleLogin,
        className: "w-full",
        disabled: isLoginPending || isCallbackPending,
        isLoading: isLoginPending || isCallbackPending,
        children: t("auth.login.cloud")
      }
    )
  ] });
};
var useHandleLogin = (isAutoLogin) => {
  const { t } = useTranslation();
  const navigate = useNavigate();
  const [isPending, setIsPending] = useState(false);
  const handleLogin = useCallback(async () => {
    setIsPending(true);
    try {
      const result = await sdk.auth.login("user", CLOUD_AUTH_PROVIDER, {
        // setting callback_url in case the admin is on a different domain, or the backend URL is set to just "/" which won't work for the callback
        callback_url: `${window.location.origin}${window.location.pathname}?auth_provider=${CLOUD_AUTH_PROVIDER}`
      });
      if (typeof result === "object" && "location" in result) {
        window.location.href = result.location;
        return;
      }
      throw new Error("Unexpected login response");
    } catch {
      toast.error(t("auth.login.authenticationFailed"));
      if (isAutoLogin) {
        navigate("/login");
      }
    }
    setIsPending(false);
  }, [t, navigate, isAutoLogin]);
  return { handleLogin, isLoginPending: isPending };
};
var useAuthCallback = (searchParams, onMfaChallenge) => {
  const { t } = useTranslation();
  const navigate = useNavigate();
  const { mutateAsync: createCloudAuthUser } = useCreateCloudAuthUser();
  const [isPending, setIsPending] = useState(false);
  const ensureCloudAuthUser = useCallback(
    async (token) => {
      const decodedToken = decodeToken(token);
      if (!decodedToken?.actor_id) {
        await createCloudAuthUser();
        const refreshedToken = await sdk.auth.refresh({
          Authorization: `Bearer ${token}`
          // passing it manually in case the auth type is session
        });
        if (!refreshedToken) {
          throw new Error("Failed to refresh token after user creation");
        }
      }
    },
    [createCloudAuthUser]
  );
  const handleCallback = useCallback(async () => {
    setIsPending(true);
    try {
      let token;
      try {
        const query = Object.fromEntries(searchParams);
        delete query.auth_provider;
        const result = await callbackWithCloudAuth(query);
        if (typeof result === "object" && "mfa_challenge" in result) {
          if (!onMfaChallenge) {
            throw new Error("MFA challenge handler is missing");
          }
          onMfaChallenge?.(result.mfa_challenge, async (verifiedToken) => {
            await ensureCloudAuthUser(verifiedToken);
            navigate("/");
          });
          return;
        }
        if (typeof result === "object" && "verification_required" in result) {
          throw new Error("Verification required but not implemented yet");
        }
        token = result;
      } catch (error) {
        throw new Error("Authentication callback failed");
      }
      await ensureCloudAuthUser(token);
      navigate("/");
    } catch (error) {
      toast.error(t("auth.login.authenticationFailed"));
      navigate("/login");
    }
    setIsPending(false);
  }, [searchParams, t, onMfaChallenge, ensureCloudAuthUser, navigate]);
  return { handleCallback, isCallbackPending: isPending };
};

// src/routes/login/components/mfa-challenge-form.tsx
import { Button as Button2, Heading, Hint, Input, OtpInput, Text } from "@medusajs/ui";
import { useState as useState2 } from "react";
import { useTranslation as useTranslation2 } from "react-i18next";
import { Link } from "react-router-dom";
import { jsx as jsx2, jsxs as jsxs2 } from "react/jsx-runtime";
var getDefaultMethod = (methods) => {
  if (methods.includes("totp")) {
    return "totp";
  }
  return methods[0] ?? "totp";
};
var MfaChallengeForm = ({
  challenge,
  onSuccess,
  onBack
}) => {
  const { t } = useTranslation2();
  const [method, setMethod] = useState2(
    getDefaultMethod(challenge.methods)
  );
  const [code, setCode] = useState2("");
  const [error, setError] = useState2(null);
  const [verifiedToken, setVerifiedToken] = useState2(null);
  const [isCompleting, setIsCompleting] = useState2(false);
  const { mutateAsync, isPending } = useVerifyAuthMfaChallenge();
  const isRecoveryCode = method === "recovery_code";
  const canUseRecoveryCode = challenge.methods.includes("recovery_code");
  const canVerify = isRecoveryCode ? !!code.trim() : code.length === 6;
  const isLoading = isPending || isCompleting;
  const handleVerify = async (nextCode = code) => {
    const verificationCode = nextCode.trim();
    if (!verifiedToken && (!verificationCode || !isRecoveryCode && verificationCode.length !== 6)) {
      return;
    }
    setError(null);
    let token = verifiedToken;
    if (!token) {
      try {
        token = await mutateAsync({
          id: challenge.id,
          method,
          code: verificationCode
        });
        setVerifiedToken(token);
      } catch (e) {
        setError(e instanceof Error ? e.message : t("login.mfa.verifyError"));
        setCode("");
        return;
      }
    }
    setIsCompleting(true);
    try {
      await onSuccess(token);
    } catch (e) {
      setError(e instanceof Error ? e.message : t("login.mfa.completeError"));
    } finally {
      setIsCompleting(false);
    }
  };
  const handleOtpChange = (value) => {
    setCode(value);
  };
  const handleMethodChange = (nextMethod) => {
    setMethod(nextMethod);
    setCode("");
    setError(null);
    setVerifiedToken(null);
  };
  return /* @__PURE__ */ jsxs2("div", { className: "flex w-full flex-col items-center", children: [
    /* @__PURE__ */ jsxs2("div", { className: "mb-6 flex flex-col items-center", children: [
      /* @__PURE__ */ jsx2(Heading, { children: t("login.mfa.title") }),
      /* @__PURE__ */ jsx2(Text, { size: "small", className: "text-ui-fg-subtle text-center", children: isRecoveryCode ? t("login.mfa.recoveryDescription") : t("login.mfa.description") })
    ] }),
    /* @__PURE__ */ jsxs2("div", { className: "flex w-full flex-col gap-y-4", children: [
      isRecoveryCode ? /* @__PURE__ */ jsx2(
        Input,
        {
          autoComplete: "one-time-code",
          className: "bg-ui-bg-field-component",
          placeholder: t("login.mfa.recoveryCodePlaceholder"),
          value: code,
          onChange: (e) => setCode(e.target.value),
          disabled: isLoading || !!verifiedToken
        }
      ) : /* @__PURE__ */ jsx2(
        OtpInput,
        {
          value: code,
          onChange: handleOtpChange,
          onComplete: handleVerify,
          disabled: isLoading || !!verifiedToken,
          autoFocus: true
        }
      ),
      error && /* @__PURE__ */ jsx2("div", { className: "text-center", children: /* @__PURE__ */ jsx2(Hint, { className: "inline-flex", variant: "error", children: error }) }),
      /* @__PURE__ */ jsx2(
        Button2,
        {
          className: "w-full",
          isLoading,
          disabled: !verifiedToken && !canVerify,
          onClick: () => handleVerify(),
          children: t("login.mfa.verify")
        }
      ),
      onBack && /* @__PURE__ */ jsx2(
        Link,
        {
          to: "/login",
          className: "txt-small text-ui-fg-muted transition-fg hover:text-ui-fg-subtle focus-visible:text-ui-fg-subtle text-center outline-none",
          onClick: onBack,
          children: t("login.mfa.backToLogin")
        }
      )
    ] }),
    canUseRecoveryCode && /* @__PURE__ */ jsxs2("div", { className: "text-ui-fg-muted txt-small mt-6 text-center", children: [
      isRecoveryCode ? t("login.mfa.useAuthenticatorPrompt") : t("login.mfa.useRecoveryCodePrompt"),
      " ",
      /* @__PURE__ */ jsx2(
        "button",
        {
          type: "button",
          className: "text-ui-fg-interactive transition-fg hover:text-ui-fg-interactive-hover focus-visible:text-ui-fg-interactive-hover font-medium outline-none",
          onClick: () => handleMethodChange(isRecoveryCode ? "totp" : "recovery_code"),
          children: isRecoveryCode ? t("login.mfa.useAuthenticator") : t("login.mfa.useRecoveryCode")
        }
      )
    ] })
  ] });
};

// src/routes/login/components/mfa-challenge-card.tsx
import { jsx as jsx3, jsxs as jsxs3 } from "react/jsx-runtime";
var MfaChallengeCard = ({
  challenge,
  onSuccess,
  onBack
}) => {
  return /* @__PURE__ */ jsxs3("div", { className: "m-4 flex w-full max-w-[280px] flex-col items-center", children: [
    /* @__PURE__ */ jsx3(AvatarBox, {}),
    /* @__PURE__ */ jsx3(
      MfaChallengeForm,
      {
        challenge,
        onSuccess,
        onBack
      }
    )
  ] });
};

// src/routes/login/login.tsx
import { jsx as jsx4, jsxs as jsxs4 } from "react/jsx-runtime";
var LoginSchema = z.object({
  email: z.string().email(),
  password: z.string()
});
var Login = () => {
  const { t } = useTranslation3();
  const location = useLocation();
  const navigate = useNavigate2();
  const { getWidgets } = useExtension();
  const [mfaChallenge, setMfaChallenge] = useState3(null);
  const [mfaSuccessHandler, setMfaSuccessHandler] = useState3(null);
  const from = location.state?.from?.pathname || "/orders";
  const form = useForm({
    resolver: zodResolver(LoginSchema),
    defaultValues: {
      email: "",
      password: ""
    }
  });
  const { mutateAsync, isPending } = useSignInWithEmailPass();
  const handleSubmit = form.handleSubmit(async ({ email, password }) => {
    await mutateAsync(
      {
        email,
        password
      },
      {
        onError: (error) => {
          if (isFetchError(error)) {
            if (error.status === 401) {
              form.setError("email", {
                type: "manual",
                message: error.message
              });
              return;
            }
          }
          form.setError("root.serverError", {
            type: "manual",
            message: error.message
          });
        },
        onSuccess: (result) => {
          if (typeof result === "object" && "mfa_challenge" in result) {
            setMfaChallenge(result.mfa_challenge);
            setMfaSuccessHandler(() => () => {
              navigate(from, { replace: true });
            });
            return;
          }
          navigate(from, { replace: true });
        }
      }
    );
  });
  const serverError = form.formState.errors?.root?.serverError?.message;
  const validationError = form.formState.errors.email?.message || form.formState.errors.password?.message;
  if (mfaChallenge) {
    return /* @__PURE__ */ jsx4("div", { className: "bg-ui-bg-subtle flex min-h-dvh w-dvw items-center justify-center", children: /* @__PURE__ */ jsx4(
      MfaChallengeCard,
      {
        challenge: mfaChallenge,
        onSuccess: (token) => {
          if (mfaSuccessHandler) {
            return mfaSuccessHandler(token);
          }
          navigate(from, { replace: true });
        },
        onBack: () => {
          setMfaChallenge(null);
          setMfaSuccessHandler(null);
        }
      }
    ) });
  }
  return /* @__PURE__ */ jsx4("div", { className: "bg-ui-bg-subtle flex min-h-dvh w-dvw items-center justify-center", children: /* @__PURE__ */ jsxs4("div", { className: "m-4 flex w-full max-w-[280px] flex-col items-center", children: [
    /* @__PURE__ */ jsx4(AvatarBox, {}),
    /* @__PURE__ */ jsxs4("div", { className: "mb-4 flex flex-col items-center", children: [
      /* @__PURE__ */ jsx4(Heading2, { children: t("login.title") }),
      /* @__PURE__ */ jsx4(Text2, { size: "small", className: "text-ui-fg-subtle text-center", children: t("login.hint") })
    ] }),
    /* @__PURE__ */ jsxs4("div", { className: "flex w-full flex-col gap-y-3", children: [
      getWidgets("login.before").map((Component, i) => {
        return /* @__PURE__ */ jsx4(Component, {}, i);
      }),
      /* @__PURE__ */ jsx4(Form, { ...form, children: /* @__PURE__ */ jsxs4(
        "form",
        {
          onSubmit: handleSubmit,
          className: "flex w-full flex-col gap-y-6",
          children: [
            /* @__PURE__ */ jsxs4("div", { className: "flex flex-col gap-y-1", children: [
              /* @__PURE__ */ jsx4(
                Form.Field,
                {
                  control: form.control,
                  name: "email",
                  render: ({ field }) => {
                    return /* @__PURE__ */ jsx4(Form.Item, { children: /* @__PURE__ */ jsx4(Form.Control, { children: /* @__PURE__ */ jsx4(
                      Input2,
                      {
                        autoComplete: "email",
                        ...field,
                        className: "bg-ui-bg-field-component",
                        placeholder: t("fields.email")
                      }
                    ) }) });
                  }
                }
              ),
              /* @__PURE__ */ jsx4(
                Form.Field,
                {
                  control: form.control,
                  name: "password",
                  render: ({ field }) => {
                    return /* @__PURE__ */ jsxs4(Form.Item, { children: [
                      /* @__PURE__ */ jsx4(Form.Label, {}),
                      /* @__PURE__ */ jsx4(Form.Control, { children: /* @__PURE__ */ jsx4(
                        Input2,
                        {
                          type: "password",
                          autoComplete: "current-password",
                          ...field,
                          className: "bg-ui-bg-field-component",
                          placeholder: t("fields.password")
                        }
                      ) })
                    ] });
                  }
                }
              )
            ] }),
            validationError && /* @__PURE__ */ jsx4("div", { className: "text-center", children: /* @__PURE__ */ jsx4(Hint2, { className: "inline-flex", variant: "error", children: validationError }) }),
            serverError && /* @__PURE__ */ jsx4(
              Alert,
              {
                className: "bg-ui-bg-base items-center p-2",
                dismissible: true,
                variant: "error",
                children: serverError
              }
            ),
            /* @__PURE__ */ jsx4(Button3, { className: "w-full", type: "submit", isLoading: isPending, children: t("actions.continueWithEmail") })
          ]
        }
      ) }),
      getWidgets("login.after").map((Component, i) => {
        return /* @__PURE__ */ jsx4(Component, {}, i);
      }),
      /* @__PURE__ */ jsx4(
        CloudAuthLogin,
        {
          onMfaChallenge: (challenge, onSuccess) => {
            setMfaChallenge(challenge);
            setMfaSuccessHandler(() => onSuccess);
          }
        }
      )
    ] }),
    /* @__PURE__ */ jsx4("span", { className: "text-ui-fg-muted txt-small my-6", children: /* @__PURE__ */ jsx4(
      Trans,
      {
        i18nKey: "login.forgotPassword",
        components: [
          /* @__PURE__ */ jsx4(
            Link2,
            {
              to: "/reset-password",
              className: "text-ui-fg-interactive transition-fg hover:text-ui-fg-interactive-hover focus-visible:text-ui-fg-interactive-hover font-medium outline-none"
            },
            "reset-password-link"
          )
        ]
      }
    ) })
  ] }) });
};
export {
  Login as Component
};
