// @medusajs/dashboard/dist/policy-detail-AYFKSIJ6.mjs
import {
  useRequireRbacFeature
} from "./chunk-Q7IJ6TPC.mjs";
import "./chunk-YOYOJU5D.mjs";
import {
  SectionRow
} from "./chunk-LFLGEXIG.mjs";
import {
  SingleColumnPage
} from "./chunk-HLYIVAII.mjs";
import {
  PermissionGuard
} from "./chunk-TO7QJZLE.mjs";
import {
  DataTable
} from "./chunk-N4L2ZHYI.mjs";
import {
  usePermissions
} from "./chunk-HFX2KPQD.mjs";
import {
  SingleColumnPageSkeleton
} from "./chunk-LPEUYMRK.mjs";
import "./chunk-PJU3RODH.mjs";
import "./chunk-BPIG7PI6.mjs";
import {
  useExtension
} from "./chunk-5SZFF255.mjs";
import "./chunk-OZPB6JBL.mjs";
import "./chunk-OC7BQLYI.mjs";
import "./chunk-S4DMV3ZT.mjs";
import {
  useQueryParams
} from "./chunk-C76H5USB.mjs";
import {
  rbacPoliciesQueryKeys,
  useRbacPolicy,
  useRbacPolicyRoles
} from "./chunk-JHQC6AFI.mjs";
import "./chunk-OCESTHZJ.mjs";
import "./chunk-BEQ5XTWH.mjs";
import {
  queryClient
} from "./chunk-FXYH54JP.mjs";
import "./chunk-774WSTCC.mjs";
import {
  sdk
} from "./chunk-NFEK63OE.mjs";
import "./chunk-QZ7TP4HQ.mjs";

// src/routes/policies/policy-detail/constants.ts
var POLICY_DETAIL_FIELDS = "id,key,resource,operation,name,description,created_at,updated_at,metadata";

// src/routes/policies/policy-detail/breadcrumb.tsx
import { jsx } from "react/jsx-runtime";
var PolicyDetailBreadcrumb = (props) => {
  const { id } = props.params || {};
  const { policy } = useRbacPolicy(
    id,
    {
      fields: POLICY_DETAIL_FIELDS
    },
    {
      initialData: props.data,
      enabled: Boolean(id)
    }
  );
  if (!policy) {
    return null;
  }
  return /* @__PURE__ */ jsx("span", { children: policy.key });
};

// src/routes/policies/policy-detail/loader.ts
var policyDetailQuery = (id) => ({
  queryKey: rbacPoliciesQueryKeys.detail(id, { fields: POLICY_DETAIL_FIELDS }),
  queryFn: async () => sdk.admin.rbacPolicy.retrieve(id, { fields: POLICY_DETAIL_FIELDS })
});
var policyLoader = async ({ params }) => {
  const id = params.id;
  const query = policyDetailQuery(id);
  return queryClient.ensureQueryData(query);
};

// src/routes/policies/policy-detail/policy-detail.tsx
import { useLoaderData, useParams } from "react-router-dom";

// src/routes/policies/policy-detail/components/policy-general-section/policy-general-section.tsx
import {
  Container,
  Heading
} from "@medusajs/ui";
import { useTranslation } from "react-i18next";
import { jsx as jsx2, jsxs } from "react/jsx-runtime";
var PolicyGeneralSection = ({ policy }) => {
  const { t } = useTranslation();
  return /* @__PURE__ */ jsxs(Container, { className: "divide-y p-0", children: [
    /* @__PURE__ */ jsx2("div", { className: "flex items-center justify-between px-6 py-4", children: /* @__PURE__ */ jsx2(Heading, { children: policy.name ?? policy.key }) }),
    /* @__PURE__ */ jsx2(SectionRow, { title: t("fields.key"), value: policy.key }),
    /* @__PURE__ */ jsx2(SectionRow, { title: t("fields.resource"), value: policy.resource }),
    /* @__PURE__ */ jsx2(SectionRow, { title: t("fields.operation"), value: policy.operation }),
    /* @__PURE__ */ jsx2(SectionRow, { title: t("fields.name"), value: policy.name }),
    /* @__PURE__ */ jsx2(SectionRow, { title: t("fields.description"), value: policy.description })
  ] });
};

// src/routes/policies/policy-detail/components/policy-roles-section/policy-roles-section.tsx
import { Container as Container2, createDataTableColumnHelper } from "@medusajs/ui";
import { keepPreviousData } from "@tanstack/react-query";
import { useMemo } from "react";
import { useTranslation as useTranslation2 } from "react-i18next";
import { jsx as jsx3 } from "react/jsx-runtime";
var PAGE_SIZE = 10;
var PREFIX = "pr";
var PolicyRolesSection = ({ policy }) => {
  const { t } = useTranslation2();
  const { offset } = useQueryParams(["offset"], PREFIX);
  const { hasPermission } = usePermissions();
  const canReadRoles = hasPermission("rbac_role:read");
  const { roles, count, isPending, isError, error } = useRbacPolicyRoles(
    policy.id,
    {
      offset: offset ? parseInt(offset) : 0,
      limit: PAGE_SIZE
    },
    {
      placeholderData: keepPreviousData,
      enabled: canReadRoles
    }
  );
  const columns = useColumns();
  if (isError) {
    throw error;
  }
  return /* @__PURE__ */ jsx3(PermissionGuard, { permission: "rbac_role:read", children: /* @__PURE__ */ jsx3(Container2, { className: "divide-y p-0", children: /* @__PURE__ */ jsx3(
    DataTable,
    {
      data: roles ?? [],
      columns,
      getRowId: (row) => row.id,
      rowHref: (row) => `/settings/roles/${row.id}`,
      rowCount: count,
      pageSize: PAGE_SIZE,
      heading: t("policies.fields.roles"),
      headingLevel: "h2",
      isLoading: isPending,
      prefix: PREFIX,
      enableSearch: false,
      emptyState: {
        empty: {
          heading: t("policies.roles.empty.heading"),
          description: t("policies.roles.empty.description")
        },
        filtered: {
          heading: t("policies.list.filtered.heading"),
          description: t("policies.list.filtered.description")
        }
      }
    }
  ) }) });
};
var columnHelper = createDataTableColumnHelper();
var useColumns = () => {
  const { t } = useTranslation2();
  return useMemo(() => {
    return [
      columnHelper.accessor("name", {
        header: t("fields.name")
      }),
      columnHelper.accessor("description", {
        header: t("fields.description"),
        cell: ({ row }) => row.original.description || "-"
      }),
      columnHelper.display({
        id: "user_count",
        header: t("policies.fields.users"),
        cell: ({ row }) => {
          const users = row.original.users ?? [];
          return /* @__PURE__ */ jsx3("span", { children: users.length });
        }
      })
    ];
  }, [t]);
};

// src/routes/policies/policy-detail/policy-detail.tsx
import { jsx as jsx4, jsxs as jsxs2 } from "react/jsx-runtime";
var PolicyDetail = () => {
  const initialData = useLoaderData();
  const { id } = useParams();
  const { getWidgets } = useExtension();
  const isRbacEnabled = useRequireRbacFeature();
  const { hasPermission } = usePermissions();
  const {
    policy,
    isPending: isLoading,
    isError,
    error
  } = useRbacPolicy(
    id,
    { fields: POLICY_DETAIL_FIELDS },
    {
      initialData,
      enabled: !!id && isRbacEnabled
    }
  );
  if (isError) {
    throw error;
  }
  if (isLoading || !policy) {
    return /* @__PURE__ */ jsx4(SingleColumnPageSkeleton, { sections: 2, showJSON: true, showMetadata: true });
  }
  return /* @__PURE__ */ jsxs2(
    SingleColumnPage,
    {
      data: policy,
      showJSON: true,
      showMetadata: true,
      widgets: {
        before: getWidgets("policy.details.before"),
        after: getWidgets("policy.details.after")
      },
      children: [
        /* @__PURE__ */ jsx4(PolicyGeneralSection, { policy }),
        hasPermission("rbac_role:read") && /* @__PURE__ */ jsx4(PolicyRolesSection, { policy })
      ]
    }
  );
};
export {
  PolicyDetailBreadcrumb as Breadcrumb,
  PolicyDetail as Component,
  policyLoader as loader
};
