// @medusajs/dashboard/dist/profile-detail-JNWYH7GB.mjs
import {
  SingleColumnPage
} from "./chunk-HLYIVAII.mjs";
import "./chunk-HFX2KPQD.mjs";
import {
  SingleColumnPageSkeleton
} from "./chunk-LPEUYMRK.mjs";
import "./chunk-PJU3RODH.mjs";
import "./chunk-BPIG7PI6.mjs";
import {
  useExtension
} from "./chunk-5SZFF255.mjs";
import {
  ActionMenu
} from "./chunk-OZPB6JBL.mjs";
import "./chunk-OC7BQLYI.mjs";
import "./chunk-S4DMV3ZT.mjs";
import {
  languages
} from "./chunk-QJ63TWAK.mjs";
import {
  useAuthMfa,
  useDisableAuthMfa,
  useGenerateAuthMfaRecoveryCodes,
  useStartAuthMfa,
  useVerifyAuthMfa
} from "./chunk-BVANDDMY.mjs";
import "./chunk-FAB3VINY.mjs";
import "./chunk-EKTC7HGF.mjs";
import "./chunk-HI6URQ7H.mjs";
import "./chunk-6CLQKVAU.mjs";
import "./chunk-Q7YBE4IP.mjs";
import "./chunk-KI7TOXBR.mjs";
import "./chunk-QBIQEVTU.mjs";
import "./chunk-WAXMT4IY.mjs";
import "./chunk-CN7JXSGW.mjs";
import "./chunk-JHQC6AFI.mjs";
import "./chunk-OCESTHZJ.mjs";
import "./chunk-LGNTHZ5Y.mjs";
import {
  useMe
} from "./chunk-CHDKZVIP.mjs";
import "./chunk-5U6AZEO7.mjs";
import "./chunk-KWWLGCSY.mjs";
import "./chunk-5AFMB7XQ.mjs";
import "./chunk-A63RZVX6.mjs";
import "./chunk-HBXV7ENS.mjs";
import "./chunk-ZBMENGLV.mjs";
import "./chunk-LGLPAUIO.mjs";
import "./chunk-CDORR33H.mjs";
import "./chunk-DE44HCDH.mjs";
import "./chunk-HP2JH45P.mjs";
import "./chunk-VCSSQVPD.mjs";
import "./chunk-SQDIZZDW.mjs";
import "./chunk-MUOBXNL6.mjs";
import "./chunk-BEQ5XTWH.mjs";
import "./chunk-DYDGGABK.mjs";
import "./chunk-LA2ILUXH.mjs";
import "./chunk-7AXHHXCX.mjs";
import "./chunk-FXYH54JP.mjs";
import "./chunk-774WSTCC.mjs";
import "./chunk-NFEK63OE.mjs";
import "./chunk-QZ7TP4HQ.mjs";

// src/routes/profile/profile-detail/components/profile-general-section/profile-general-section.tsx
import { PencilSquare } from "@medusajs/icons";
import { Container, Heading, Text } from "@medusajs/ui";
import { useTranslation } from "react-i18next";
import { jsx, jsxs } from "react/jsx-runtime";
var ProfileGeneralSection = ({ user }) => {
  const { i18n, t } = useTranslation();
  const name = [user.first_name, user.last_name].filter(Boolean).join(" ");
  return /* @__PURE__ */ jsxs(Container, { className: "divide-y p-0", children: [
    /* @__PURE__ */ jsxs("div", { className: "flex items-center justify-between px-6 py-4", children: [
      /* @__PURE__ */ jsxs("div", { children: [
        /* @__PURE__ */ jsx(Heading, { children: t("profile.domain") }),
        /* @__PURE__ */ jsx(Text, { className: "text-ui-fg-subtle", size: "small", children: t("profile.manageYourProfileDetails") })
      ] }),
      /* @__PURE__ */ jsx(
        ActionMenu,
        {
          groups: [
            {
              actions: [
                {
                  label: t("actions.edit"),
                  to: "edit",
                  icon: /* @__PURE__ */ jsx(PencilSquare, {})
                }
              ]
            }
          ]
        }
      )
    ] }),
    /* @__PURE__ */ jsxs("div", { className: "text-ui-fg-subtle grid grid-cols-2 items-center px-6 py-4", children: [
      /* @__PURE__ */ jsx(Text, { size: "small", leading: "compact", weight: "plus", children: t("fields.name") }),
      /* @__PURE__ */ jsx(Text, { size: "small", leading: "compact", children: name || "-" })
    ] }),
    /* @__PURE__ */ jsxs("div", { className: "grid grid-cols-2 items-center px-6 py-4", children: [
      /* @__PURE__ */ jsx(Text, { size: "small", leading: "compact", weight: "plus", children: t("fields.email") }),
      /* @__PURE__ */ jsx(Text, { size: "small", leading: "compact", children: user.email })
    ] }),
    /* @__PURE__ */ jsxs("div", { className: "grid grid-cols-2 items-center px-6 py-4", children: [
      /* @__PURE__ */ jsx(Text, { size: "small", leading: "compact", weight: "plus", children: t("profile.fields.languageLabel") }),
      /* @__PURE__ */ jsx(Text, { size: "small", leading: "compact", children: languages.find((lang) => lang.code === i18n.language)?.display_name || "-" })
    ] })
  ] });
};

// src/routes/profile/profile-detail/components/profile-mfa-section/profile-mfa-section.tsx
import { Key, ShieldCheck } from "@medusajs/icons";
import {
  Badge,
  Button as Button3,
  Container as Container2,
  Heading as Heading4,
  Text as Text5,
  toast,
  usePrompt
} from "@medusajs/ui";
import { useMemo, useState as useState4 } from "react";
import { useTranslation as useTranslation5 } from "react-i18next";

// src/routes/profile/profile-detail/components/profile-mfa-section/mfa-disable-modal.tsx
import { Button, FocusModal, Heading as Heading2, Hint, OtpInput, Text as Text2 } from "@medusajs/ui";
import { useState } from "react";
import { useTranslation as useTranslation2 } from "react-i18next";
import { jsx as jsx2, jsxs as jsxs2 } from "react/jsx-runtime";
var MfaDisableModal = ({
  factor,
  onClose,
  onSuccess
}) => {
  const { t } = useTranslation2();
  const [code, setCode] = useState("");
  const [error, setError] = useState(null);
  const { mutateAsync, isPending } = useDisableAuthMfa(factor.id);
  const handleDisable = async (nextCode = code) => {
    if (nextCode.length !== 6) {
      return;
    }
    setError(null);
    try {
      await mutateAsync({
        method: factor.provider,
        code: nextCode
      });
      onSuccess();
    } catch (e) {
      setError(e instanceof Error ? e.message : t("profile.mfa.disableError"));
      setCode("");
    }
  };
  const handleOtpChange = (value) => {
    setCode(value);
  };
  return /* @__PURE__ */ jsx2(FocusModal, { open: true, onOpenChange: (open) => !open && onClose(), children: /* @__PURE__ */ jsxs2(FocusModal.Content, { children: [
    /* @__PURE__ */ jsx2(FocusModal.Header, { children: /* @__PURE__ */ jsx2(FocusModal.Title, { children: t("profile.mfa.disableTitle") }) }),
    /* @__PURE__ */ jsx2(FocusModal.Body, { className: "flex flex-1 items-center justify-center", children: /* @__PURE__ */ jsxs2("div", { className: "flex w-full max-w-[360px] flex-col items-center gap-y-6 text-center", children: [
      /* @__PURE__ */ jsxs2("div", { children: [
        /* @__PURE__ */ jsx2(Heading2, { children: t("profile.mfa.disableTitle") }),
        /* @__PURE__ */ jsx2(Text2, { size: "small", className: "text-ui-fg-subtle", children: t("profile.mfa.disableChallengeDescription") })
      ] }),
      /* @__PURE__ */ jsx2(
        OtpInput,
        {
          value: code,
          onChange: handleOtpChange,
          onComplete: handleDisable,
          disabled: isPending,
          autoFocus: true
        }
      ),
      error && /* @__PURE__ */ jsx2(Hint, { className: "inline-flex", variant: "error", children: error })
    ] }) }),
    /* @__PURE__ */ jsxs2(FocusModal.Footer, { children: [
      /* @__PURE__ */ jsx2(Button, { variant: "secondary", onClick: onClose, children: t("actions.cancel") }),
      /* @__PURE__ */ jsx2(
        Button,
        {
          variant: "danger",
          isLoading: isPending,
          disabled: code.length !== 6,
          onClick: () => handleDisable(),
          children: t("actions.disable")
        }
      )
    ] })
  ] }) });
};

// src/routes/profile/profile-detail/components/profile-mfa-section/mfa-setup-modal.tsx
import {
  Button as Button2,
  Copy,
  FocusModal as FocusModal2,
  Heading as Heading3,
  Hint as Hint2,
  OtpInput as OtpInput2,
  Text as Text4
} from "@medusajs/ui";
import { useState as useState3 } from "react";
import { useTranslation as useTranslation4 } from "react-i18next";

// src/routes/profile/profile-detail/components/profile-mfa-section/mfa-qr-code.tsx
import { Text as Text3 } from "@medusajs/ui";
import QRCode from "qrcode";
import { useEffect, useRef, useState as useState2 } from "react";
import { useTranslation as useTranslation3 } from "react-i18next";
import { jsx as jsx3 } from "react/jsx-runtime";
var MfaQrCode = ({ value }) => {
  const { t } = useTranslation3();
  const canvasRef = useRef(null);
  const [error, setError] = useState2(false);
  useEffect(() => {
    const canvas = canvasRef.current;
    if (!canvas) {
      return;
    }
    setError(false);
    QRCode.toCanvas(canvas, value, {
      margin: 2,
      width: 220,
      errorCorrectionLevel: "M"
    }).catch(() => {
      setError(true);
    });
  }, [value]);
  if (error) {
    return /* @__PURE__ */ jsx3("div", { className: "border-ui-border-base bg-ui-bg-subtle flex size-[220px] items-center justify-center rounded-md border", children: /* @__PURE__ */ jsx3(Text3, { size: "small", className: "text-ui-fg-subtle text-center", children: t("profile.mfa.qrError") }) });
  }
  return /* @__PURE__ */ jsx3(
    "canvas",
    {
      ref: canvasRef,
      className: "border-ui-border-base bg-ui-bg-base rounded-md border",
      height: 220,
      width: 220
    }
  );
};

// src/routes/profile/profile-detail/components/profile-mfa-section/mfa-setup-modal.tsx
import { Fragment, jsx as jsx4, jsxs as jsxs3 } from "react/jsx-runtime";
var MfaSetupModal = ({ setup, onClose }) => {
  const { t } = useTranslation4();
  const [step, setStep] = useState3("verify");
  const [code, setCode] = useState3("");
  const [error, setError] = useState3(null);
  const [isVerified, setIsVerified] = useState3(false);
  const [recoveryCodes, setRecoveryCodes] = useState3([]);
  const secret = setup.secret;
  const otpauthUrl = setup.otpauth_url;
  const { mutateAsync: verify, isPending: isVerifying } = useVerifyAuthMfa(
    setup.mfa_factor.id
  );
  const { mutateAsync: generateRecoveryCodes, isPending: isGenerating } = useGenerateAuthMfaRecoveryCodes();
  const generateAndShowRecoveryCodes = async () => {
    setError(null);
    try {
      const { recovery_codes } = await generateRecoveryCodes();
      setRecoveryCodes(recovery_codes);
      setStep("recovery-codes");
    } catch (e) {
      setError(
        e instanceof Error ? e.message : t("profile.mfa.recoveryCodesError")
      );
    }
  };
  const handleVerify = async (nextCode = code) => {
    if (isVerified) {
      await generateAndShowRecoveryCodes();
      return;
    }
    if (nextCode.length !== 6) {
      return;
    }
    setError(null);
    try {
      await verify({ code: nextCode });
      setIsVerified(true);
    } catch (e) {
      setError(e instanceof Error ? e.message : t("profile.mfa.verifyError"));
      setCode("");
      return;
    }
    await generateAndShowRecoveryCodes();
  };
  const handleOtpChange = (value) => {
    setCode(value);
  };
  const handleDownloadRecoveryCodes = () => {
    const blob = new Blob([recoveryCodes.join("\n")], {
      type: "text/plain;charset=utf-8"
    });
    const url = URL.createObjectURL(blob);
    const link = document.createElement("a");
    link.href = url;
    link.download = "medusa-recovery-codes.txt";
    link.click();
    URL.revokeObjectURL(url);
  };
  return /* @__PURE__ */ jsx4(FocusModal2, { open: true, onOpenChange: (open) => !open && onClose(), children: /* @__PURE__ */ jsxs3(FocusModal2.Content, { className: "inset-auto left-1/2 top-1/2 max-h-[calc(100vh-32px)] w-[calc(100vw-32px)] max-w-[560px] -translate-x-1/2 -translate-y-1/2", children: [
    /* @__PURE__ */ jsx4(FocusModal2.Header, { children: /* @__PURE__ */ jsx4(FocusModal2.Title, { asChild: true, children: /* @__PURE__ */ jsx4("span", { className: "sr-only", children: step === "verify" ? t("profile.mfa.setupTitle") : t("profile.mfa.recoveryCodesTitle") }) }) }),
    /* @__PURE__ */ jsx4(FocusModal2.Body, { className: "overflow-y-auto p-6", children: step === "verify" ? /* @__PURE__ */ jsxs3("div", { className: "flex w-full flex-col items-center gap-y-6", children: [
      /* @__PURE__ */ jsxs3("div", { className: "flex flex-col items-center gap-y-2 text-center", children: [
        /* @__PURE__ */ jsx4(Heading3, { children: t("profile.mfa.setupAuthenticatorApp") }),
        /* @__PURE__ */ jsx4(Text4, { size: "small", className: "text-ui-fg-subtle", children: t("profile.mfa.setupDescription") })
      ] }),
      secret && otpauthUrl ? /* @__PURE__ */ jsxs3("div", { className: "border-ui-border-base flex w-full flex-col items-center gap-y-4 rounded-lg border p-4", children: [
        /* @__PURE__ */ jsxs3("div", { className: "bg-ui-bg-subtle txt-compact-small text-ui-fg-base border-ui-border-base flex max-w-full items-center gap-x-2 rounded-md border px-3 py-2 font-mono", children: [
          /* @__PURE__ */ jsx4("span", { className: "truncate", children: secret }),
          /* @__PURE__ */ jsx4(Copy, { content: secret, variant: "mini" })
        ] }),
        /* @__PURE__ */ jsx4(MfaQrCode, { value: otpauthUrl })
      ] }) : /* @__PURE__ */ jsx4(Hint2, { className: "inline-flex", variant: "error", children: t("profile.mfa.setupError") }),
      /* @__PURE__ */ jsxs3("div", { className: "flex flex-col items-center gap-y-3", children: [
        /* @__PURE__ */ jsx4(
          OtpInput2,
          {
            value: code,
            onChange: handleOtpChange,
            onComplete: handleVerify,
            disabled: isVerifying || isGenerating || !secret || isVerified,
            autoFocus: true
          }
        ),
        error && /* @__PURE__ */ jsx4(Hint2, { className: "inline-flex", variant: "error", children: error })
      ] })
    ] }) : /* @__PURE__ */ jsxs3("div", { className: "flex w-full flex-col gap-y-4", children: [
      /* @__PURE__ */ jsxs3("div", { children: [
        /* @__PURE__ */ jsx4(Heading3, { children: t("profile.mfa.recoveryCodesTitle") }),
        /* @__PURE__ */ jsx4(Text4, { size: "small", className: "text-ui-fg-subtle", children: t("profile.mfa.recoveryCodesDescription") })
      ] }),
      /* @__PURE__ */ jsx4("div", { className: "border-ui-border-base bg-ui-bg-subtle grid grid-cols-2 gap-3 rounded-lg border p-4", children: recoveryCodes.map((recoveryCode) => {
        return /* @__PURE__ */ jsx4(Text4, { size: "small", className: "font-mono", children: recoveryCode }, recoveryCode);
      }) })
    ] }) }),
    /* @__PURE__ */ jsx4(FocusModal2.Footer, { children: step === "verify" ? /* @__PURE__ */ jsxs3(Fragment, { children: [
      /* @__PURE__ */ jsx4(Button2, { variant: "secondary", onClick: onClose, children: t("actions.cancel") }),
      /* @__PURE__ */ jsx4(
        Button2,
        {
          isLoading: isVerifying || isGenerating,
          disabled: !isVerified && code.length !== 6 || !secret,
          onClick: () => handleVerify(),
          children: t("actions.confirm")
        }
      )
    ] }) : /* @__PURE__ */ jsxs3(Fragment, { children: [
      /* @__PURE__ */ jsx4(Button2, { variant: "secondary", onClick: handleDownloadRecoveryCodes, children: t("actions.download") }),
      /* @__PURE__ */ jsx4(Copy, { content: recoveryCodes.join("\n"), asChild: true, children: /* @__PURE__ */ jsx4(Button2, { variant: "secondary", children: t("actions.copy") }) }),
      /* @__PURE__ */ jsx4(Button2, { onClick: onClose, children: t("actions.complete") })
    ] }) })
  ] }) });
};

// src/routes/profile/profile-detail/components/profile-mfa-section/profile-mfa-section.tsx
import { Fragment as Fragment2, jsx as jsx5, jsxs as jsxs4 } from "react/jsx-runtime";
var MFA_DISABLE_CODE_REQUIRED_ERROR = "MFA verification code is required to disable MFA";
var ProfileMfaSection = () => {
  const { t } = useTranslation5();
  const prompt = usePrompt();
  const { mfa_factors: factors = [], isPending } = useAuthMfa();
  const [setupResponse, setSetupResponse] = useState4(null);
  const [disableChallengeFactor, setDisableChallengeFactor] = useState4(null);
  const { mutateAsync: startMfa, isPending: isStarting } = useStartAuthMfa();
  const enabledFactor = useMemo(() => {
    return factors.find((factor) => factor.status === "enabled");
  }, [factors]);
  const pendingFactor = useMemo(() => {
    return factors.find((factor) => factor.status === "pending");
  }, [factors]);
  const { mutateAsync: disableMfa, isPending: isDisabling } = useDisableAuthMfa(
    enabledFactor?.id ?? ""
  );
  const { mutateAsync: cancelPendingMfa, isPending: isCancellingPending } = useDisableAuthMfa(pendingFactor?.id ?? "");
  const handleSetup = async () => {
    try {
      if (pendingFactor) {
        await cancelPendingMfa();
      }
      const response = await startMfa({
        provider: "totp",
        label: t("profile.mfa.authenticatorApp")
      });
      setSetupResponse(response);
    } catch (e) {
      toast.error(e instanceof Error ? e.message : t("profile.mfa.setupError"));
    }
  };
  const handleDisable = async () => {
    if (!enabledFactor) {
      return;
    }
    const confirmed = await prompt({
      title: t("profile.mfa.disableTitle"),
      description: t("profile.mfa.disableDescription"),
      confirmText: t("actions.disable"),
      cancelText: t("actions.cancel"),
      variant: "danger"
    });
    if (!confirmed) {
      return;
    }
    try {
      await disableMfa();
      toast.success(t("profile.mfa.disableSuccess"));
    } catch (e) {
      if (e instanceof Error && e.message.includes(MFA_DISABLE_CODE_REQUIRED_ERROR)) {
        setDisableChallengeFactor(enabledFactor);
        return;
      }
      toast.error(
        e instanceof Error ? e.message : t("profile.mfa.disableError")
      );
    }
  };
  return /* @__PURE__ */ jsxs4(Fragment2, { children: [
    /* @__PURE__ */ jsxs4(Container2, { className: "divide-y p-0", children: [
      /* @__PURE__ */ jsxs4("div", { className: "flex items-center justify-between px-6 py-4", children: [
        /* @__PURE__ */ jsxs4("div", { children: [
          /* @__PURE__ */ jsx5(Heading4, { children: t("profile.mfa.title") }),
          /* @__PURE__ */ jsx5(Text5, { className: "text-ui-fg-subtle", size: "small", children: t("profile.mfa.description") })
        ] }),
        enabledFactor ? /* @__PURE__ */ jsx5(
          Button3,
          {
            size: "small",
            variant: "danger",
            isLoading: isDisabling,
            onClick: handleDisable,
            children: t("actions.disable")
          }
        ) : /* @__PURE__ */ jsx5(
          Button3,
          {
            size: "small",
            variant: "secondary",
            isLoading: isStarting || isCancellingPending,
            disabled: isPending,
            onClick: handleSetup,
            children: t("actions.enable")
          }
        )
      ] }),
      /* @__PURE__ */ jsxs4("div", { className: "grid grid-cols-2 items-center px-6 py-4", children: [
        /* @__PURE__ */ jsx5(Text5, { size: "small", leading: "compact", weight: "plus", children: t("profile.mfa.status") }),
        /* @__PURE__ */ jsx5("div", { children: enabledFactor ? /* @__PURE__ */ jsx5(Badge, { color: "green", size: "2xsmall", children: t("profile.mfa.enabled") }) : pendingFactor ? /* @__PURE__ */ jsx5(Badge, { color: "orange", size: "2xsmall", children: t("profile.mfa.pending") }) : /* @__PURE__ */ jsx5(Badge, { color: "grey", size: "2xsmall", children: t("profile.mfa.disabled") }) })
      ] }),
      /* @__PURE__ */ jsxs4("div", { className: "grid grid-cols-2 items-center px-6 py-4", children: [
        /* @__PURE__ */ jsx5(Text5, { size: "small", leading: "compact", weight: "plus", children: t("profile.mfa.method") }),
        /* @__PURE__ */ jsx5("div", { className: "flex items-center gap-x-2", children: enabledFactor || pendingFactor ? /* @__PURE__ */ jsxs4(Fragment2, { children: [
          /* @__PURE__ */ jsx5(ShieldCheck, { className: "text-ui-fg-subtle" }),
          /* @__PURE__ */ jsx5(Text5, { size: "small", leading: "compact", children: t("profile.mfa.authenticatorApp") })
        ] }) : /* @__PURE__ */ jsxs4(Fragment2, { children: [
          /* @__PURE__ */ jsx5(Key, { className: "text-ui-fg-muted" }),
          /* @__PURE__ */ jsx5(
            Text5,
            {
              size: "small",
              leading: "compact",
              className: "text-ui-fg-subtle",
              children: t("profile.mfa.noMethod")
            }
          )
        ] }) })
      ] })
    ] }),
    setupResponse && /* @__PURE__ */ jsx5(
      MfaSetupModal,
      {
        setup: setupResponse,
        onClose: () => setSetupResponse(null)
      }
    ),
    disableChallengeFactor && /* @__PURE__ */ jsx5(
      MfaDisableModal,
      {
        factor: disableChallengeFactor,
        onClose: () => setDisableChallengeFactor(null),
        onSuccess: () => {
          setDisableChallengeFactor(null);
          toast.success(t("profile.mfa.disableSuccess"));
        }
      }
    )
  ] });
};

// src/routes/profile/profile-detail/profile-detail.tsx
import { jsx as jsx6, jsxs as jsxs5 } from "react/jsx-runtime";
var ProfileDetail = () => {
  const { user, isPending: isLoading, isError, error } = useMe();
  const { getWidgets } = useExtension();
  if (isLoading || !user) {
    return /* @__PURE__ */ jsx6(SingleColumnPageSkeleton, { sections: 2 });
  }
  if (isError) {
    throw error;
  }
  return /* @__PURE__ */ jsxs5(
    SingleColumnPage,
    {
      widgets: {
        after: getWidgets("profile.details.after"),
        before: getWidgets("profile.details.before")
      },
      children: [
        /* @__PURE__ */ jsx6(ProfileGeneralSection, { user }),
        /* @__PURE__ */ jsx6(ProfileMfaSection, {})
      ]
    }
  );
};
export {
  ProfileDetail as Component
};
