// @medusajs/dashboard/dist/role-permissions-ZHZVNCDE.mjs
import "./chunk-IUCDCPJU.mjs";
import {
  KeyboundForm
} from "./chunk-6HTZNHPT.mjs";
import {
  DataTable
} from "./chunk-N4L2ZHYI.mjs";
import "./chunk-PJU3RODH.mjs";
import "./chunk-OZPB6JBL.mjs";
import "./chunk-OC7BQLYI.mjs";
import "./chunk-S4DMV3ZT.mjs";
import {
  useQueryParams
} from "./chunk-C76H5USB.mjs";
import {
  RouteDrawer
} from "./chunk-OV6YUU3D.mjs";
import {
  useRouteModal
} from "./chunk-ULMMKZ3U.mjs";
import "./chunk-OBQI23QM.mjs";
import {
  useRbacAssignablePolicies
} from "./chunk-JHQC6AFI.mjs";
import {
  rbacRolesQueryKeys,
  useAddRbacRolePolicies,
  useRbacRole
} from "./chunk-OCESTHZJ.mjs";
import "./chunk-BEQ5XTWH.mjs";
import {
  queryClient
} from "./chunk-FXYH54JP.mjs";
import "./chunk-774WSTCC.mjs";
import {
  sdk
} from "./chunk-NFEK63OE.mjs";
import "./chunk-QZ7TP4HQ.mjs";

// src/routes/roles/role-permissions/role-permissions.tsx
import { Heading } from "@medusajs/ui";
import { useTranslation as useTranslation2 } from "react-i18next";
import { useParams } from "react-router-dom";

// src/routes/roles/role-permissions/components/edit-role-permissions-form/edit-role-permissions-form.tsx
import { zodResolver } from "@hookform/resolvers/zod";
import {
  Button,
  createDataTableColumnHelper,
  toast
} from "@medusajs/ui";
import { keepPreviousData, useMutation } from "@tanstack/react-query";
import { useEffect, useMemo, useState } from "react";
import { useForm } from "react-hook-form";
import { useTranslation } from "react-i18next";
import * as zod from "zod";
import { jsx, jsxs } from "react/jsx-runtime";
var EditRolePermissionsSchema = zod.object({
  policies: zod.array(zod.string()).optional()
});
var PAGE_SIZE = 20;
var PREFIX = "rp";
var EditRolePermissionsForm = ({
  role
}) => {
  const { t } = useTranslation();
  const { handleSuccess } = useRouteModal();
  const form = useForm({
    defaultValues: {
      policies: role.policies?.map((policy) => policy.id) ?? []
    },
    resolver: zodResolver(EditRolePermissionsSchema)
  });
  const initialState = role.policies?.reduce((acc, policy) => {
    acc[policy.id] = true;
    return acc;
  }, {}) ?? {};
  const [rowSelection, setRowSelection] = useState(initialState);
  useEffect(() => {
    const ids = Object.keys(rowSelection).filter((id) => rowSelection[id]);
    form.setValue("policies", ids, {
      shouldDirty: true,
      shouldTouch: true
    });
  }, [rowSelection, form]);
  const { q, order, offset } = useQueryParams(["q", "order", "offset"], PREFIX);
  const {
    data: pageData,
    isPending: isLoading,
    isError,
    error
  } = useRbacAssignablePolicies(
    {
      q,
      order,
      offset: offset ? parseInt(offset) : 0,
      limit: PAGE_SIZE,
      fields: "id,key,resource,operation,description"
    },
    {
      placeholderData: keepPreviousData
    }
  );
  const visiblePolicies = pageData?.policies ?? [];
  const count = pageData?.count ?? 0;
  const { data: allAssignable, isPending: isAssignableSetLoading } = useRbacAssignablePolicies();
  const assignableIds = useMemo(
    () => new Set((allAssignable?.policies ?? []).map((p) => p.id)),
    [allAssignable?.policies]
  );
  const columns = usePolicyColumns();
  const { mutateAsync: addPolicies, isPending: isAdding } = useAddRbacRolePolicies(role.id);
  const { mutateAsync: removePolicies, isPending: isRemoving } = useMutation({
    mutationFn: async (policyIds) => {
      await Promise.all(
        policyIds.map(
          (policyId) => sdk.admin.rbacRole.removePolicy(role.id, policyId)
        )
      );
    },
    onSuccess: () => {
      queryClient.invalidateQueries({
        queryKey: rbacRolesQueryKeys.policies(role.id)
      });
      queryClient.invalidateQueries({
        queryKey: rbacRolesQueryKeys.detail(role.id)
      });
      queryClient.invalidateQueries({ queryKey: rbacRolesQueryKeys.lists() });
    }
  });
  const handleSubmit = form.handleSubmit(async (data) => {
    const selectedPolicies = data.policies ?? [];
    const existingPolicies = role.policies?.map((policy) => policy.id) ?? [];
    const toAdd = selectedPolicies.filter(
      (policyId) => !existingPolicies.includes(policyId)
    );
    const toRemove = existingPolicies.filter((policyId) => assignableIds.has(policyId)).filter((policyId) => !selectedPolicies.includes(policyId));
    try {
      if (toAdd.length) {
        await addPolicies({ policies: toAdd });
      }
      if (toRemove.length) {
        await removePolicies(toRemove);
      }
      handleSuccess();
    } catch (e) {
      const errorMessage = e instanceof Error ? e.message : "Unknown error";
      toast.error(errorMessage);
    }
  });
  if (isError) {
    throw error;
  }
  return /* @__PURE__ */ jsx(RouteDrawer.Form, { form, children: /* @__PURE__ */ jsxs(
    KeyboundForm,
    {
      className: "flex flex-1 flex-col overflow-hidden",
      onSubmit: handleSubmit,
      children: [
        /* @__PURE__ */ jsx(RouteDrawer.Body, { className: "-mx-4 flex flex-1 flex-col overflow-hidden p-0", children: /* @__PURE__ */ jsx(
          DataTable,
          {
            data: visiblePolicies,
            columns,
            getRowId: (row) => row.id,
            rowCount: count,
            isLoading,
            pageSize: PAGE_SIZE,
            rowSelection: {
              state: rowSelection,
              onRowSelectionChange: setRowSelection
            },
            autoFocusSearch: true,
            layout: "fill",
            emptyState: {
              empty: {
                heading: t("roles.permissions.empty.heading"),
                description: t("roles.permissions.empty.description")
              },
              filtered: {
                heading: t("roles.permissions.filtered.heading"),
                description: t("roles.permissions.filtered.description")
              }
            },
            prefix: PREFIX
          }
        ) }),
        /* @__PURE__ */ jsx(RouteDrawer.Footer, { className: "shrink-0", children: /* @__PURE__ */ jsxs("div", { className: "flex items-center justify-end gap-x-2", children: [
          /* @__PURE__ */ jsx(RouteDrawer.Close, { asChild: true, children: /* @__PURE__ */ jsx(Button, { size: "small", variant: "secondary", children: t("actions.cancel") }) }),
          /* @__PURE__ */ jsx(
            Button,
            {
              size: "small",
              type: "submit",
              isLoading: isAdding || isRemoving,
              disabled: isAssignableSetLoading,
              children: t("actions.save")
            }
          )
        ] }) })
      ]
    }
  ) });
};
var columnHelper = createDataTableColumnHelper();
var usePolicyColumns = () => {
  const { t } = useTranslation();
  return useMemo(() => {
    return [
      columnHelper.select(),
      columnHelper.accessor("key", {
        header: t("fields.key")
      }),
      columnHelper.accessor("resource", {
        header: t("fields.resource"),
        cell: ({ row }) => {
          const resource = row.original.resource;
          return t(`permissions.resources.${resource}`, {
            defaultValue: resource
          });
        }
      }),
      columnHelper.accessor("operation", {
        header: t("fields.operation"),
        cell: ({ row }) => {
          const operation = row.original.operation;
          return t(`permissions.actions.${operation}`, {
            defaultValue: operation
          });
        }
      }),
      columnHelper.accessor("description", {
        header: t("fields.description"),
        cell: ({ row }) => {
          return row.original.description || "-";
        }
      })
    ];
  }, [t]);
};

// src/routes/roles/role-permissions/role-permissions.tsx
import { Fragment, jsx as jsx2, jsxs as jsxs2 } from "react/jsx-runtime";
var RolePermissions = () => {
  const { id } = useParams();
  const { t } = useTranslation2();
  const { role, isPending, isError, error } = useRbacRole(id, {
    fields: "id,name,policies.id"
  });
  if (isError) {
    throw error;
  }
  return /* @__PURE__ */ jsx2(RouteDrawer, { children: !isPending && role && /* @__PURE__ */ jsxs2(Fragment, { children: [
    /* @__PURE__ */ jsxs2(RouteDrawer.Header, { children: [
      /* @__PURE__ */ jsx2(RouteDrawer.Title, { asChild: true, children: /* @__PURE__ */ jsx2(Heading, { children: t("roles.permissions.header", { name: role.name }) }) }),
      /* @__PURE__ */ jsx2(RouteDrawer.Description, { className: "sr-only", children: t("roles.permissions.hint") })
    ] }),
    /* @__PURE__ */ jsx2(EditRolePermissionsForm, { role })
  ] }) });
};
export {
  RolePermissions as Component
};
