// @medusajs/dashboard/dist/role-detail-QZS3AQLS.mjs
import {
  useRequireRbacFeature
} from "./chunk-Q7IJ6TPC.mjs";
import {
  ListSummary
} from "./chunk-I3VB6NM2.mjs";
import "./chunk-YOYOJU5D.mjs";
import {
  SectionRow
} from "./chunk-LFLGEXIG.mjs";
import "./chunk-LQTHYS2Z.mjs";
import {
  _DataTable,
  useDataTable
} from "./chunk-CR5DSAVZ.mjs";
import "./chunk-HQKGZADC.mjs";
import "./chunk-EMIHDNB7.mjs";
import {
  SingleColumnPage
} from "./chunk-HLYIVAII.mjs";
import {
  PermissionGuard
} from "./chunk-TO7QJZLE.mjs";
import {
  usePermissions
} from "./chunk-HFX2KPQD.mjs";
import {
  SingleColumnPageSkeleton
} from "./chunk-LPEUYMRK.mjs";
import "./chunk-PJU3RODH.mjs";
import "./chunk-BPIG7PI6.mjs";
import {
  useExtension
} from "./chunk-5SZFF255.mjs";
import {
  ActionMenu
} from "./chunk-OZPB6JBL.mjs";
import "./chunk-OC7BQLYI.mjs";
import "./chunk-S4DMV3ZT.mjs";
import "./chunk-XRM7PIRS.mjs";
import {
  useQueryParams
} from "./chunk-C76H5USB.mjs";
import {
  useDate
} from "./chunk-3BQAAPDR.mjs";
import "./chunk-QJ63TWAK.mjs";
import {
  rbacRolesQueryKeys,
  useDeleteRbacRole,
  useRbacAssignableRoles,
  useRbacRole,
  useRbacRoleUsers,
  useRemoveRbacRoleUsers
} from "./chunk-OCESTHZJ.mjs";
import {
  queryClient
} from "./chunk-FXYH54JP.mjs";
import "./chunk-774WSTCC.mjs";
import {
  sdk
} from "./chunk-NFEK63OE.mjs";
import "./chunk-QZ7TP4HQ.mjs";

// src/routes/roles/role-detail/constants.ts
var ROLE_DETAIL_FIELDS = "id,name,description,created_at,updated_at,metadata,policies.id,policies.key,policies.resource,policies.operation,policies.description,users_link.user.id,users_link.user.first_name,users_link.user.last_name,users_link.user.email";

// src/routes/roles/role-detail/breadcrumb.tsx
import { jsx } from "react/jsx-runtime";
var RoleDetailBreadcrumb = (props) => {
  const { id } = props.params || {};
  const { role } = useRbacRole(
    id,
    {
      fields: ROLE_DETAIL_FIELDS
    },
    {
      initialData: props.data,
      enabled: Boolean(id)
    }
  );
  if (!role) {
    return null;
  }
  return /* @__PURE__ */ jsx("span", { children: role.name });
};

// src/routes/roles/role-detail/loader.ts
var roleDetailQuery = (id) => ({
  queryKey: rbacRolesQueryKeys.detail(id, { fields: ROLE_DETAIL_FIELDS }),
  queryFn: async () => sdk.admin.rbacRole.retrieve(id, { fields: ROLE_DETAIL_FIELDS })
});
var roleLoader = async ({ params }) => {
  const id = params.id;
  const query = roleDetailQuery(id);
  return queryClient.ensureQueryData(query);
};

// src/routes/roles/role-detail/role-detail.tsx
import { useLoaderData, useParams } from "react-router-dom";

// src/routes/roles/role-detail/components/role-general-section/role-general-section.tsx
import { Key, PencilSquare, Trash } from "@medusajs/icons";
import { Container, Heading, toast, usePrompt } from "@medusajs/ui";
import { useMemo } from "react";
import { useTranslation } from "react-i18next";
import { useNavigate } from "react-router-dom";
import { jsx as jsx2, jsxs } from "react/jsx-runtime";
var RoleGeneralSection = ({ role }) => {
  const { t } = useTranslation();
  const { hasPermission } = usePermissions();
  const prompt = usePrompt();
  const navigate = useNavigate();
  const { mutateAsync: deleteRole, isPending: isDeleting } = useDeleteRbacRole(
    role.id
  );
  const handleDelete = async () => {
    const confirmed = await prompt({
      title: t("roles.delete.title"),
      description: t("roles.delete.description", { name: role.name }),
      confirmText: t("actions.delete"),
      cancelText: t("actions.cancel")
    });
    if (!confirmed) {
      return;
    }
    try {
      await deleteRole();
      toast.success(t("roles.delete.successToast", { name: role.name }));
      navigate("/settings/roles", { replace: true });
    } catch (error) {
      toast.error(error.message);
    }
  };
  const users = useMemo(() => {
    return role.users_link?.map((link) => link.user).filter((user) => !!user) ?? [];
  }, [role.users_link]);
  const userLabels = useMemo(() => {
    return users.map((user) => {
      const fullName = `${user.first_name || ""} ${user.last_name || ""}`.trim();
      return fullName || user.email || user.id;
    });
  }, [users]);
  const permissionLabels = useMemo(() => {
    return role.policies?.map((policy) => {
      if (policy.key) {
        return policy.key;
      }
      const resource = policy.resource ?? "";
      const operation = policy.operation ?? "";
      if (resource && operation) {
        return `${resource}:${operation}`;
      }
      return policy.id;
    }) ?? [];
  }, [role.policies]);
  const canUpdate = hasPermission("rbac_role:update");
  const canDelete = hasPermission("rbac_role:delete");
  const groups = [];
  if (canUpdate) {
    groups.push({
      actions: [
        {
          icon: /* @__PURE__ */ jsx2(PencilSquare, {}),
          label: t("actions.edit"),
          to: `edit`
        },
        {
          icon: /* @__PURE__ */ jsx2(Key, {}),
          label: t("roles.actions.managePermissions"),
          to: `permissions`
        }
      ]
    });
  }
  if (canDelete) {
    groups.push({
      actions: [
        {
          icon: /* @__PURE__ */ jsx2(Trash, {}),
          label: t("actions.delete"),
          onClick: handleDelete,
          disabled: isDeleting
        }
      ]
    });
  }
  return /* @__PURE__ */ jsxs(Container, { className: "divide-y p-0", children: [
    /* @__PURE__ */ jsxs("div", { className: "flex items-center justify-between px-6 py-4", children: [
      /* @__PURE__ */ jsx2(Heading, { children: role.name }),
      groups.length && /* @__PURE__ */ jsx2(ActionMenu, { groups })
    ] }),
    /* @__PURE__ */ jsx2(SectionRow, { title: t("fields.description"), value: role.description }),
    hasPermission("user:read") && /* @__PURE__ */ jsx2(
      SectionRow,
      {
        title: t("users.domain"),
        value: userLabels.length ? /* @__PURE__ */ jsx2("div", { className: "inline-flex", children: /* @__PURE__ */ jsx2(ListSummary, { list: userLabels }) }) : "-"
      }
    ),
    hasPermission("rbac_policy:read") && /* @__PURE__ */ jsx2(
      SectionRow,
      {
        title: t("roles.fields.permissions"),
        value: permissionLabels.length ? /* @__PURE__ */ jsx2("div", { className: "inline-flex", children: /* @__PURE__ */ jsx2(ListSummary, { list: permissionLabels }) }) : "-"
      }
    )
  ] });
};

// src/routes/roles/role-detail/components/role-users-section/role-users-section.tsx
import { Trash as Trash2 } from "@medusajs/icons";
import { Button, Checkbox, Container as Container2, Heading as Heading2, usePrompt as usePrompt2 } from "@medusajs/ui";
import { createColumnHelper } from "@tanstack/react-table";
import { useEffect, useMemo as useMemo2, useState } from "react";
import { useTranslation as useTranslation2 } from "react-i18next";
import { Link } from "react-router-dom";
import { jsx as jsx3, jsxs as jsxs2 } from "react/jsx-runtime";
var PAGE_SIZE = 10;
var RoleUsersSection = ({ role }) => {
  const [rowSelection, setRowSelection] = useState({});
  const { t } = useTranslation2();
  const prompt = usePrompt2();
  const { offset, order } = useQueryParams(["offset", "order"]);
  const { hasAllPermissions } = usePermissions();
  const canManageRole = hasAllPermissions(["user:update", "rbac_role:update"]);
  const { data: assignableData } = useRbacAssignableRoles(void 0, {
    enabled: canManageRole
  });
  const isRoleAssignable = (assignableData?.roles ?? []).some(
    (r) => r.id === role.id
  );
  useEffect(() => {
    if (!isRoleAssignable && Object.keys(rowSelection).length) {
      setRowSelection({});
    }
  }, [isRoleAssignable, rowSelection, setRowSelection]);
  const {
    users,
    count,
    isPending: isLoading,
    isError,
    error
  } = useRbacRoleUsers(role.id, {
    limit: PAGE_SIZE,
    offset: offset ? parseInt(offset) : 0,
    order
  });
  const columns = useColumns(isRoleAssignable);
  const { table } = useDataTable({
    data: users ?? [],
    columns,
    count,
    getRowId: (row) => row.id,
    enablePagination: true,
    enableRowSelection: isRoleAssignable,
    pageSize: PAGE_SIZE,
    rowSelection: {
      state: rowSelection,
      updater: setRowSelection
    },
    meta: {
      roleId: role.id,
      isRoleAssignable
    }
  });
  if (isError) {
    throw error;
  }
  const { mutateAsync } = useRemoveRbacRoleUsers(role.id);
  const handleRemove = async () => {
    if (!isRoleAssignable) {
      return;
    }
    const keys = Object.keys(rowSelection);
    if (!keys.length) {
      return;
    }
    const res = await prompt({
      title: t("roles.users.remove.title", {
        count: keys.length
      }),
      description: t("roles.users.remove.description", {
        count: keys.length
      }),
      confirmText: t("actions.continue"),
      cancelText: t("actions.cancel")
    });
    if (!res) {
      return;
    }
    await mutateAsync(keys, {
      onSuccess: () => {
        setRowSelection({});
      }
    });
  };
  return /* @__PURE__ */ jsxs2(Container2, { className: "divide-y p-0", children: [
    /* @__PURE__ */ jsxs2("div", { className: "flex items-center justify-between px-6 py-4", children: [
      /* @__PURE__ */ jsx3(Heading2, { level: "h2", children: t("users.domain") }),
      isRoleAssignable && /* @__PURE__ */ jsx3(Link, { to: "add-users", children: /* @__PURE__ */ jsx3(Button, { variant: "secondary", size: "small", children: t("general.add") }) })
    ] }),
    /* @__PURE__ */ jsx3(
      _DataTable,
      {
        table,
        columns,
        pageSize: PAGE_SIZE,
        isLoading,
        count,
        navigateTo: (row) => `/settings/users/${row.original.id}`,
        orderBy: [
          { key: "email", label: t("fields.email") },
          { key: "first_name", label: t("fields.firstName") },
          { key: "last_name", label: t("fields.lastName") },
          { key: "created_at", label: t("fields.createdAt") },
          { key: "updated_at", label: t("fields.updatedAt") }
        ],
        commands: isRoleAssignable ? [
          {
            action: handleRemove,
            label: t("actions.remove"),
            shortcut: "r"
          }
        ] : void 0,
        noRecords: {
          message: t("roles.users.list.noRecordsMessage")
        },
        pagination: true
      }
    )
  ] });
};
var RoleUserActions = ({
  user,
  roleId,
  isRoleAssignable
}) => {
  const { t } = useTranslation2();
  const prompt = usePrompt2();
  const { mutateAsync } = useRemoveRbacRoleUsers(roleId);
  if (!isRoleAssignable) {
    return null;
  }
  const handleRemove = async () => {
    const res = await prompt({
      title: t("roles.users.remove.title", {
        count: 1
      }),
      description: t("roles.users.remove.description", {
        count: 1
      }),
      confirmText: t("actions.continue"),
      cancelText: t("actions.cancel")
    });
    if (!res) {
      return;
    }
    await mutateAsync([user.id]);
  };
  return /* @__PURE__ */ jsx3(
    ActionMenu,
    {
      groups: [
        {
          actions: [
            {
              icon: /* @__PURE__ */ jsx3(Trash2, {}),
              label: t("actions.remove"),
              onClick: handleRemove
            }
          ]
        }
      ]
    }
  );
};
var columnHelper = createColumnHelper();
var useColumns = (isRoleAssignable) => {
  const { t } = useTranslation2();
  const { getFullDate } = useDate();
  return useMemo2(
    () => [
      columnHelper.display({
        id: "select",
        header: ({ table }) => {
          return /* @__PURE__ */ jsx3(
            Checkbox,
            {
              checked: table.getIsSomePageRowsSelected() ? "indeterminate" : table.getIsAllPageRowsSelected(),
              disabled: !isRoleAssignable,
              onCheckedChange: (value) => table.toggleAllPageRowsSelected(!!value)
            }
          );
        },
        cell: ({ row }) => {
          return /* @__PURE__ */ jsx3(
            Checkbox,
            {
              checked: row.getIsSelected(),
              disabled: !row.getCanSelect(),
              onCheckedChange: (value) => row.toggleSelected(!!value),
              onClick: (e) => {
                e.stopPropagation();
              }
            }
          );
        }
      }),
      columnHelper.accessor("first_name", {
        header: t("fields.firstName"),
        cell: ({ row }) => row.original.first_name || "-"
      }),
      columnHelper.accessor("last_name", {
        header: t("fields.lastName"),
        cell: ({ row }) => row.original.last_name || "-"
      }),
      columnHelper.accessor("email", {
        header: t("fields.email"),
        cell: ({ row }) => row.original.email
      }),
      columnHelper.accessor("created_at", {
        header: t("fields.createdAt"),
        cell: ({ row }) => getFullDate({ date: row.original.created_at })
      }),
      columnHelper.display({
        id: "actions",
        cell: ({ row, table }) => {
          const { roleId, isRoleAssignable: rowisRoleAssignable } = table.options.meta;
          return /* @__PURE__ */ jsx3(
            RoleUserActions,
            {
              user: row.original,
              roleId,
              isRoleAssignable: rowisRoleAssignable
            }
          );
        }
      })
    ],
    [t, getFullDate, isRoleAssignable]
  );
};

// src/routes/roles/role-detail/role-detail.tsx
import { jsx as jsx4, jsxs as jsxs3 } from "react/jsx-runtime";
var RoleDetail = () => {
  const initialData = useLoaderData();
  const { id } = useParams();
  const { getWidgets } = useExtension();
  const isRbacEnabled = useRequireRbacFeature();
  const {
    role,
    isPending: isLoading,
    isError,
    error
  } = useRbacRole(
    id,
    { fields: ROLE_DETAIL_FIELDS },
    {
      initialData,
      enabled: !!id && isRbacEnabled
    }
  );
  if (!isRbacEnabled) {
    return null;
  }
  if (isError) {
    throw error;
  }
  if (isLoading || !role) {
    return /* @__PURE__ */ jsx4(SingleColumnPageSkeleton, { sections: 2, showJSON: true, showMetadata: true });
  }
  return /* @__PURE__ */ jsxs3(
    SingleColumnPage,
    {
      data: role,
      showJSON: true,
      showMetadata: true,
      widgets: {
        before: getWidgets("role.details.before"),
        after: getWidgets("role.details.after")
      },
      children: [
        /* @__PURE__ */ jsx4(RoleGeneralSection, { role }),
        /* @__PURE__ */ jsx4(PermissionGuard, { permission: "user:read", children: /* @__PURE__ */ jsx4(RoleUsersSection, { role }) })
      ]
    }
  );
};
export {
  RoleDetailBreadcrumb as Breadcrumb,
  RoleDetail as Component,
  roleLoader as loader
};
