// @medusajs/dashboard/dist/role-add-users-TQBAKL6I.mjs
import "./chunk-LQTHYS2Z.mjs";
import {
  _DataTable,
  useDataTable
} from "./chunk-CR5DSAVZ.mjs";
import "./chunk-HQKGZADC.mjs";
import "./chunk-EMIHDNB7.mjs";
import "./chunk-IUCDCPJU.mjs";
import {
  KeyboundForm
} from "./chunk-6HTZNHPT.mjs";
import {
  usePermissions
} from "./chunk-HFX2KPQD.mjs";
import "./chunk-LPEUYMRK.mjs";
import "./chunk-S4DMV3ZT.mjs";
import "./chunk-XRM7PIRS.mjs";
import {
  useQueryParams
} from "./chunk-C76H5USB.mjs";
import {
  useDate
} from "./chunk-3BQAAPDR.mjs";
import "./chunk-QJ63TWAK.mjs";
import "./chunk-OV6YUU3D.mjs";
import {
  RouteFocusModal,
  useRouteModal
} from "./chunk-ULMMKZ3U.mjs";
import "./chunk-OBQI23QM.mjs";
import {
  useAddRbacRoleUsers
} from "./chunk-OCESTHZJ.mjs";
import {
  useUsers
} from "./chunk-CHDKZVIP.mjs";
import "./chunk-FXYH54JP.mjs";
import "./chunk-774WSTCC.mjs";
import "./chunk-NFEK63OE.mjs";
import "./chunk-QZ7TP4HQ.mjs";

// src/routes/roles/role-add-users/role-add-users.tsx
import { useParams } from "react-router-dom";

// src/routes/roles/role-add-users/components/add-users-form/add-users-form.tsx
import { zodResolver } from "@hookform/resolvers/zod";
import { Button, Checkbox, Hint, Tooltip, toast } from "@medusajs/ui";
import {
  createColumnHelper
} from "@tanstack/react-table";
import { useEffect, useMemo, useState } from "react";
import { useForm } from "react-hook-form";
import { useTranslation } from "react-i18next";
import * as zod from "zod";
import { jsx, jsxs } from "react/jsx-runtime";
var AddUsersSchema = zod.object({
  user_ids: zod.array(zod.string()).min(1)
});
var PAGE_SIZE = 10;
var AddUsersForm = ({ roleId }) => {
  const { t } = useTranslation();
  const { handleSuccess } = useRouteModal();
  const { hasAllPermissions } = usePermissions();
  const canManageRole = hasAllPermissions(["user:update", "rbac_role:update"]);
  const form = useForm({
    defaultValues: {
      user_ids: []
    },
    resolver: zodResolver(AddUsersSchema)
  });
  const { setValue } = form;
  const [rowSelection, setRowSelection] = useState({});
  useEffect(() => {
    setValue(
      "user_ids",
      Object.keys(rowSelection).filter((k) => rowSelection[k]),
      {
        shouldDirty: true,
        shouldTouch: true
      }
    );
  }, [rowSelection, setValue]);
  useEffect(() => {
    if (!canManageRole && Object.keys(rowSelection).length) {
      setRowSelection({});
    }
  }, [canManageRole, rowSelection, setRowSelection]);
  const queryObject = useQueryParams(["offset", "q", "order"]);
  const { offset, q, order } = queryObject;
  const {
    users,
    count,
    isPending: isLoading,
    isError,
    error
  } = useUsers({
    limit: PAGE_SIZE,
    offset: offset ? Number(offset) : 0,
    q,
    order,
    fields: "id,email,first_name,last_name,created_at,rbac_roles.id"
  });
  const updater = (fn) => {
    const state = typeof fn === "function" ? fn(rowSelection) : fn;
    const ids = Object.keys(state);
    setValue("user_ids", ids, {
      shouldDirty: true,
      shouldTouch: true
    });
    setRowSelection(state);
  };
  const columns = useColumns({ roleId });
  const { table } = useDataTable({
    data: users ?? [],
    columns,
    count,
    enablePagination: true,
    enableRowSelection: (row) => {
      const rowRoles = row.original?.rbac_roles ?? [];
      return canManageRole && !rowRoles.some((rbacRole) => rbacRole.id === roleId);
    },
    getRowId: (row) => row.id,
    pageSize: PAGE_SIZE,
    rowSelection: {
      state: rowSelection,
      updater
    }
  });
  const { mutateAsync, isPending } = useAddRbacRoleUsers(roleId);
  const handleSubmit = form.handleSubmit(async (data) => {
    if (!canManageRole) {
      return;
    }
    await mutateAsync(data.user_ids, {
      onSuccess: () => {
        toast.success(
          t("roles.users.add.successToast", {
            count: data.user_ids.length
          })
        );
        handleSuccess(`/settings/roles/${roleId}`);
      },
      onError: (error2) => {
        toast.error(error2.message);
      }
    });
  });
  if (isError) {
    throw error;
  }
  return /* @__PURE__ */ jsx(RouteFocusModal.Form, { form, children: /* @__PURE__ */ jsxs(
    KeyboundForm,
    {
      className: "flex h-full flex-col overflow-hidden",
      onSubmit: handleSubmit,
      children: [
        /* @__PURE__ */ jsx(RouteFocusModal.Header, { children: /* @__PURE__ */ jsx("div", { className: "flex items-center justify-end gap-x-2", children: form.formState.errors.user_ids && /* @__PURE__ */ jsx(Hint, { variant: "error", children: form.formState.errors.user_ids.message }) }) }),
        /* @__PURE__ */ jsx(RouteFocusModal.Body, { className: "size-full overflow-hidden", children: /* @__PURE__ */ jsx(
          _DataTable,
          {
            table,
            columns,
            pageSize: PAGE_SIZE,
            count,
            isLoading,
            layout: "fill",
            search: "autofocus",
            orderBy: [
              { key: "email", label: t("fields.email") },
              { key: "first_name", label: t("fields.firstName") },
              { key: "last_name", label: t("fields.lastName") },
              { key: "created_at", label: t("fields.createdAt") },
              { key: "updated_at", label: t("fields.updatedAt") }
            ],
            queryObject,
            noRecords: {
              message: t("roles.users.add.list.noRecordsMessage")
            }
          }
        ) }),
        /* @__PURE__ */ jsxs(RouteFocusModal.Footer, { children: [
          /* @__PURE__ */ jsx(RouteFocusModal.Close, { asChild: true, children: /* @__PURE__ */ jsx(Button, { variant: "secondary", size: "small", children: t("actions.cancel") }) }),
          canManageRole && /* @__PURE__ */ jsx(
            Button,
            {
              type: "submit",
              variant: "primary",
              size: "small",
              isLoading: isPending,
              children: t("actions.save")
            }
          )
        ] })
      ]
    }
  ) });
};
var columnHelper = createColumnHelper();
var useColumns = ({ roleId }) => {
  const { t } = useTranslation();
  const { getFullDate } = useDate();
  return useMemo(
    () => [
      columnHelper.display({
        id: "select",
        header: ({ table }) => {
          return /* @__PURE__ */ jsx(
            Checkbox,
            {
              checked: table.getIsSomePageRowsSelected() ? "indeterminate" : table.getIsAllPageRowsSelected(),
              onCheckedChange: (value) => table.toggleAllPageRowsSelected(!!value)
            }
          );
        },
        cell: ({ row }) => {
          const rowRoles = row.original?.rbac_roles ?? [];
          const isAlreadyAdded = rowRoles.some(
            (rbacRole) => rbacRole.id === roleId
          );
          const isSelected = row.getIsSelected() || isAlreadyAdded;
          const Component = /* @__PURE__ */ jsx(
            Checkbox,
            {
              checked: isSelected,
              disabled: isAlreadyAdded,
              onCheckedChange: (value) => row.toggleSelected(!!value),
              onClick: (e) => {
                e.stopPropagation();
              }
            }
          );
          if (isAlreadyAdded) {
            return /* @__PURE__ */ jsx(
              Tooltip,
              {
                content: t("roles.users.alreadyAddedTooltip"),
                side: "right",
                children: Component
              }
            );
          }
          return Component;
        }
      }),
      columnHelper.accessor("first_name", {
        header: t("fields.firstName"),
        cell: ({ row }) => row.original.first_name || "-"
      }),
      columnHelper.accessor("last_name", {
        header: t("fields.lastName"),
        cell: ({ row }) => row.original.last_name || "-"
      }),
      columnHelper.accessor("email", {
        header: t("fields.email"),
        cell: ({ row }) => row.original.email
      }),
      columnHelper.accessor("created_at", {
        header: t("fields.createdAt"),
        cell: ({ row }) => getFullDate({ date: row.original.created_at })
      })
    ],
    [t, getFullDate, roleId]
  );
};

// src/routes/roles/role-add-users/role-add-users.tsx
import { jsx as jsx2 } from "react/jsx-runtime";
var RoleAddUsers = () => {
  const { id } = useParams();
  return /* @__PURE__ */ jsx2(RouteFocusModal, { children: /* @__PURE__ */ jsx2(AddUsersForm, { roleId: id }) });
};
export {
  RoleAddUsers as Component
};
