// @medusajs/dashboard/dist/role-list-PL7GHASQ.mjs
import {
  useRequireRbacFeature
} from "./chunk-Q7IJ6TPC.mjs";
import {
  ListSummary
} from "./chunk-I3VB6NM2.mjs";
import {
  SingleColumnPage
} from "./chunk-HLYIVAII.mjs";
import {
  useDataTableDateFilters
} from "./chunk-V2OKIIO7.mjs";
import {
  DataTable
} from "./chunk-N4L2ZHYI.mjs";
import {
  usePermissions
} from "./chunk-HFX2KPQD.mjs";
import "./chunk-PJU3RODH.mjs";
import "./chunk-BPIG7PI6.mjs";
import {
  useExtension
} from "./chunk-5SZFF255.mjs";
import "./chunk-OZPB6JBL.mjs";
import "./chunk-OC7BQLYI.mjs";
import "./chunk-S4DMV3ZT.mjs";
import {
  useQueryParams
} from "./chunk-C76H5USB.mjs";
import {
  useDate
} from "./chunk-3BQAAPDR.mjs";
import "./chunk-QJ63TWAK.mjs";
import {
  useDeleteRbacRoleLazy,
  useRbacRoles
} from "./chunk-OCESTHZJ.mjs";
import "./chunk-BEQ5XTWH.mjs";
import "./chunk-FXYH54JP.mjs";
import "./chunk-774WSTCC.mjs";
import "./chunk-NFEK63OE.mjs";
import "./chunk-QZ7TP4HQ.mjs";

// src/routes/roles/role-list/components/role-list-table/role-list-table.tsx
import { PencilSquare, Trash } from "@medusajs/icons";
import {
  Container,
  createDataTableColumnHelper,
  toast,
  usePrompt
} from "@medusajs/ui";
import { keepPreviousData } from "@tanstack/react-query";
import { useCallback, useMemo } from "react";
import { useTranslation } from "react-i18next";
import { useNavigate } from "react-router-dom";
import { jsx } from "react/jsx-runtime";
var PAGE_SIZE = 20;
var RoleListTable = () => {
  const { t } = useTranslation();
  const { hasPermission } = usePermissions();
  const { q, order, offset, created_at } = useQueryParams([
    "q",
    "order",
    "offset",
    "created_at"
  ]);
  const { roles, count, isPending, isError, error } = useRbacRoles(
    {
      q,
      order,
      offset: offset ? parseInt(offset) : void 0,
      limit: PAGE_SIZE,
      created_at: created_at ? JSON.parse(created_at) : void 0,
      fields: "id,name,description,created_at,users.id,users.first_name,users.last_name,users.email"
    },
    {
      placeholderData: keepPreviousData
    }
  );
  const columns = useColumns({ hasPermission });
  const filters = useFilters();
  if (isError) {
    throw error;
  }
  const canCreate = hasPermission("rbac_role:create");
  return /* @__PURE__ */ jsx(Container, { className: "divide-y p-0", children: /* @__PURE__ */ jsx(
    DataTable,
    {
      data: roles,
      columns,
      filters,
      getRowId: (row) => row.id,
      rowHref: (row) => `${row.id}`,
      rowCount: count,
      pageSize: PAGE_SIZE,
      heading: t("roles.domain"),
      subHeading: t("roles.subtitle"),
      isLoading: isPending,
      action: canCreate ? {
        label: t("actions.create"),
        to: "create"
      } : void 0,
      emptyState: {
        empty: {
          heading: t("roles.list.empty.heading"),
          description: t("roles.list.empty.description")
        },
        filtered: {
          heading: t("roles.list.filtered.heading"),
          description: t("roles.list.filtered.description")
        }
      }
    }
  ) });
};
var columnHelper = createDataTableColumnHelper();
var useColumns = ({
  hasPermission
}) => {
  const { t } = useTranslation();
  const navigate = useNavigate();
  const prompt = usePrompt();
  const { getFullDate } = useDate();
  const { mutateAsync: deleteRole } = useDeleteRbacRoleLazy();
  const handleEdit = useCallback(
    (role) => {
      navigate(`/settings/roles/${role.id}/edit`);
    },
    [navigate]
  );
  const handleDelete = useCallback(
    async (role) => {
      const confirmed = await prompt({
        title: t("roles.delete.title"),
        description: t("roles.delete.description", { name: role.name }),
        confirmText: t("actions.delete"),
        cancelText: t("actions.cancel")
      });
      if (!confirmed) {
        return;
      }
      try {
        await deleteRole(role.id);
        toast.success(t("roles.delete.successToast", { name: role.name }));
      } catch (error) {
        toast.error(error.message);
      }
    },
    [prompt, deleteRole, t]
  );
  const canUpdate = hasPermission("rbac_role:update");
  const canDelete = hasPermission("rbac_role:delete");
  return useMemo(() => {
    const baseColumns = [
      columnHelper.accessor("name", {
        header: t("fields.name"),
        enableSorting: true,
        sortAscLabel: t("filters.sorting.alphabeticallyAsc"),
        sortDescLabel: t("filters.sorting.alphabeticallyDesc")
      }),
      columnHelper.accessor("description", {
        header: t("fields.description"),
        cell: ({ row }) => {
          return row.original.description || "-";
        }
      }),
      columnHelper.display({
        id: "users",
        header: t("users.domain"),
        cell: ({ row }) => {
          const users = row.original.users_link?.map((link) => link.user).filter((user) => !!user) ?? [];
          if (!users.length) {
            return "-";
          }
          const labels = users.map((user) => {
            const fullName = `${user.first_name || ""} ${user.last_name || ""}`.trim();
            return fullName || user.email || user.id;
          });
          return /* @__PURE__ */ jsx("div", { className: "flex items-center", children: /* @__PURE__ */ jsx(ListSummary, { inline: true, n: 1, list: labels }) });
        }
      }),
      columnHelper.accessor("created_at", {
        header: t("fields.createdAt"),
        cell: ({ row }) => {
          return /* @__PURE__ */ jsx("span", { children: getFullDate({ date: row.original.created_at }) });
        },
        enableSorting: true,
        sortAscLabel: t("filters.sorting.dateAsc"),
        sortDescLabel: t("filters.sorting.dateDesc")
      })
    ];
    if (!canUpdate && !canDelete) {
      return baseColumns;
    }
    const groups = [];
    if (canUpdate) {
      groups.push([
        {
          label: t("actions.edit"),
          icon: /* @__PURE__ */ jsx(PencilSquare, {}),
          onClick: (ctx) => handleEdit(ctx.row.original)
        }
      ]);
    }
    if (canDelete) {
      groups.push([
        {
          label: t("actions.delete"),
          icon: /* @__PURE__ */ jsx(Trash, {}),
          onClick: (ctx) => handleDelete(ctx.row.original)
        }
      ]);
    }
    return [
      ...baseColumns,
      columnHelper.action({
        actions: groups
      })
    ];
  }, [t, getFullDate, handleEdit, handleDelete, canUpdate, canDelete]);
};
var useFilters = () => {
  const dateFilters = useDataTableDateFilters();
  return useMemo(() => {
    return dateFilters.filter((filter) => filter.id === "created_at");
  }, [dateFilters]);
};

// src/routes/roles/role-list/role-list.tsx
import { jsx as jsx2 } from "react/jsx-runtime";
var RoleList = () => {
  const { getWidgets } = useExtension();
  const isRbacEnabled = useRequireRbacFeature();
  if (!isRbacEnabled) {
    return null;
  }
  return /* @__PURE__ */ jsx2(
    SingleColumnPage,
    {
      widgets: {
        before: getWidgets("role.list.before"),
        after: getWidgets("role.list.after")
      },
      children: /* @__PURE__ */ jsx2(RoleListTable, {})
    }
  );
};
export {
  RoleList as Component
};
