// @medusajs/dashboard/dist/role-create-XIBPRQW5.mjs
import {
  useRequireRbacFeature
} from "./chunk-Q7IJ6TPC.mjs";
import "./chunk-IUCDCPJU.mjs";
import {
  KeyboundForm
} from "./chunk-6HTZNHPT.mjs";
import {
  DataTable
} from "./chunk-N4L2ZHYI.mjs";
import "./chunk-PJU3RODH.mjs";
import "./chunk-OZPB6JBL.mjs";
import "./chunk-OC7BQLYI.mjs";
import {
  useDocumentDirection
} from "./chunk-S4DMV3ZT.mjs";
import {
  useQueryParams
} from "./chunk-C76H5USB.mjs";
import "./chunk-OV6YUU3D.mjs";
import {
  RouteFocusModal,
  useRouteModal
} from "./chunk-ULMMKZ3U.mjs";
import {
  Form
} from "./chunk-OBQI23QM.mjs";
import {
  useRbacAssignablePolicies
} from "./chunk-JHQC6AFI.mjs";
import {
  useCreateRbacRole
} from "./chunk-OCESTHZJ.mjs";
import "./chunk-BEQ5XTWH.mjs";
import "./chunk-FXYH54JP.mjs";
import "./chunk-774WSTCC.mjs";
import "./chunk-NFEK63OE.mjs";
import "./chunk-QZ7TP4HQ.mjs";

// src/routes/roles/role-create/components/create-role-form/create-role-form.tsx
import { zodResolver } from "@hookform/resolvers/zod";
import {
  Button,
  createDataTableColumnHelper,
  Heading,
  Input,
  ProgressTabs,
  Text,
  Textarea,
  toast
} from "@medusajs/ui";
import { keepPreviousData } from "@tanstack/react-query";
import { useCallback, useEffect, useMemo, useState } from "react";
import { useForm } from "react-hook-form";
import { useTranslation } from "react-i18next";
import { z } from "zod";
import { jsx, jsxs } from "react/jsx-runtime";
var CreateRoleDetailsSchema = z.object({
  name: z.string().min(1),
  description: z.string().optional()
});
var CreateRoleSchema = CreateRoleDetailsSchema.extend({
  policies: z.array(z.string()).optional()
});
var PAGE_SIZE = 20;
var PREFIX = "rp";
var columnHelper = createDataTableColumnHelper();
var CreateRoleForm = () => {
  const { t } = useTranslation();
  const { handleSuccess } = useRouteModal();
  const direction = useDocumentDirection();
  const [tab, setTab] = useState("details" /* DETAILS */);
  const [tabState, setTabState] = useState({
    ["details" /* DETAILS */]: "in-progress",
    ["permissions" /* PERMISSIONS */]: "not-started"
  });
  const [rowSelection, setRowSelection] = useState(
    {}
  );
  const form = useForm({
    defaultValues: {
      name: "",
      description: "",
      policies: []
    },
    resolver: zodResolver(CreateRoleSchema)
  });
  const handleTabChange = async (nextTab) => {
    const valid = await form.trigger();
    if (!valid) {
      return;
    }
    setTab(nextTab);
  };
  const { mutateAsync: createRole, isPending: isCreating } = useCreateRbacRole();
  const onRowSelectionChange = useCallback(
    (selection) => {
      const ids = Object.keys(selection).filter((id) => selection[id]);
      form.setValue("policies", ids, {
        shouldDirty: true,
        shouldTouch: true
      });
      setRowSelection(selection);
    },
    [form]
  );
  const { q, order, offset } = useQueryParams(["q", "order", "offset"], PREFIX);
  const {
    data: pageData,
    isPending: isPoliciesLoading,
    isError: isPoliciesError,
    error: policiesError
  } = useRbacAssignablePolicies(
    {
      q,
      order,
      offset: offset ? parseInt(offset) : 0,
      limit: PAGE_SIZE,
      fields: "id,key,resource,operation,description"
    },
    {
      placeholderData: keepPreviousData
    }
  );
  if (isPoliciesError) {
    throw policiesError;
  }
  const visiblePolicies = pageData?.policies ?? [];
  const count = pageData?.count ?? 0;
  const columns = usePolicyColumns();
  const onNext = async (currentTab) => {
    const valid = await form.trigger();
    if (!valid) {
      return;
    }
    if (currentTab === "details" /* DETAILS */) {
      setTab("permissions" /* PERMISSIONS */);
    }
  };
  useEffect(() => {
    const currentState = { ...tabState };
    currentState["details" /* DETAILS */] = tab === "details" /* DETAILS */ ? "in-progress" : "completed";
    currentState["permissions" /* PERMISSIONS */] = tab === "details" /* DETAILS */ ? "not-started" : "in-progress";
    setTabState({ ...currentState });
  }, [tab]);
  const handleSubmit = form.handleSubmit(async (values) => {
    try {
      const { role } = await createRole({
        name: values.name.trim(),
        description: values.description?.trim() || null,
        policy_ids: values.policies?.length ? values.policies : void 0
      });
      toast.success(
        t("roles.create.successToast", {
          name: role.name
        })
      );
      handleSuccess(`/settings/roles/${role.id}`);
    } catch (error) {
      toast.error(error.message);
    }
  });
  return /* @__PURE__ */ jsx(RouteFocusModal.Form, { form, children: /* @__PURE__ */ jsxs(
    KeyboundForm,
    {
      onSubmit: handleSubmit,
      className: "flex size-full flex-col overflow-hidden",
      children: [
        /* @__PURE__ */ jsxs(
          ProgressTabs,
          {
            dir: direction,
            value: tab,
            onValueChange: (tab2) => handleTabChange(tab2),
            className: "flex h-full flex-col overflow-hidden",
            children: [
              /* @__PURE__ */ jsx(RouteFocusModal.Header, { children: /* @__PURE__ */ jsx("div", { className: "-my-2 w-full border-l", children: /* @__PURE__ */ jsxs(ProgressTabs.List, { className: "justify-start-start flex w-full items-center", children: [
                /* @__PURE__ */ jsx(
                  ProgressTabs.Trigger,
                  {
                    status: tabState["details" /* DETAILS */],
                    value: "details" /* DETAILS */,
                    className: "max-w-[200px] truncate",
                    children: t("roles.create.tabs.details")
                  }
                ),
                /* @__PURE__ */ jsx(
                  ProgressTabs.Trigger,
                  {
                    status: tabState["permissions" /* PERMISSIONS */],
                    value: "permissions" /* PERMISSIONS */,
                    className: "max-w-[200px] truncate",
                    children: t("roles.create.tabs.permissions")
                  }
                )
              ] }) }) }),
              /* @__PURE__ */ jsxs(RouteFocusModal.Body, { className: "size-full overflow-hidden", children: [
                /* @__PURE__ */ jsx(
                  ProgressTabs.Content,
                  {
                    className: "size-full overflow-y-auto",
                    value: "details" /* DETAILS */,
                    children: /* @__PURE__ */ jsx("div", { className: "flex flex-col items-center overflow-auto p-16", children: /* @__PURE__ */ jsxs("div", { className: "flex w-full max-w-[720px] flex-col gap-y-8", children: [
                      /* @__PURE__ */ jsxs("div", { children: [
                        /* @__PURE__ */ jsx(Heading, { children: t("roles.create.header") }),
                        /* @__PURE__ */ jsx(Text, { size: "small", className: "text-ui-fg-subtle", children: t("roles.create.hint") })
                      ] }),
                      /* @__PURE__ */ jsxs("div", { className: "grid grid-cols-1 gap-4", children: [
                        /* @__PURE__ */ jsx(
                          Form.Field,
                          {
                            control: form.control,
                            name: "name",
                            render: ({ field }) => {
                              return /* @__PURE__ */ jsxs(Form.Item, { children: [
                                /* @__PURE__ */ jsx(Form.Label, { children: t("fields.name") }),
                                /* @__PURE__ */ jsx(Form.Control, { children: /* @__PURE__ */ jsx(Input, { ...field }) }),
                                /* @__PURE__ */ jsx(Form.ErrorMessage, {})
                              ] });
                            }
                          }
                        ),
                        /* @__PURE__ */ jsx(
                          Form.Field,
                          {
                            control: form.control,
                            name: "description",
                            render: ({ field }) => {
                              return /* @__PURE__ */ jsxs(Form.Item, { children: [
                                /* @__PURE__ */ jsx(Form.Label, { optional: true, children: t("fields.description") }),
                                /* @__PURE__ */ jsx(Form.Control, { children: /* @__PURE__ */ jsx(Textarea, { ...field, rows: 4 }) }),
                                /* @__PURE__ */ jsx(Form.ErrorMessage, {})
                              ] });
                            }
                          }
                        )
                      ] })
                    ] }) })
                  }
                ),
                /* @__PURE__ */ jsx(
                  ProgressTabs.Content,
                  {
                    className: "size-full overflow-y-auto",
                    value: "permissions" /* PERMISSIONS */,
                    children: /* @__PURE__ */ jsx(
                      DataTable,
                      {
                        data: visiblePolicies,
                        columns,
                        getRowId: (row) => row.id,
                        rowCount: count,
                        pageSize: PAGE_SIZE,
                        isLoading: isPoliciesLoading,
                        prefix: PREFIX,
                        rowSelection: {
                          state: rowSelection,
                          onRowSelectionChange
                        },
                        layout: "fill",
                        emptyState: {
                          empty: {
                            heading: t("roles.create.permissions.empty.heading"),
                            description: t(
                              "roles.create.permissions.empty.description"
                            )
                          },
                          filtered: {
                            heading: t("roles.create.permissions.filtered.heading"),
                            description: t(
                              "roles.create.permissions.filtered.description"
                            )
                          }
                        }
                      }
                    )
                  }
                )
              ] })
            ]
          }
        ),
        /* @__PURE__ */ jsx(RouteFocusModal.Footer, { children: /* @__PURE__ */ jsxs("div", { className: "flex items-center justify-end gap-x-2", children: [
          /* @__PURE__ */ jsx(RouteFocusModal.Close, { asChild: true, children: /* @__PURE__ */ jsx(Button, { variant: "secondary", size: "small", children: t("actions.cancel") }) }),
          /* @__PURE__ */ jsx(PrimaryButton, { tab, next: onNext, isLoading: isCreating })
        ] }) })
      ]
    }
  ) });
};
var usePolicyColumns = () => {
  const { t } = useTranslation();
  return useMemo(() => {
    return [
      columnHelper.select(),
      columnHelper.accessor("key", {
        header: t("fields.key")
      }),
      columnHelper.accessor("resource", {
        header: t("fields.resource"),
        cell: ({ row }) => {
          const resource = row.original.resource;
          return t(`permissions.resources.${resource}`, {
            defaultValue: resource
          });
        }
      }),
      columnHelper.accessor("operation", {
        header: t("fields.operation"),
        cell: ({ row }) => {
          const operation = row.original.operation;
          return t(`permissions.actions.${operation}`, {
            defaultValue: operation
          });
        }
      }),
      columnHelper.accessor("description", {
        header: t("fields.description"),
        cell: ({ row }) => {
          return row.original.description || "-";
        }
      })
    ];
  }, [t]);
};
var PrimaryButton = ({ tab, next, isLoading }) => {
  const { t } = useTranslation();
  if (tab === "permissions" /* PERMISSIONS */) {
    return /* @__PURE__ */ jsx(
      Button,
      {
        type: "submit",
        variant: "primary",
        size: "small",
        isLoading,
        children: t("actions.create")
      },
      "submit-button"
    );
  }
  return /* @__PURE__ */ jsx(
    Button,
    {
      type: "button",
      variant: "primary",
      size: "small",
      onClick: () => next(tab),
      children: t("actions.continue")
    },
    "next-button"
  );
};

// src/routes/roles/role-create/role-create.tsx
import { jsx as jsx2 } from "react/jsx-runtime";
var RoleCreate = () => {
  const isRbacEnabled = useRequireRbacFeature();
  if (!isRbacEnabled) {
    return null;
  }
  return /* @__PURE__ */ jsx2(RouteFocusModal, { children: /* @__PURE__ */ jsx2(CreateRoleForm, {}) });
};
export {
  RoleCreate as Component
};
