// @medusajs/dashboard/dist/user-invite-RD3EBKNP.mjs
import {
  ListSummary
} from "./chunk-I3VB6NM2.mjs";
import "./chunk-LQTHYS2Z.mjs";
import {
  _DataTable,
  useDataTable
} from "./chunk-CR5DSAVZ.mjs";
import "./chunk-HQKGZADC.mjs";
import "./chunk-EMIHDNB7.mjs";
import {
  Combobox
} from "./chunk-RDM5FCA6.mjs";
import {
  KeyboundForm
} from "./chunk-6HTZNHPT.mjs";
import {
  isFetchError
} from "./chunk-ONB3JEHR.mjs";
import {
  usePermissions
} from "./chunk-HFX2KPQD.mjs";
import "./chunk-LPEUYMRK.mjs";
import {
  useFeatureFlag
} from "./chunk-PJU3RODH.mjs";
import {
  ActionMenu
} from "./chunk-OZPB6JBL.mjs";
import "./chunk-OC7BQLYI.mjs";
import "./chunk-S4DMV3ZT.mjs";
import {
  useUserInviteTableQuery
} from "./chunk-CEYKNZTH.mjs";
import "./chunk-XRM7PIRS.mjs";
import "./chunk-C76H5USB.mjs";
import "./chunk-3BQAAPDR.mjs";
import "./chunk-QJ63TWAK.mjs";
import "./chunk-OV6YUU3D.mjs";
import {
  RouteFocusModal
} from "./chunk-ULMMKZ3U.mjs";
import {
  Form
} from "./chunk-OBQI23QM.mjs";
import {
  useCreateInvite,
  useDeleteInvite,
  useInvites,
  useResendInvite
} from "./chunk-CN7JXSGW.mjs";
import {
  useRbacAssignableRoles
} from "./chunk-OCESTHZJ.mjs";
import "./chunk-FXYH54JP.mjs";
import "./chunk-774WSTCC.mjs";
import "./chunk-NFEK63OE.mjs";
import "./chunk-QZ7TP4HQ.mjs";

// src/routes/users/user-invite/components/invite-user-form/invite-user-form.tsx
import { zodResolver } from "@hookform/resolvers/zod";
import { ArrowPath, Link, Trash } from "@medusajs/icons";
import {
  Alert,
  Button,
  Container,
  Heading,
  Input,
  StatusBadge,
  Text,
  Tooltip,
  usePrompt
} from "@medusajs/ui";
import { createColumnHelper } from "@tanstack/react-table";
import copy from "copy-to-clipboard";
import { format } from "date-fns";
import { useMemo } from "react";
import { useForm } from "react-hook-form";
import { Trans, useTranslation } from "react-i18next";
import * as zod from "zod";
import { jsx, jsxs } from "react/jsx-runtime";
var InviteUserSchema = zod.object({
  email: zod.string().email(),
  roles: zod.array(zod.string()).optional()
});
var PAGE_SIZE = 10;
var PREFIX = "usr_invite";
var INVITE_URL = `${window.location.origin}${__BASE__ === "/" ? "" : __BASE__}/invite?token=`;
var InviteUserForm = () => {
  const { t } = useTranslation();
  const isRbacEnabled = useFeatureFlag("rbac");
  const { hasPermission } = usePermissions();
  const canReadRbacRoles = hasPermission("rbac_role:read");
  const showRbacRolesField = isRbacEnabled && canReadRbacRoles;
  const form = useForm({
    defaultValues: {
      email: "",
      roles: []
    },
    resolver: zodResolver(InviteUserSchema)
  });
  const { data: assignableData, isPending: isRolesLoading } = useRbacAssignableRoles(
    { limit: 200, order: "name" },
    { enabled: showRbacRolesField }
  );
  const roleOptions = useMemo(() => {
    return (assignableData?.roles ?? []).map((role) => ({
      label: role.name,
      value: role.id
    }));
  }, [assignableData?.roles]);
  const inviteFields = useMemo(() => {
    if (!showRbacRolesField) {
      return void 0;
    }
    return [
      "id",
      "email",
      "accepted",
      "token",
      "expires_at",
      "created_at",
      "updated_at",
      "rbac_roles.id",
      "rbac_roles.name"
    ].join(",");
  }, [showRbacRolesField]);
  const { raw, searchParams } = useUserInviteTableQuery({
    prefix: PREFIX,
    pageSize: PAGE_SIZE,
    fields: inviteFields
  });
  const {
    invites,
    count,
    isPending: isLoading,
    isError,
    error
  } = useInvites(searchParams);
  const columns = useColumns({ isRbacEnabled: showRbacRolesField });
  const { table } = useDataTable({
    data: invites ?? [],
    columns,
    count,
    enablePagination: true,
    getRowId: (row) => row.id,
    pageSize: PAGE_SIZE,
    prefix: PREFIX
  });
  const { mutateAsync, isPending } = useCreateInvite();
  const handleSubmit = form.handleSubmit(async (values) => {
    try {
      const payload = {
        email: values.email
      };
      if (showRbacRolesField && values.roles?.length) {
        payload.roles = values.roles;
      }
      await mutateAsync(payload);
      form.reset();
    } catch (error2) {
      if (isFetchError(error2) && error2.status === 400) {
        form.setError("root", {
          type: "manual",
          message: error2.message
        });
        return;
      }
    }
  });
  if (isError) {
    throw error;
  }
  return /* @__PURE__ */ jsx(RouteFocusModal.Form, { form, children: /* @__PURE__ */ jsxs(
    KeyboundForm,
    {
      onSubmit: handleSubmit,
      className: "flex h-full flex-col overflow-hidden",
      children: [
        /* @__PURE__ */ jsx(RouteFocusModal.Header, {}),
        /* @__PURE__ */ jsx(RouteFocusModal.Body, { className: "flex flex-1 flex-col overflow-hidden", children: /* @__PURE__ */ jsx("div", { className: "flex flex-1 flex-col items-center overflow-y-auto", children: /* @__PURE__ */ jsxs("div", { className: "flex w-full max-w-[720px] flex-col gap-y-8 px-2 py-16", children: [
          /* @__PURE__ */ jsxs("div", { children: [
            /* @__PURE__ */ jsx(Heading, { children: t("users.inviteUser") }),
            /* @__PURE__ */ jsx(Text, { size: "small", className: "text-ui-fg-subtle", children: t("users.inviteUserHint") })
          ] }),
          form.formState.errors.root && /* @__PURE__ */ jsx(
            Alert,
            {
              variant: "error",
              dismissible: false,
              className: "text-balance",
              children: form.formState.errors.root.message
            }
          ),
          /* @__PURE__ */ jsxs("div", { className: "flex flex-col gap-y-4", children: [
            /* @__PURE__ */ jsxs("div", { className: "grid grid-cols-2 gap-4", children: [
              /* @__PURE__ */ jsx(
                Form.Field,
                {
                  control: form.control,
                  name: "email",
                  render: ({ field }) => {
                    return /* @__PURE__ */ jsxs(Form.Item, { children: [
                      /* @__PURE__ */ jsx(Form.Label, { children: t("fields.email") }),
                      /* @__PURE__ */ jsx(Form.Control, { children: /* @__PURE__ */ jsx(
                        Input,
                        {
                          ...field,
                          placeholder: "john.doe@example.com"
                        }
                      ) }),
                      /* @__PURE__ */ jsx(Form.ErrorMessage, {})
                    ] });
                  }
                }
              ),
              showRbacRolesField && /* @__PURE__ */ jsx(
                Form.Field,
                {
                  control: form.control,
                  name: "roles",
                  render: ({ field }) => {
                    return /* @__PURE__ */ jsxs(Form.Item, { children: [
                      /* @__PURE__ */ jsx(
                        Form.Label,
                        {
                          optional: true,
                          tooltip: t("users.inviteRolesTooltip"),
                          children: t("roles.domain")
                        }
                      ),
                      /* @__PURE__ */ jsx(Form.Control, { children: /* @__PURE__ */ jsx(
                        Combobox,
                        {
                          ...field,
                          value: field.value ?? [],
                          onChange: (value) => {
                            field.onChange(value ?? []);
                          },
                          options: roleOptions,
                          placeholder: t("labels.selectValues"),
                          disabled: isRolesLoading
                        }
                      ) }),
                      /* @__PURE__ */ jsx(Form.ErrorMessage, {})
                    ] });
                  }
                }
              )
            ] }),
            /* @__PURE__ */ jsx("div", { className: "flex items-center justify-end", children: /* @__PURE__ */ jsx(
              Button,
              {
                size: "small",
                variant: "secondary",
                type: "submit",
                isLoading: isPending,
                children: t("users.sendInvite")
              }
            ) })
          ] }),
          /* @__PURE__ */ jsxs("div", { className: "flex flex-col gap-y-4", children: [
            /* @__PURE__ */ jsx(Heading, { level: "h2", children: t("users.pendingInvites") }),
            /* @__PURE__ */ jsx(Container, { className: "overflow-hidden p-0", children: /* @__PURE__ */ jsx(
              _DataTable,
              {
                table,
                columns,
                count,
                pageSize: PAGE_SIZE,
                pagination: true,
                search: "autofocus",
                isLoading,
                queryObject: raw,
                prefix: PREFIX,
                orderBy: [
                  { key: "email", label: t("fields.email") },
                  { key: "created_at", label: t("fields.createdAt") },
                  { key: "updated_at", label: t("fields.updatedAt") }
                ]
              }
            ) })
          ] })
        ] }) }) })
      ]
    }
  ) });
};
var InviteActions = ({ invite }) => {
  const { mutateAsync: revokeAsync } = useDeleteInvite(invite.id);
  const { mutateAsync: resendAsync } = useResendInvite(invite.id);
  const prompt = usePrompt();
  const { t } = useTranslation();
  const handleDelete = async () => {
    const res = await prompt({
      title: t("general.areYouSure"),
      description: t("users.deleteInviteWarning", {
        email: invite.email
      }),
      cancelText: t("actions.cancel"),
      confirmText: t("actions.delete")
    });
    if (!res) {
      return;
    }
    await revokeAsync();
  };
  const handleResend = async () => {
    await resendAsync();
  };
  const handleCopyInviteLink = () => {
    const inviteUrl = `${INVITE_URL}${invite.token}`;
    copy(inviteUrl);
  };
  return /* @__PURE__ */ jsx(
    ActionMenu,
    {
      groups: [
        {
          actions: [
            {
              icon: /* @__PURE__ */ jsx(ArrowPath, {}),
              label: t("users.resendInvite"),
              onClick: handleResend
            }
          ]
        },
        {
          actions: [
            {
              icon: /* @__PURE__ */ jsx(Link, {}),
              label: t("users.copyInviteLink"),
              onClick: handleCopyInviteLink
            }
          ]
        },
        {
          actions: [
            {
              icon: /* @__PURE__ */ jsx(Trash, {}),
              label: t("actions.delete"),
              onClick: handleDelete
            }
          ]
        }
      ]
    }
  );
};
var columnHelper = createColumnHelper();
var useColumns = ({ isRbacEnabled }) => {
  const { t } = useTranslation();
  return useMemo(
    () => [
      columnHelper.accessor("email", {
        header: t("fields.email"),
        cell: ({ getValue }) => {
          return getValue();
        }
      }),
      ...isRbacEnabled ? [
        columnHelper.display({
          id: "roles",
          header: t("roles.domain"),
          cell: ({ row }) => {
            const roleNames = row.original.rbac_roles?.map((role) => role.name) ?? [];
            if (!roleNames.length) {
              return /* @__PURE__ */ jsx(Text, { size: "small", className: "text-ui-fg-subtle", children: "-" });
            }
            return /* @__PURE__ */ jsx("div", { className: "flex items-center", children: /* @__PURE__ */ jsx(ListSummary, { inline: true, n: 1, list: roleNames }) });
          }
        })
      ] : [],
      columnHelper.accessor("accepted", {
        header: t("fields.status"),
        cell: ({ getValue, row }) => {
          const accepted = getValue();
          const expired = new Date(row.original.expires_at) < /* @__PURE__ */ new Date();
          if (accepted) {
            return /* @__PURE__ */ jsx(
              Tooltip,
              {
                content: t("users.acceptedOnDate", {
                  date: format(
                    new Date(row.original.updated_at),
                    "dd MMM, yyyy"
                  )
                }),
                children: /* @__PURE__ */ jsx(StatusBadge, { color: "green", children: t("users.inviteStatus.accepted") })
              }
            );
          }
          if (expired) {
            return /* @__PURE__ */ jsx(
              Tooltip,
              {
                content: t("users.expiredOnDate", {
                  date: format(
                    new Date(row.original.expires_at),
                    "dd MMM, yyyy"
                  )
                }),
                children: /* @__PURE__ */ jsx(StatusBadge, { color: "red", children: t("users.inviteStatus.expired") })
              }
            );
          }
          return /* @__PURE__ */ jsx(
            Tooltip,
            {
              content: /* @__PURE__ */ jsx(
                Trans,
                {
                  i18nKey: "users.validFromUntil",
                  components: [
                    /* @__PURE__ */ jsx("span", { className: "font-medium" }, "from"),
                    /* @__PURE__ */ jsx("span", { className: "font-medium" }, "untill")
                  ],
                  values: {
                    from: format(
                      new Date(row.original.created_at),
                      "dd MMM, yyyy"
                    ),
                    until: format(
                      new Date(row.original.expires_at),
                      "dd MMM, yyyy"
                    )
                  }
                }
              ),
              children: /* @__PURE__ */ jsx(StatusBadge, { color: "orange", children: t("users.inviteStatus.pending") })
            }
          );
        }
      }),
      columnHelper.display({
        id: "actions",
        cell: ({ row }) => /* @__PURE__ */ jsx(InviteActions, { invite: row.original })
      })
    ],
    [t, isRbacEnabled]
  );
};

// src/routes/users/user-invite/user-invite.tsx
import { jsx as jsx2 } from "react/jsx-runtime";
var UserInvite = () => {
  return /* @__PURE__ */ jsx2(RouteFocusModal, { children: /* @__PURE__ */ jsx2(InviteUserForm, {}) });
};
export {
  UserInvite as Component
};
