// @medusajs/medusa/dist/api/admin/rbac/policies/assignable/route.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GET = void 0;
const core_flows_1 = require("@medusajs/core-flows");
const utils_1 = require("@medusajs/framework/utils");
const rbac_1 = __importDefault(require("../../../../../feature-flags/rbac"));
/**
 * Returns the subset of `rbac_policy` rows the authenticated actor is allowed to assign.
 *
 * @ignore
 * @featureFlag rbac
 */
const GET = async (req, res) => {
    const { result } = await (0, core_flows_1.getAssignablePoliciesWorkflow)(req.scope).run({
        input: {
            actor_id: req.auth_context.actor_id,
            actor: req.auth_context.actor_type,
            filters: req.filterableFields,
            pagination: req.queryConfig?.pagination,
        },
    });
    res.status(200).json(result);
};
exports.GET = GET;
(0, utils_1.defineFileConfig)({
    isDisabled: () => !utils_1.FeatureFlag.isFeatureEnabled(rbac_1.default.key),
});
