// @medusajs/core-flows/dist/auth/steps/create-password-reset-token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPasswordResetTokenStep = exports.createPasswordResetTokenStepId = void 0;
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
/**
 * The ID of the create password reset token step.
 *
 * @since 2.16.0
 */
exports.createPasswordResetTokenStepId = "create-password-reset-token";
/**
 * Issues a single-use password reset token via the auth module. The token's
 * `jti` is meant to be embedded in the JWT delivered to the user; the auth
 * module stores its hash in a dedicated table so the consume side can verify
 * and atomically delete it on first use.
 *
 * Issuing a new token automatically invalidates any prior password-reset
 * tokens for the same provider identity. Compensation consumes the token
 * (best-effort) if the workflow rolls back.
 *
 * @since 2.16.0
 */
exports.createPasswordResetTokenStep = (0, workflows_sdk_1.createStep)(exports.createPasswordResetTokenStepId, async (data, { container }) => {
    const authModule = container.resolve(utils_1.Modules.AUTH);
    const { jti, expires_at } = await authModule.createPasswordResetToken({
        provider: data.provider,
        entity_id: data.entityId,
        ttl_seconds: data.ttlSeconds,
    });
    return new workflows_sdk_1.StepResponse({ jti, expiresAt: new Date(expires_at).toISOString() }, { jti, provider: data.provider, entityId: data.entityId });
}, async (compensateData, { container }) => {
    if (!compensateData) {
        return;
    }
    const authModule = container.resolve(utils_1.Modules.AUTH);
    try {
        await authModule.consumePasswordResetToken({
            jti: compensateData.jti,
            provider: compensateData.provider,
            entity_id: compensateData.entityId,
        });
    }
    catch {
        // Token may already be consumed; compensation is best-effort.
    }
});
