// @medusajs/medusa/dist/api/auth/utils/generate-jwt-token.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateJwtTokenWithChecks = generateJwtTokenWithChecks;
exports.generateJwtTokenForAuthIdentity = generateJwtTokenForAuthIdentity;
const utils_1 = require("@medusajs/framework/utils");
const rbac_1 = __importDefault(require("../../../feature-flags/rbac"));
const validate_verification_1 = require("./validate-verification");
async function generateJwtTokenWithChecks(container, { authIdentity, mfaChallenge, actorType, authProvider, }) {
    const config = container.resolve(utils_1.ContainerRegistrationKeys.CONFIG_MODULE);
    const { http } = config.projectConfig;
    const actorlessToken = await generateJwtTokenForAuthIdentity({
        authIdentity,
        actorType: actorType,
        authProvider: authProvider,
        container: container,
    }, {
        secret: http.jwtSecret,
        expiresIn: http.jwtExpiresIn,
        // Running a verification is about the auth identity, so we return a token to be able to authenticate the requests
        // without having an actor tied to it until the verification is completed.
        skipActorType: true,
        options: http.jwtOptions,
    });
    // Check if verification of the provider entity data is required (such as email verification)
    const { requiresVerification, verification } = await (0, validate_verification_1.validateVerification)(container, {
        actor_type: actorType,
        auth_provider: authProvider,
        auth_identity: authIdentity,
    });
    if (requiresVerification && !verification?.verified_at) {
        return {
            verification_required: true,
            verification,
            token: actorlessToken,
        };
    }
    if (mfaChallenge) {
        return {
            mfa_required: true,
            mfa_challenge: mfaChallenge,
            token: actorlessToken,
        };
    }
    const token = await generateJwtTokenForAuthIdentity({
        authIdentity,
        actorType: actorType,
        authProvider: authProvider,
        container,
    }, {
        secret: http.jwtSecret,
        expiresIn: http.jwtExpiresIn,
        options: http.jwtOptions,
    });
    return { token };
}
async function generateJwtTokenForAuthIdentity({ authIdentity, actorType, authProvider, container, }, { secret, expiresIn, skipActorType, options, }) {
    const expiresIn_ = expiresIn ?? options?.expiresIn;
    const entityIdKey = `${actorType}_id`;
    const entityId = skipActorType
        ? undefined
        : authIdentity?.app_metadata?.[entityIdKey];
    const providerIdentity = !authProvider
        ? undefined
        : authIdentity.provider_identities?.filter((identity) => identity.provider === authProvider)[0];
    let roles;
    if (utils_1.FeatureFlag.isFeatureEnabled(rbac_1.default.key)) {
        if (container && entityId) {
            try {
                const query = container.resolve(utils_1.ContainerRegistrationKeys.QUERY);
                const { data: userRoles } = await query.graph({
                    entity: actorType,
                    fields: ["rbac_roles.id"],
                    filters: {
                        id: entityId,
                    },
                });
                if (userRoles?.[0]?.rbac_roles) {
                    roles = userRoles[0].rbac_roles.map((role) => role.id);
                }
            }
            catch {
                // ignore
            }
        }
    }
    return (0, utils_1.generateJwtToken)({
        actor_id: entityId ?? "",
        actor_type: actorType,
        auth_identity_id: authIdentity?.id ?? "",
        ...(authProvider ? { auth_provider: authProvider } : {}),
        app_metadata: {
            ...(authIdentity.app_metadata ?? {}),
            [entityIdKey]: entityId,
            roles,
        },
        user_metadata: providerIdentity?.user_metadata ?? {},
    }, {
        secret,
        expiresIn: expiresIn_,
        jwtOptions: options,
    });
}
