// @medusajs/medusa/dist/api/auth/utils/validate-verification.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateVerification = void 0;
const utils_1 = require("@medusajs/framework/utils");
// The method checks for whether a verification is required for the given auth provider and identity.
const validateVerification = async (container, { actor_type, auth_provider, auth_identity, }) => {
    const config = container.resolve(utils_1.ContainerRegistrationKeys.CONFIG_MODULE);
    const service = container.resolve(utils_1.Modules.AUTH);
    const verificationsForActor = config.projectConfig.http.authVerificationsPerActor?.[actor_type];
    if (!verificationsForActor || verificationsForActor.length === 0) {
        return {
            requiresVerification: false,
            verification: undefined,
        };
    }
    const verificationForAuthProvider = verificationsForActor.find((verification) => verification.auth_provider === auth_provider);
    // This auth provider does not require verification.
    if (!verificationForAuthProvider) {
        return {
            requiresVerification: false,
            verification: undefined,
        };
    }
    const providerIdentity = auth_identity.provider_identities?.filter((identity) => identity.provider === auth_provider)[0];
    if (!providerIdentity) {
        throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_FOUND, "Provider identity not found while checking verification");
    }
    const verification = await service.listAuthVerifications({
        auth_identity_id: auth_identity.id,
        entity_id: providerIdentity.entity_id,
        entity_type: verificationForAuthProvider.entity_type,
    });
    // We return whether the verification even configured/required, and the verification itself and its status.
    return {
        requiresVerification: true,
        verification: verification[0],
    };
};
exports.validateVerification = validateVerification;
