// @medusajs/core-flows/dist/user/steps/get-assignable-policies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAssignablePoliciesStep = exports.getAssignablePoliciesStepId = void 0;
const framework_1 = require("@medusajs/framework");
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
/**
 * @ignore
 * @featureFlag rbac
 * @since 2.16.0
 */
exports.getAssignablePoliciesStepId = "get-assignable-policies";
/**
 * Resolves the set of policies the actor is allowed to assign.
 *
 * @ignore
 * @featureFlag rbac
 * @since 2.16.0
 */
exports.getAssignablePoliciesStep = (0, workflows_sdk_1.createStep)(exports.getAssignablePoliciesStepId, async (data, { container }) => {
    const { actor_id, actor, filters, pagination } = data;
    const query = container.resolve(utils_1.ContainerRegistrationKeys.QUERY);
    const { data: actors } = await query.graph({
        entity: actor ?? "user",
        fields: ["rbac_roles.id"],
        filters: { id: actor_id },
    });
    const actorRoleIds = actors?.[0]?.rbac_roles?.map((r) => r.id).filter(Boolean) ?? [];
    if (!actorRoleIds.length) {
        return new workflows_sdk_1.StepResponse({ policies: [], count: 0 });
    }
    const { data: candidates } = await query.graph({
        entity: "rbac_policy",
        fields: ["id", "key", "resource", "operation", "description"],
        filters: {
            ...(filters ?? {}),
            resource: { $ne: null },
            operation: { $ne: null },
        },
    });
    const granted = await (0, framework_1.resolvePermissions)({
        roles: actorRoleIds,
        universe: (candidates ?? []).map((p) => ({
            resource: p.resource,
            operation: p.operation,
        })),
        container,
    });
    const assignable = [];
    for (const policy of candidates ?? []) {
        if (granted.has(`${policy.resource}:${policy.operation}`)) {
            assignable.push({
                id: policy.id,
                key: policy.key,
                resource: policy.resource,
                operation: policy.operation,
                description: policy.description ?? null,
            });
        }
    }
    const { skip = 0, take } = pagination ?? {};
    const page = typeof take === "number"
        ? assignable.slice(skip, skip + take)
        : assignable.slice(skip);
    return new workflows_sdk_1.StepResponse({ policies: page, count: assignable.length });
});
