// @medusajs/core-flows/dist/user/steps/get-assignable-roles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAssignableRolesStep = exports.getAssignableRolesStepId = void 0;
const framework_1 = require("@medusajs/framework");
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
/**
 * @ignore
 * @featureFlag rbac
 * @since 2.16.0
 */
exports.getAssignableRolesStepId = "get-assignable-roles";
/**
 * Resolves the set of roles the actor is allowed to assign.
 *
 * @ignore
 * @featureFlag rbac
 * @since 2.16.0
 */
exports.getAssignableRolesStep = (0, workflows_sdk_1.createStep)(exports.getAssignableRolesStepId, async (data, { container }) => {
    const { actor_id, actor, filters, pagination } = data;
    const query = container.resolve(utils_1.ContainerRegistrationKeys.QUERY);
    const { data: actors } = await query.graph({
        entity: actor ?? "user",
        fields: ["rbac_roles.id"],
        filters: { id: actor_id },
    });
    const actorRoleIds = actors?.[0]?.rbac_roles?.map((r) => r.id).filter(Boolean) ?? [];
    if (!actorRoleIds.length) {
        return new workflows_sdk_1.StepResponse({ roles: [], count: 0 });
    }
    const { data: candidates } = await query.graph({
        entity: "rbac_role",
        fields: [
            "id",
            "name",
            "description",
            "policies.resource",
            "policies.operation",
        ],
        filters: filters ?? {},
        pagination: pagination ?? {},
    });
    const assignable = [];
    for (const role of candidates ?? []) {
        const actions = (role.policies ?? []).filter((p) => p.resource != null && p.operation != null);
        const allowed = await (0, framework_1.hasPermission)({
            roles: actorRoleIds,
            actions,
            container,
        });
        if (allowed) {
            assignable.push({
                id: role.id,
                name: role.name,
                description: role.description ?? null,
            });
        }
    }
    return new workflows_sdk_1.StepResponse({ roles: assignable, count: assignable.length });
});
