// @medusajs/medusa/dist/api/auth/validators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthMfaGenerateRecoveryCodesRequest = exports.AuthMfaDisableFactorRequest = exports.AuthMfaVerifyFactorRequest = exports.AuthMfaCreateFactorRequest = exports.AuthMfaVerifyChallengeRequest = exports.VerificationConfirmRequest = exports.VerificationRequest = exports.ResetPasswordRequest = void 0;
const zod_1 = require("@medusajs/framework/zod");
exports.ResetPasswordRequest = zod_1.z.object({
    identifier: zod_1.z.string(),
    metadata: zod_1.z.record(zod_1.z.string(), zod_1.z.unknown()).optional().default({}),
});
exports.VerificationRequest = zod_1.z.object({
    entity_id: zod_1.z.string().min(1),
    entity_type: zod_1.z.string().min(1),
    code_provider: zod_1.z.string().min(1).default("token"),
    metadata: zod_1.z.record(zod_1.z.string(), zod_1.z.unknown()).optional().default({}),
});
exports.VerificationConfirmRequest = zod_1.z.object({
    code: zod_1.z.string().min(1),
    code_provider: zod_1.z.string().min(1).optional(),
});
exports.AuthMfaVerifyChallengeRequest = zod_1.z.object({
    method: zod_1.z.string().min(1),
    code: zod_1.z.string().min(1),
});
exports.AuthMfaCreateFactorRequest = zod_1.z.object({
    provider: zod_1.z.string().min(1),
    label: zod_1.z.string().nullable().optional(),
    issuer: zod_1.z.string().optional(),
    metadata: zod_1.z.record(zod_1.z.string(), zod_1.z.unknown()).nullable().optional(),
});
exports.AuthMfaVerifyFactorRequest = zod_1.z.object({
    code: zod_1.z.string().min(1),
});
exports.AuthMfaDisableFactorRequest = zod_1.z.object({
    method: zod_1.z.string().min(1).optional(),
    code: zod_1.z.string().min(1).optional(),
});
exports.AuthMfaGenerateRecoveryCodesRequest = zod_1.z.object({
    count: zod_1.z.number().int().min(1).max(50).optional(),
});
