// @medusajs/js-sdk/dist/admin/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Admin = void 0;
const api_key_js_1 = require("./api-key.js");
const campaign_js_1 = require("./campaign.js");
const claim_js_1 = require("./claim.js");
const currency_js_1 = require("./currency.js");
const customer_js_1 = require("./customer.js");
const customer_group_js_1 = require("./customer-group.js");
const draft_order_js_1 = require("./draft-order.js");
const exchange_js_1 = require("./exchange.js");
const fulfillment_js_1 = require("./fulfillment.js");
const fulfillment_provider_js_1 = require("./fulfillment-provider.js");
const fulfillment_set_js_1 = require("./fulfillment-set.js");
const inventory_item_js_1 = require("./inventory-item.js");
const invite_js_1 = require("./invite.js");
const notification_js_1 = require("./notification.js");
const order_js_1 = require("./order.js");
const order_edit_js_1 = require("./order-edit.js");
const payment_js_1 = require("./payment.js");
const payment_collection_js_1 = require("./payment-collection.js");
const plugin_js_1 = require("./plugin.js");
const price_list_js_1 = require("./price-list.js");
const price_preference_js_1 = require("./price-preference.js");
const product_js_1 = require("./product.js");
const product_category_js_1 = require("./product-category.js");
const product_collection_js_1 = require("./product-collection.js");
const product_tag_js_1 = require("./product-tag.js");
const product_type_js_1 = require("./product-type.js");
const product_variant_js_1 = require("./product-variant.js");
const promotion_js_1 = require("./promotion.js");
const refund_reasons_js_1 = require("./refund-reasons.js");
const region_js_1 = require("./region.js");
const reservation_js_1 = __importDefault(require("./reservation.js"));
const return_js_1 = require("./return.js");
const return_reason_js_1 = require("./return-reason.js");
const sales_channel_js_1 = require("./sales-channel.js");
const shipping_option_js_1 = require("./shipping-option.js");
const shipping_profile_js_1 = require("./shipping-profile.js");
const stock_location_js_1 = require("./stock-location.js");
const store_js_1 = require("./store.js");
const tax_provider_js_1 = require("./tax-provider.js");
const tax_rate_js_1 = require("./tax-rate.js");
const tax_region_js_1 = require("./tax-region.js");
const upload_js_1 = require("./upload.js");
const user_js_1 = require("./user.js");
const views_js_1 = require("./views.js");
const workflow_execution_js_1 = require("./workflow-execution.js");
const shipping_option_type_js_1 = require("./shipping-option-type.js");
const locale_js_1 = require("./locale.js");
const translation_js_1 = require("./translation.js");
const rbac_role_js_1 = require("./rbac-role.js");
const rbac_policy_js_1 = require("./rbac-policy.js");
/**
 * The admin client provides access to admin-related resources.
 */
class Admin {
    constructor(client) {
        this.invite = new invite_js_1.Invite(client);
        this.customer = new customer_js_1.Customer(client);
        this.productCollection = new product_collection_js_1.ProductCollection(client);
        this.productCategory = new product_category_js_1.ProductCategory(client);
        this.priceList = new price_list_js_1.PriceList(client);
        this.pricePreference = new price_preference_js_1.PricePreference(client);
        this.product = new product_js_1.Product(client);
        this.productType = new product_type_js_1.ProductType(client);
        this.upload = new upload_js_1.Upload(client);
        this.region = new region_js_1.Region(client);
        this.returnReason = new return_reason_js_1.ReturnReason(client);
        this.stockLocation = new stock_location_js_1.StockLocation(client);
        this.salesChannel = new sales_channel_js_1.SalesChannel(client);
        this.fulfillmentSet = new fulfillment_set_js_1.FulfillmentSet(client);
        this.fulfillment = new fulfillment_js_1.Fulfillment(client);
        this.fulfillmentProvider = new fulfillment_provider_js_1.FulfillmentProvider(client);
        this.shippingOption = new shipping_option_js_1.ShippingOption(client);
        this.shippingOptionType = new shipping_option_type_js_1.ShippingOptionType(client);
        this.shippingProfile = new shipping_profile_js_1.ShippingProfile(client);
        this.inventoryItem = new inventory_item_js_1.InventoryItem(client);
        this.notification = new notification_js_1.Notification(client);
        this.order = new order_js_1.Order(client);
        this.draftOrder = new draft_order_js_1.DraftOrder(client);
        this.orderEdit = new order_edit_js_1.OrderEdit(client);
        this.return = new return_js_1.Return(client);
        this.claim = new claim_js_1.Claim(client);
        this.taxRate = new tax_rate_js_1.TaxRate(client);
        this.taxRegion = new tax_region_js_1.TaxRegion(client);
        this.translation = new translation_js_1.Translation(client);
        this.store = new store_js_1.Store(client);
        this.productTag = new product_tag_js_1.ProductTag(client);
        this.user = new user_js_1.User(client);
        this.locale = new locale_js_1.Locale(client);
        this.currency = new currency_js_1.Currency(client);
        this.payment = new payment_js_1.Payment(client);
        this.productVariant = new product_variant_js_1.ProductVariant(client);
        this.refundReason = new refund_reasons_js_1.RefundReason(client);
        this.exchange = new exchange_js_1.Exchange(client);
        this.paymentCollection = new payment_collection_js_1.PaymentCollection(client);
        this.apiKey = new api_key_js_1.ApiKey(client);
        this.workflowExecution = new workflow_execution_js_1.WorkflowExecution(client);
        this.reservation = new reservation_js_1.default(client);
        this.customerGroup = new customer_group_js_1.CustomerGroup(client);
        this.promotion = new promotion_js_1.Promotion(client);
        this.campaign = new campaign_js_1.Campaign(client);
        this.plugin = new plugin_js_1.Plugin(client);
        this.taxProvider = new tax_provider_js_1.TaxProvider(client);
        this.views = new views_js_1.Views(client);
        this.rbacRole = new rbac_role_js_1.RbacRole(client);
        this.rbacPolicy = new rbac_policy_js_1.RbacPolicy(client);
    }
}
exports.Admin = Admin;
